// @carverjs/core/dist/chunk-R4J4O2UJ.mjs
import {
  useSceneStore
} from "./chunk-333G7WFK.mjs";

// src/systems/SceneManagerImpl.ts
var SceneManagerImpl = class {
  /** Navigate to a scene, clearing the stack */
  go(name, data, transition) {
    useSceneStore.getState().go(name, data, transition);
  }
  /** Push a scene onto the stack (current scene sleeps) */
  push(name, data, transition) {
    useSceneStore.getState().push(name, data, transition);
  }
  /** Pop the top scene, returning to the previous one */
  pop(transition) {
    useSceneStore.getState().pop(transition);
  }
  /** Replace the current scene without changing stack depth */
  replace(name, data, transition) {
    useSceneStore.getState().replace(name, data, transition);
  }
  /** Get the current active scene name */
  getCurrent() {
    const stack = useSceneStore.getState().stack;
    return stack[stack.length - 1] ?? null;
  }
  /** Get the full navigation stack */
  getStack() {
    return useSceneStore.getState().stack;
  }
  /** Whether a transition is currently in progress */
  isTransitioning() {
    return useSceneStore.getState().transition.active;
  }
  /** Set shared cross-scene data */
  setShared(key, value) {
    useSceneStore.getState().setShared(key, value);
  }
  /** Get shared cross-scene data */
  getShared(key) {
    return useSceneStore.getState().getShared(key);
  }
  /** Reset the store (called on destroy) */
  destroy() {
    useSceneStore.setState({
      scenes: /* @__PURE__ */ new Map(),
      stack: [],
      transition: {
        active: false,
        from: null,
        to: null,
        progress: 0,
        config: { type: "none" },
        swapped: false,
        targetStack: null
      },
      shared: {}
    });
  }
};
var _instance = null;
function getSceneManager() {
  if (!_instance) _instance = new SceneManagerImpl();
  return _instance;
}
function destroySceneManager() {
  _instance?.destroy();
  _instance = null;
}

export {
  getSceneManager,
  destroySceneManager
};
