// @carverjs/core/dist/chunk-HPYY6PNE.mjs
// src/hooks/useCamera.ts
import { useRef, useCallback } from "react";
import { useFrame, useThree } from "@react-three/fiber";
import { Vector3 } from "three";
function useCamera(options = {}) {
  const { follow } = options;
  const { camera } = useThree();
  const _targetPos = useRef(new Vector3());
  const _desiredPos = useRef(new Vector3());
  const _lookAtPos = useRef(new Vector3());
  const shakeRef = useRef({ active: false, intensity: 0, remaining: 0 });
  const transitionRef = useRef(null);
  useFrame((_, delta) => {
    if (follow?.target?.current) {
      const obj = follow.target.current;
      const offset = follow.offset ?? [0, 5, 10];
      const smoothing = follow.smoothing ?? 0.1;
      const shouldLookAt = follow.lookAt ?? true;
      const lookAtOffset = follow.lookAtOffset ?? [0, 0, 0];
      obj.getWorldPosition(_targetPos.current);
      _desiredPos.current.set(
        _targetPos.current.x + offset[0],
        _targetPos.current.y + offset[1],
        _targetPos.current.z + offset[2]
      );
      camera.position.lerp(_desiredPos.current, smoothing);
      if (shouldLookAt) {
        _lookAtPos.current.set(
          _targetPos.current.x + lookAtOffset[0],
          _targetPos.current.y + lookAtOffset[1],
          _targetPos.current.z + lookAtOffset[2]
        );
        camera.lookAt(_lookAtPos.current);
      }
    }
    if (transitionRef.current?.active) {
      const t = transitionRef.current;
      t.elapsed += delta;
      const progress = Math.min(t.elapsed / t.duration, 1);
      const eased = progress * progress * (3 - 2 * progress);
      camera.position.lerpVectors(t.from, t.to, eased);
      if (progress >= 1) {
        transitionRef.current = null;
      }
    }
    if (shakeRef.current.active) {
      shakeRef.current.remaining -= delta;
      if (shakeRef.current.remaining <= 0) {
        shakeRef.current.active = false;
      } else {
        const intensity = shakeRef.current.intensity;
        camera.position.x += (Math.random() - 0.5) * intensity;
        camera.position.y += (Math.random() - 0.5) * intensity;
      }
    }
  });
  const shake = useCallback((intensity = 0.3, duration = 0.3) => {
    shakeRef.current = { active: true, intensity, remaining: duration };
  }, []);
  const moveTo = useCallback(
    (position, duration = 1) => {
      transitionRef.current = {
        active: true,
        from: camera.position.clone(),
        to: new Vector3(...position),
        duration,
        elapsed: 0
      };
    },
    [camera]
  );
  const lookAtFn = useCallback(
    (target) => {
      camera.lookAt(new Vector3(...target));
    },
    [camera]
  );
  return { shake, moveTo, lookAt: lookAtFn };
}

// src/hooks/useAnimation.ts
import { useRef as useRef2, useEffect, useCallback as useCallback2 } from "react";
import { useAnimations } from "@react-three/drei";
import {
  LoopRepeat,
  LoopOnce
} from "three";
function useAnimation(clips, options = {}) {
  const { clipName, paused = false, speed = 1, loop = true } = options;
  const { ref, mixer, actions, names } = useAnimations(clips);
  const previousAction = useRef2(null);
  useEffect(() => {
    if (!clipName || !actions[clipName]) return;
    const action = actions[clipName];
    if (previousAction.current && previousAction.current !== action) {
      previousAction.current.fadeOut(0.2);
    }
    action.reset();
    action.setLoop(loop ? LoopRepeat : LoopOnce, loop ? Infinity : 1);
    action.clampWhenFinished = !loop;
    action.fadeIn(0.2).play();
    previousAction.current = action;
  }, [clipName, actions, loop]);
  useEffect(() => {
    if (previousAction.current) {
      previousAction.current.paused = paused;
    }
  }, [paused]);
  useEffect(() => {
    mixer.timeScale = speed;
  }, [speed, mixer]);
  const play = useCallback2(
    (name) => {
      const action = actions[name];
      if (!action) return;
      if (previousAction.current && previousAction.current !== action) {
        previousAction.current.fadeOut(0.2);
      }
      action.reset();
      action.setLoop(loop ? LoopRepeat : LoopOnce, loop ? Infinity : 1);
      action.clampWhenFinished = !loop;
      action.fadeIn(0.2).play();
      previousAction.current = action;
    },
    [actions, loop]
  );
  const stopAll = useCallback2(() => {
    mixer.stopAllAction();
    previousAction.current = null;
  }, [mixer]);
  const crossFadeTo = useCallback2(
    (name, duration = 0.3) => {
      const action = actions[name];
      if (!action) return;
      if (previousAction.current) {
        previousAction.current.crossFadeTo(action, duration, true);
      }
      action.reset();
      action.setLoop(loop ? LoopRepeat : LoopOnce, loop ? Infinity : 1);
      action.clampWhenFinished = !loop;
      action.play();
      previousAction.current = action;
    },
    [actions, loop]
  );
  return {
    ref,
    clipNames: names,
    activeAction: previousAction.current,
    mixer,
    actions,
    play,
    stopAll,
    crossFadeTo
  };
}

export {
  useCamera,
  useAnimation
};
