// @carverjs/core/dist/chunk-332ISZNQ.mjs
// src/systems/GridCollisionManager.ts
var GridCollisionManager = class {
  constructor() {
    this._grid = new Int32Array(0);
    this._width = 0;
    this._height = 0;
    this._cellSize = 1;
    this._originX = 0;
    this._originY = 0;
  }
  create(config) {
    this._width = config.width;
    this._height = config.height;
    this._cellSize = config.cellSize ?? 1;
    this._originX = config.origin?.[0] ?? 0;
    this._originY = config.origin?.[1] ?? 0;
    this._grid = new Int32Array(this._width * this._height);
  }
  _inBounds(x, y) {
    return x >= 0 && x < this._width && y >= 0 && y < this._height;
  }
  _index(x, y) {
    return y * this._width + x;
  }
  setCell(x, y, value) {
    if (!this._inBounds(x, y)) return;
    this._grid[this._index(x, y)] = value;
  }
  getCell(x, y) {
    if (!this._inBounds(x, y)) return 0;
    return this._grid[this._index(x, y)];
  }
  clearCell(x, y) {
    this.setCell(x, y, 0);
  }
  isCellOccupied(x, y) {
    return this.getCell(x, y) !== 0;
  }
  worldToGrid(worldX, worldY) {
    return [
      Math.floor((worldX - this._originX) / this._cellSize),
      Math.floor((worldY - this._originY) / this._cellSize)
    ];
  }
  gridToWorld(gridX, gridY) {
    return [
      this._originX + gridX * this._cellSize + this._cellSize / 2,
      this._originY + gridY * this._cellSize + this._cellSize / 2
    ];
  }
  getNeighbors4(x, y) {
    return [
      this.getCell(x, y - 1),
      // up
      this.getCell(x + 1, y),
      // right
      this.getCell(x, y + 1),
      // down
      this.getCell(x - 1, y)
      // left
    ];
  }
  getNeighbors8(x, y) {
    return [
      this.getCell(x, y - 1),
      // up
      this.getCell(x + 1, y - 1),
      // up-right
      this.getCell(x + 1, y),
      // right
      this.getCell(x + 1, y + 1),
      // down-right
      this.getCell(x, y + 1),
      // down
      this.getCell(x - 1, y + 1),
      // down-left
      this.getCell(x - 1, y),
      // left
      this.getCell(x - 1, y - 1)
      // up-left
    ];
  }
  clearAll() {
    this._grid.fill(0);
  }
  get width() {
    return this._width;
  }
  get height() {
    return this._height;
  }
  get cellSize() {
    return this._cellSize;
  }
};
var _instance = null;
function getGridCollisionManager() {
  if (!_instance) _instance = new GridCollisionManager();
  return _instance;
}
function destroyGridCollisionManager() {
  _instance = null;
}

export {
  getGridCollisionManager,
  destroyGridCollisionManager
};
