// @carverjs/core/dist/components.mjs
import {
  useAnimation,
  useCamera
} from "./chunk-HPYY6PNE.mjs";
import {
  destroySceneManager
} from "./chunk-R4J4O2UJ.mjs";
import {
  PhysicsBodyWrapper,
  PhysicsGate,
  destroyAudioManager,
  destroyCollisionManager,
  destroyInputManager,
  destroyParticleManager,
  destroyTweenManager,
  detectAssetType,
  getActorRegistry,
  getAssetManager,
  getAudioManager,
  getCollisionManager,
  getInputManager,
  getParticleManager,
  getParticlePreset,
  getTweenManager
} from "./chunk-4JMHDLMK.mjs";
import {
  useGameStore,
  useSceneStore
} from "./chunk-333G7WFK.mjs";

// src/components/Game.tsx
import { useMemo, useEffect } from "react";
import { Canvas, useFrame, useThree } from "@react-three/fiber";
import { Sky, Environment } from "@react-three/drei";

// src/components/GameContext.ts
import { createContext, useContext } from "react";
var GameContext = createContext(null);
function useGameContext() {
  const ctx = useContext(GameContext);
  if (!ctx) {
    throw new Error("useGameContext must be used inside a <Game> component.");
  }
  return ctx;
}

// src/components/Game.tsx
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
var MAX_DELTA = 0.1;
function GameLoopTick() {
  useFrame((_, rawDelta) => {
    useGameStore.getState()._tick(Math.min(rawDelta, MAX_DELTA));
  }, -5);
  return null;
}
function CollisionFlush() {
  const { mode } = useGameContext();
  useEffect(() => {
    getCollisionManager().setMode(mode);
  }, [mode]);
  useEffect(() => {
    return () => destroyCollisionManager();
  }, []);
  useFrame(() => {
    getCollisionManager().tick();
  }, -25);
  return null;
}
function AudioFlush() {
  const camera = useThree((s) => s.camera);
  useEffect(() => {
    const mgr = getAudioManager();
    mgr.init();
    return () => destroyAudioManager();
  }, []);
  useEffect(() => {
    getAudioManager().setListenerSource(camera);
  }, [camera]);
  useFrame(() => {
    getAudioManager().flush(useGameStore.getState().phase);
  }, -48);
  return null;
}
function TweenFlush() {
  useEffect(() => {
    return () => destroyTweenManager();
  }, []);
  useFrame((_, rawDelta) => {
    getTweenManager().tick(Math.min(rawDelta, MAX_DELTA));
  }, -15);
  return null;
}
function ParticleFlush() {
  const { mode } = useGameContext();
  useEffect(() => {
    getParticleManager().setMode(mode);
  }, [mode]);
  useEffect(() => {
    return () => destroyParticleManager();
  }, []);
  useFrame((_, rawDelta) => {
    const { phase } = useGameStore.getState();
    if (phase === "loading" || phase === "paused") return;
    getParticleManager().tick(Math.min(rawDelta, MAX_DELTA));
  }, -12);
  return null;
}
function InputFlush() {
  const gl = useThree((s) => s.gl);
  useEffect(() => {
    const mgr = getInputManager();
    mgr.attach(gl.domElement);
    return () => destroyInputManager();
  }, [gl.domElement]);
  useFrame(() => {
    getInputManager().flush();
  }, -50);
  return null;
}
function Game({
  mode = "3d",
  ambientLightProps,
  directionalLightProps,
  skyProps,
  environmentProps,
  children,
  ...canvasProps
}) {
  const is2D = mode === "2d";
  const contextValue = useMemo(() => ({ mode }), [mode]);
  return /* @__PURE__ */ jsx(Canvas, { shadows: !is2D, ...canvasProps, children: /* @__PURE__ */ jsxs(GameContext.Provider, { value: contextValue, children: [
    /* @__PURE__ */ jsx(GameLoopTick, {}),
    /* @__PURE__ */ jsx(InputFlush, {}),
    /* @__PURE__ */ jsx(CollisionFlush, {}),
    /* @__PURE__ */ jsx(AudioFlush, {}),
    /* @__PURE__ */ jsx(TweenFlush, {}),
    /* @__PURE__ */ jsx(ParticleFlush, {}),
    /* @__PURE__ */ jsx("ambientLight", { intensity: is2D ? 1 : 0.4, ...ambientLightProps }),
    !is2D && /* @__PURE__ */ jsxs(Fragment, { children: [
      /* @__PURE__ */ jsx(
        "directionalLight",
        {
          position: [10, 15, 10],
          intensity: 1,
          castShadow: true,
          "shadow-mapSize-width": 1024,
          "shadow-mapSize-height": 1024,
          ...directionalLightProps
        }
      ),
      /* @__PURE__ */ jsx(Sky, { sunPosition: [100, 20, 100], ...skyProps }),
      /* @__PURE__ */ jsx(Environment, { preset: "sunset", background: true, ...environmentProps })
    ] }),
    children
  ] }) });
}

// src/components/World.tsx
import { Children, isValidElement } from "react";
import {
  OrthographicCamera as OrthographicCamera2,
  PerspectiveCamera as PerspectiveCamera2,
  OrbitControls as OrbitControls2
} from "@react-three/drei";

// src/components/Camera.tsx
import { forwardRef, useEffect as useEffect2, useRef } from "react";
import {
  OrthographicCamera,
  PerspectiveCamera,
  OrbitControls,
  MapControls,
  PointerLockControls
} from "@react-three/drei";
import { useFrame as useFrame2 } from "@react-three/fiber";
import { Vector3 } from "three";
import { Fragment as Fragment2, jsx as jsx2, jsxs as jsxs2 } from "react/jsx-runtime";
function OrbitFollow({ follow }) {
  const controlsRef = useRef(null);
  const _targetPos = useRef(new Vector3());
  useFrame2(() => {
    if (!controlsRef.current || !follow.target?.current) return;
    const obj = follow.target.current;
    const lookAtOffset = follow.lookAtOffset ?? [0, 0, 0];
    obj.getWorldPosition(_targetPos.current);
    _targetPos.current.x += lookAtOffset[0];
    _targetPos.current.y += lookAtOffset[1];
    _targetPos.current.z += lookAtOffset[2];
    const smoothing = follow.smoothing ?? 0.1;
    controlsRef.current.target.lerp(_targetPos.current, smoothing);
    controlsRef.current.update();
  });
  return /* @__PURE__ */ jsx2(OrbitControls, { ref: controlsRef, enableZoom: true });
}
var Camera = forwardRef(
  ({
    type = "perspective",
    controls = "none",
    follow,
    perspectiveProps,
    orthographicProps,
    orbitControlsProps,
    mapControlsProps,
    pointerLockControlsProps,
    onReady
  }, ref) => {
    const useOrbitFollow = controls === "orbit" && !!follow;
    const api = useCamera({ follow: useOrbitFollow ? void 0 : follow });
    useEffect2(() => {
      onReady?.(api);
    }, [onReady, api]);
    return /* @__PURE__ */ jsxs2(Fragment2, { children: [
      type === "perspective" ? /* @__PURE__ */ jsx2(
        PerspectiveCamera,
        {
          ref,
          makeDefault: true,
          position: [0, 5, 10],
          fov: 75,
          near: 0.1,
          far: 1e3,
          ...perspectiveProps
        }
      ) : /* @__PURE__ */ jsx2(
        OrthographicCamera,
        {
          ref,
          makeDefault: true,
          position: [0, 0, 100],
          zoom: 50,
          near: 0.1,
          far: 1e3,
          ...orthographicProps
        }
      ),
      controls === "orbit" && (useOrbitFollow ? /* @__PURE__ */ jsx2(OrbitFollow, { follow }) : /* @__PURE__ */ jsx2(OrbitControls, { enableZoom: true, ...orbitControlsProps })),
      controls === "map" && /* @__PURE__ */ jsx2(MapControls, { enableZoom: true, ...mapControlsProps }),
      controls === "pointerlock" && /* @__PURE__ */ jsx2(PointerLockControls, { ...pointerLockControlsProps })
    ] });
  }
);
Camera.displayName = "Camera";

// src/components/World.tsx
import { jsx as jsx3, jsxs as jsxs3 } from "react/jsx-runtime";
function World({
  cameraProps2D,
  cameraProps3D,
  orbitControlsProps,
  physics,
  active = true,
  children
}) {
  const { mode } = useGameContext();
  const is2D = mode === "2d";
  const hasCustomCamera = Children.toArray(children).some(
    (child) => isValidElement(child) && child.type === Camera
  );
  return /* @__PURE__ */ jsxs3("group", { visible: active, children: [
    !hasCustomCamera && (is2D ? /* @__PURE__ */ jsx3(
      OrthographicCamera2,
      {
        makeDefault: true,
        position: [0, 0, 100],
        zoom: 50,
        near: 0.1,
        far: 1e3,
        ...cameraProps2D
      }
    ) : /* @__PURE__ */ jsx3(
      PerspectiveCamera2,
      {
        makeDefault: true,
        position: [0, 5, 10],
        fov: 75,
        near: 0.1,
        far: 1e3,
        ...cameraProps3D
      }
    )),
    !is2D && !hasCustomCamera && /* @__PURE__ */ jsx3(OrbitControls2, { enableZoom: true, ...orbitControlsProps }),
    /* @__PURE__ */ jsx3(PhysicsGate, { config: physics, mode, children })
  ] });
}

// src/components/Actor.tsx
import { forwardRef as forwardRef2, useMemo as useMemo2, useEffect as useEffect3, useRef as useRef2, useCallback } from "react";
import {
  useGLTF,
  useTexture,
  SpriteAnimator,
  Billboard
} from "@react-three/drei";
import { useFrame as useFrame3 } from "@react-three/fiber";
import { SkeletonUtils } from "three-stdlib";
import { Mesh, RepeatWrapping } from "three";
import { jsx as jsx4, jsxs as jsxs4 } from "react/jsx-runtime";
var DEFAULT_GEOMETRY_ARGS = {
  box: [1, 1, 1],
  sphere: [0.5, 32, 32],
  cylinder: [0.5, 0.5, 1, 32],
  cone: [0.5, 1, 32],
  torus: [0.5, 0.2, 16, 32],
  plane: [1, 1],
  circle: [0.5, 32],
  capsule: [0.5, 1, 4, 16],
  ring: [0.3, 0.5, 32]
};
function ModelRenderer({
  src,
  useDraco,
  useMeshopt,
  animationName,
  animationPaused,
  animationSpeed,
  animationLoop,
  castShadow,
  receiveShadow
}) {
  const { scene, animations } = useGLTF(src, useDraco, useMeshopt);
  const clone = useMemo2(() => {
    return animations.length > 0 ? SkeletonUtils.clone(scene) : scene.clone(true);
  }, [scene, animations]);
  const { ref } = useAnimation(animations, {
    clipName: animationName,
    paused: animationPaused,
    speed: animationSpeed,
    loop: animationLoop ?? true
  });
  useEffect3(() => {
    clone.traverse((child) => {
      if (child instanceof Mesh) {
        child.castShadow = castShadow ?? false;
        child.receiveShadow = receiveShadow ?? false;
      }
    });
  }, [clone, castShadow, receiveShadow]);
  return /* @__PURE__ */ jsx4("primitive", { object: clone, ref });
}
function SpriteRenderer({
  src,
  atlasUrl,
  billboard = true,
  billboardLockX,
  billboardLockY,
  billboardLockZ,
  spriteAnimation
}) {
  if (atlasUrl && spriteAnimation) {
    const content = /* @__PURE__ */ jsx4(
      SpriteAnimator,
      {
        textureImageURL: src,
        textureDataURL: atlasUrl,
        autoPlay: true,
        ...spriteAnimation
      }
    );
    if (billboard) {
      return /* @__PURE__ */ jsx4(
        Billboard,
        {
          follow: true,
          lockX: billboardLockX,
          lockY: billboardLockY,
          lockZ: billboardLockZ,
          children: content
        }
      );
    }
    return content;
  }
  if (spriteAnimation && !atlasUrl) {
    const content = /* @__PURE__ */ jsx4(
      AnimatedGridSprite,
      {
        src,
        numberOfFrames: spriteAnimation.numberOfFrames ?? 1,
        fps: spriteAnimation.fps ?? 10,
        loop: spriteAnimation.loop ?? true,
        startFrame: spriteAnimation.startFrame,
        endFrame: spriteAnimation.endFrame,
        autoPlay: spriteAnimation.autoPlay ?? true,
        flipX: spriteAnimation.flipX
      }
    );
    if (billboard) {
      return /* @__PURE__ */ jsx4(
        Billboard,
        {
          follow: true,
          lockX: billboardLockX,
          lockY: billboardLockY,
          lockZ: billboardLockZ,
          children: content
        }
      );
    }
    return content;
  }
  return /* @__PURE__ */ jsx4(
    StaticSprite,
    {
      src,
      billboard,
      billboardLockX,
      billboardLockY,
      billboardLockZ
    }
  );
}
function AnimatedGridSprite({
  src,
  numberOfFrames,
  fps = 10,
  loop = true,
  startFrame = 0,
  endFrame,
  autoPlay = true,
  flipX = false
}) {
  const texture = useTexture(src);
  const matRef = useRef2(null);
  const frameRef = useRef2(startFrame);
  const elapsedRef = useRef2(0);
  const grid = useMemo2(() => {
    const img = texture.image;
    const cols = Math.round(Math.sqrt(numberOfFrames * (img.width / img.height)));
    const rows = Math.ceil(numberOfFrames / cols);
    return { cols, rows };
  }, [texture, numberOfFrames]);
  useEffect3(() => {
    texture.wrapS = RepeatWrapping;
    texture.wrapT = RepeatWrapping;
    const repeatX = 1 / grid.cols;
    const repeatY = 1 / grid.rows;
    texture.repeat.set(flipX ? -repeatX : repeatX, repeatY);
    texture.needsUpdate = true;
  }, [texture, grid, flipX]);
  useFrame3((_, delta) => {
    if (!autoPlay || !matRef.current?.map) return;
    elapsedRef.current += delta;
    const interval = 1 / fps;
    if (elapsedRef.current < interval) return;
    elapsedRef.current -= interval;
    const last = endFrame ?? numberOfFrames - 1;
    const frame = frameRef.current;
    const col = frame % grid.cols;
    const row = Math.floor(frame / grid.cols);
    const map = matRef.current.map;
    map.offset.x = (flipX ? col + 1 : col) / grid.cols;
    map.offset.y = 1 - (row + 1) / grid.rows;
    let next = frame + 1;
    if (next > last) {
      next = loop ? startFrame : last;
    }
    frameRef.current = next;
  });
  return /* @__PURE__ */ jsx4("sprite", { children: /* @__PURE__ */ jsx4("spriteMaterial", { ref: matRef, map: texture }) });
}
function StaticSprite({
  src,
  billboard = true,
  billboardLockX,
  billboardLockY,
  billboardLockZ
}) {
  const texture = useTexture(src);
  const content = /* @__PURE__ */ jsx4("sprite", { children: /* @__PURE__ */ jsx4("spriteMaterial", { map: texture }) });
  if (billboard) {
    return /* @__PURE__ */ jsx4(
      Billboard,
      {
        follow: true,
        lockX: billboardLockX,
        lockY: billboardLockY,
        lockZ: billboardLockZ,
        children: content
      }
    );
  }
  return content;
}
function PrimitiveRenderer({
  shape = "box",
  geometryArgs,
  materialType = "standard",
  color = "#6366f1",
  materialProps,
  wireframe,
  castShadow,
  receiveShadow
}) {
  const args = geometryArgs ?? DEFAULT_GEOMETRY_ARGS[shape ?? "box"];
  const geometryElement = {
    box: /* @__PURE__ */ jsx4("boxGeometry", { args }),
    sphere: /* @__PURE__ */ jsx4("sphereGeometry", { args }),
    cylinder: /* @__PURE__ */ jsx4("cylinderGeometry", { args }),
    cone: /* @__PURE__ */ jsx4("coneGeometry", { args }),
    torus: /* @__PURE__ */ jsx4("torusGeometry", { args }),
    plane: /* @__PURE__ */ jsx4("planeGeometry", { args }),
    circle: /* @__PURE__ */ jsx4("circleGeometry", { args }),
    capsule: /* @__PURE__ */ jsx4("capsuleGeometry", { args }),
    ring: /* @__PURE__ */ jsx4("ringGeometry", { args })
  }[shape ?? "box"];
  const sharedMaterialProps = { color, wireframe, ...materialProps };
  const materialElement = {
    standard: /* @__PURE__ */ jsx4("meshStandardMaterial", { ...sharedMaterialProps }),
    basic: /* @__PURE__ */ jsx4("meshBasicMaterial", { ...sharedMaterialProps }),
    phong: /* @__PURE__ */ jsx4("meshPhongMaterial", { ...sharedMaterialProps }),
    lambert: /* @__PURE__ */ jsx4("meshLambertMaterial", { ...sharedMaterialProps }),
    toon: /* @__PURE__ */ jsx4("meshToonMaterial", { ...sharedMaterialProps })
  }[materialType ?? "standard"];
  return /* @__PURE__ */ jsxs4("mesh", { castShadow, receiveShadow, children: [
    geometryElement,
    materialElement
  ] });
}
var Actor = forwardRef2((props, ref) => {
  const {
    type,
    name,
    position,
    rotation,
    scale,
    size,
    visible = true,
    userData,
    renderOrder,
    castShadow,
    receiveShadow,
    physics,
    networked,
    children,
    onClick,
    onContextMenu,
    onDoubleClick,
    onPointerUp,
    onPointerDown,
    onPointerOver,
    onPointerOut,
    onPointerEnter,
    onPointerLeave,
    onPointerMove,
    onPointerMissed,
    onWheel
  } = props;
  const internalRef = useRef2(null);
  const setRefs = useCallback((node) => {
    internalRef.current = node;
    if (typeof ref === "function") ref(node);
    else if (ref) ref.current = node;
  }, [ref]);
  useEffect3(() => {
    if (!name || !internalRef.current) return;
    const registry = getActorRegistry();
    registry.register(name, {
      id: name,
      object3D: internalRef.current,
      rigidBody: void 0,
      // Will be set by PhysicsBodyWrapper
      userData: { ...userData, networked }
    });
    return () => {
      registry.unregister(name);
    };
  }, [name, networked]);
  return /* @__PURE__ */ jsx4(PhysicsBodyWrapper, { physics, name, userData, children: /* @__PURE__ */ jsxs4(
    "group",
    {
      ref: setRefs,
      name,
      position,
      rotation,
      scale: scale ?? (size != null ? [size, size, size] : void 0),
      visible,
      userData,
      renderOrder,
      onClick,
      onContextMenu,
      onDoubleClick,
      onPointerUp,
      onPointerDown,
      onPointerOver,
      onPointerOut,
      onPointerEnter,
      onPointerLeave,
      onPointerMove,
      onPointerMissed,
      onWheel,
      children: [
        type === "model" && /* @__PURE__ */ jsx4(
          ModelRenderer,
          {
            src: props.src,
            useDraco: props.useDraco,
            useMeshopt: props.useMeshopt,
            animationName: props.animationName,
            animationPaused: props.animationPaused,
            animationSpeed: props.animationSpeed,
            animationLoop: props.animationLoop,
            castShadow,
            receiveShadow
          }
        ),
        type === "sprite" && /* @__PURE__ */ jsx4(
          SpriteRenderer,
          {
            src: props.src,
            atlasUrl: props.atlasUrl,
            billboard: props.billboard,
            billboardLockX: props.billboardLockX,
            billboardLockY: props.billboardLockY,
            billboardLockZ: props.billboardLockZ,
            spriteAnimation: props.spriteAnimation
          }
        ),
        type === "primitive" && /* @__PURE__ */ jsx4(
          PrimitiveRenderer,
          {
            shape: props.shape,
            geometryArgs: props.geometryArgs,
            materialType: props.materialType,
            color: props.color,
            materialProps: props.materialProps,
            wireframe: props.wireframe,
            castShadow,
            receiveShadow
          }
        ),
        children
      ]
    }
  ) });
});
Actor.displayName = "Actor";

// src/components/AudioListener.tsx
import { useEffect as useEffect4 } from "react";
import { useFrame as useFrame4 } from "@react-three/fiber";
function AudioListener({ listenerRef }) {
  useFrame4(() => {
    getAudioManager().setCustomListener(listenerRef?.current ?? null);
  }, -49);
  useEffect4(() => {
    return () => {
      getAudioManager().setCustomListener(null);
    };
  }, []);
  return null;
}

// src/components/SceneManager.tsx
import {
  useEffect as useEffect5,
  useRef as useRef3,
  useMemo as useMemo3,
  lazy,
  Suspense,
  Component
} from "react";
import { useFrame as useFrame5 } from "@react-three/fiber";
import { Color } from "three";
import { Fragment as Fragment3, jsx as jsx5, jsxs as jsxs5 } from "react/jsx-runtime";
var FADE_VERTEX = (
  /* glsl */
  `
void main() {
  gl_Position = vec4(position.xy, 0.0, 1.0);
}
`
);
var FADE_FRAGMENT = (
  /* glsl */
  `
uniform float uProgress;
uniform vec3 uColor;

void main() {
  // 0\u21920.5: fade to color (opacity 0\u21921)
  // 0.5\u21921: fade from color (opacity 1\u21920)
  float opacity = uProgress < 0.5
    ? uProgress * 2.0
    : (1.0 - uProgress) * 2.0;
  gl_FragColor = vec4(uColor, opacity);
}
`
);
function SceneTransitionFlush() {
  useFrame5((_, rawDelta) => {
    const delta = Math.min(rawDelta, 0.1);
    useSceneStore.getState()._tickTransition(delta);
  }, -15);
  return null;
}
function TransitionOverlay() {
  const active = useSceneStore((s) => s.transition.active);
  const config = useSceneStore((s) => s.transition.config);
  const materialRef = useRef3(null);
  const colorVec = useMemo3(() => {
    return new Color(config.color ?? "#000000");
  }, [config.color]);
  useFrame5(() => {
    if (!materialRef.current) return;
    const { progress } = useSceneStore.getState().transition;
    materialRef.current.uniforms.uProgress.value = progress;
  }, -14);
  if (!active || config.type === "none") return null;
  const isCustom = config.type === "custom" && config.fragmentShader;
  return /* @__PURE__ */ jsxs5("mesh", { renderOrder: 9999, frustumCulled: false, children: [
    /* @__PURE__ */ jsx5("planeGeometry", { args: [2, 2] }),
    /* @__PURE__ */ jsx5(
      "shaderMaterial",
      {
        ref: materialRef,
        transparent: true,
        depthTest: false,
        depthWrite: false,
        uniforms: {
          uProgress: { value: 0 },
          uColor: { value: colorVec },
          ...isCustom ? config.uniforms : {}
        },
        vertexShader: isCustom && config.vertexShader ? config.vertexShader : FADE_VERTEX,
        fragmentShader: isCustom ? config.fragmentShader : FADE_FRAGMENT
      }
    )
  ] });
}
var SceneErrorBoundary = class extends Component {
  constructor() {
    super(...arguments);
    this.state = { hasError: false };
  }
  static getDerivedStateFromError() {
    return { hasError: true };
  }
  componentDidCatch(error) {
    console.error(
      `[CarverJS Scene] Error in scene "${this.props.name}":`,
      error
    );
    useSceneStore.getState()._setStatus(this.props.name, "destroyed");
  }
  render() {
    if (this.state.hasError) return null;
    return this.props.children;
  }
};
var _lazyCache = /* @__PURE__ */ new Map();
function SceneContent({ entry }) {
  const { config, data } = entry;
  const SceneComponent = useMemo3(() => {
    if (config.component) return config.component;
    if (config.loader) {
      let cached = _lazyCache.get(config.name);
      if (!cached) {
        cached = lazy(config.loader);
        _lazyCache.set(config.name, cached);
      }
      return cached;
    }
    return null;
  }, [config]);
  if (!SceneComponent) return null;
  return /* @__PURE__ */ jsx5(SceneComponent, { data });
}
function SceneRenderer({
  entry,
  loadingFallback
}) {
  const { config, status } = entry;
  const name = config.name;
  const visible = status === "running" || status === "shutting_down" || status === "paused";
  const mounted = status === "running" || status === "shutting_down" || status === "paused" || status === "sleeping" || status === "preloading";
  if (!mounted) return null;
  return /* @__PURE__ */ jsx5("group", { visible, children: /* @__PURE__ */ jsx5(SceneErrorBoundary, { name, children: /* @__PURE__ */ jsx5(Suspense, { fallback: loadingFallback ?? null, children: /* @__PURE__ */ jsx5(SceneContent, { entry }) }) }) }, name);
}
function SceneManager({
  initial,
  loadingFallback,
  persistent,
  children
}) {
  const initializedRef = useRef3(false);
  const scenes = useSceneStore((s) => s.scenes);
  useEffect5(() => {
    if (initializedRef.current) return;
    initializedRef.current = true;
    const id = requestAnimationFrame(() => {
      useSceneStore.getState()._startInitial(initial);
    });
    return () => cancelAnimationFrame(id);
  }, []);
  useEffect5(() => {
    return () => destroySceneManager();
  }, []);
  const sceneEntries = useMemo3(() => {
    const entries = [];
    for (const [, entry] of scenes) {
      if (entry.status !== "created" && entry.status !== "destroyed") {
        entries.push(entry);
      }
    }
    return entries;
  }, [scenes]);
  return /* @__PURE__ */ jsxs5(Fragment3, { children: [
    children,
    /* @__PURE__ */ jsx5(SceneTransitionFlush, {}),
    sceneEntries.map((entry) => /* @__PURE__ */ jsx5(
      SceneRenderer,
      {
        entry,
        loadingFallback
      },
      entry.config.name
    )),
    /* @__PURE__ */ jsx5(TransitionOverlay, {}),
    persistent
  ] });
}

// src/components/Scene.tsx
import { useEffect as useEffect6, useRef as useRef4 } from "react";
function Scene(props) {
  const { name, component, loader, transition, persistent } = props;
  const configRef = useRef4({
    name,
    component,
    loader,
    transition,
    persistent
  });
  useEffect6(() => {
    useSceneStore.getState()._registerScene(configRef.current);
    return () => {
      useSceneStore.getState()._unregisterScene(configRef.current.name);
    };
  }, []);
  return null;
}

// src/components/AssetLoader.tsx
import { useState, useEffect as useEffect7, useRef as useRef5 } from "react";
import { useGLTF as useGLTF2, useTexture as useTexture2, Html } from "@react-three/drei";
import { Fragment as Fragment4, jsx as jsx6, jsxs as jsxs6 } from "react/jsx-runtime";
var IDLE_PROGRESS = {
  phase: "idle",
  loaded: 0,
  total: 0,
  progress: 0,
  currentAsset: null,
  currentGroup: null,
  errors: []
};
function resolveEntries(manifest) {
  if (Array.isArray(manifest)) {
    return { entries: manifest, baseUrl: "" };
  }
  return { entries: manifest.assets, baseUrl: manifest.baseUrl ?? "" };
}
function resolveUrl(url, baseUrl) {
  if (url.startsWith("http") || url.startsWith("/") || url.startsWith("data:")) {
    return url;
  }
  return baseUrl + url;
}
function AssetLoader({
  manifest,
  fallback,
  concurrency,
  retries,
  retryDelay,
  timeout,
  minLoadTime = 0,
  onComplete,
  onError,
  onProgress,
  children
}) {
  const [progress, setProgress] = useState(IDLE_PROGRESS);
  const [ready, setReady] = useState(false);
  const startRef = useRef5(0);
  useEffect7(() => {
    const manager = getAssetManager();
    manager.configure({
      ...concurrency !== void 0 && { maxConcurrent: concurrency },
      ...retries !== void 0 && { retries },
      ...retryDelay !== void 0 && { retryDelay },
      ...timeout !== void 0 && { timeout }
    });
    if (typeof manifest === "string") {
      fetch(manifest).then((r) => r.json()).then((json) => startLoad(json)).catch((err) => {
        const errorProgress = {
          ...IDLE_PROGRESS,
          phase: "error",
          errors: [
            {
              key: manifest,
              url: manifest,
              message: err instanceof Error ? err.message : String(err),
              retries: 0,
              recoverable: false
            }
          ]
        };
        setProgress(errorProgress);
        onError?.(errorProgress.errors);
      });
      return;
    }
    startLoad(manifest);
    function startLoad(m) {
      const { entries, baseUrl } = resolveEntries(m);
      for (const entry of entries) {
        const url = resolveUrl(entry.url, baseUrl);
        const type = entry.type ?? detectAssetType(url);
        if (type === "gltf") {
          useGLTF2.preload(
            url,
            entry.loaderOptions?.draco,
            entry.loaderOptions?.meshopt
          );
        }
        if (type === "texture") {
          useTexture2.preload(url);
        }
      }
      startRef.current = performance.now();
      manager.loadManifest(m, {
        onProgress: (p) => {
          const cloned = {
            ...p,
            errors: [...p.errors]
          };
          setProgress(cloned);
          onProgress?.(cloned);
        },
        onComplete: () => {
          const elapsed = performance.now() - startRef.current;
          const remaining = Math.max(0, minLoadTime - elapsed);
          setTimeout(() => {
            setReady(true);
            const state = useGameStore.getState();
            if (state.phase === "loading") {
              state.setPhase("playing");
            }
            onComplete?.();
          }, remaining);
        },
        onError
      });
    }
  }, []);
  return /* @__PURE__ */ jsxs6(Fragment4, { children: [
    !ready && fallback && /* @__PURE__ */ jsx6(Html, { fullscreen: true, zIndexRange: [100, 0], children: typeof fallback === "function" ? fallback(progress) : fallback }),
    /* @__PURE__ */ jsx6("group", { visible: ready, children })
  ] });
}

// src/components/LoadingScreen.tsx
import { useState as useState2, useEffect as useEffect8 } from "react";
import { jsx as jsx7, jsxs as jsxs7 } from "react/jsx-runtime";
var S = {
  root: {
    position: "absolute",
    inset: 0,
    display: "flex",
    flexDirection: "column",
    alignItems: "center",
    justifyContent: "center",
    fontFamily: "system-ui, -apple-system, sans-serif",
    color: "#e0e0e0"
  },
  logo: {
    marginBottom: 32,
    maxWidth: 200,
    maxHeight: 80
  },
  barTrack: {
    width: 300,
    height: 6,
    background: "rgba(255,255,255,0.15)",
    borderRadius: 3,
    overflow: "hidden"
  },
  barFill: {
    height: "100%",
    borderRadius: 3,
    transition: "width 0.3s ease"
  },
  pct: {
    marginTop: 12,
    fontSize: 14,
    opacity: 0.8
  },
  asset: {
    marginTop: 8,
    fontSize: 12,
    opacity: 0.5,
    maxWidth: 300,
    overflow: "hidden",
    textOverflow: "ellipsis",
    whiteSpace: "nowrap"
  },
  count: {
    marginTop: 4,
    fontSize: 12,
    opacity: 0.4
  },
  tip: {
    marginTop: 24,
    fontSize: 13,
    opacity: 0.6,
    maxWidth: 400,
    textAlign: "center",
    minHeight: 40,
    transition: "opacity 0.4s ease"
  },
  spinner: {
    width: 32,
    height: 32,
    border: "3px solid rgba(255,255,255,0.2)",
    borderRadius: "50%",
    marginBottom: 16,
    animation: "carverjs-spin 0.8s linear infinite"
  }
};
var _injected = false;
function injectKeyframes() {
  if (_injected || typeof document === "undefined") return;
  _injected = true;
  const style = document.createElement("style");
  style.textContent = `@keyframes carverjs-spin{to{transform:rotate(360deg)}}`;
  document.head.appendChild(style);
}
function LoadingScreen({
  progress,
  theme = "default",
  background = "#000000",
  accentColor = "#6366f1",
  logo,
  tips,
  tipInterval = 5e3,
  showCurrentAsset = true,
  showCount = true,
  children
}) {
  injectKeyframes();
  const [tipIndex, setTipIndex] = useState2(0);
  useEffect8(() => {
    if (!tips || tips.length <= 1) return;
    const id = setInterval(() => {
      setTipIndex((i) => (i + 1) % tips.length);
    }, tipInterval);
    return () => clearInterval(id);
  }, [tips, tipInterval]);
  if (theme === "custom" && children) {
    return /* @__PURE__ */ jsx7("div", { style: { ...S.root, background }, children: children(progress) });
  }
  const pct = Math.round(progress.progress * 100);
  const barWidth = `${progress.progress * 100}%`;
  const logoEl = logo ? typeof logo === "string" ? /* @__PURE__ */ jsx7("img", { src: logo, alt: "", style: S.logo }) : /* @__PURE__ */ jsx7("div", { style: S.logo, children: logo }) : null;
  if (theme === "minimal") {
    return /* @__PURE__ */ jsxs7("div", { style: { ...S.root, background }, children: [
      /* @__PURE__ */ jsx7("div", { style: { ...S.spinner, borderTopColor: accentColor } }),
      /* @__PURE__ */ jsxs7("div", { style: S.pct, children: [
        pct,
        "%"
      ] })
    ] });
  }
  const isGaming = theme === "gaming";
  return /* @__PURE__ */ jsxs7("div", { style: { ...S.root, background }, children: [
    logoEl,
    /* @__PURE__ */ jsx7(
      "div",
      {
        style: {
          ...S.barTrack,
          ...isGaming ? { width: 400, height: 8 } : {}
        },
        children: /* @__PURE__ */ jsx7(
          "div",
          {
            style: {
              ...S.barFill,
              width: barWidth,
              background: accentColor
            }
          }
        )
      }
    ),
    /* @__PURE__ */ jsxs7("div", { style: S.pct, children: [
      pct,
      "%"
    ] }),
    showCurrentAsset && progress.currentAsset && /* @__PURE__ */ jsxs7("div", { style: S.asset, children: [
      "Loading ",
      progress.currentAsset,
      "..."
    ] }),
    showCount && progress.total > 0 && /* @__PURE__ */ jsxs7("div", { style: S.count, children: [
      progress.loaded,
      "/",
      progress.total
    ] }),
    isGaming && tips && tips.length > 0 && /* @__PURE__ */ jsx7("div", { style: S.tip, children: tips[tipIndex] }),
    progress.errors.length > 0 && /* @__PURE__ */ jsxs7("div", { style: { ...S.count, color: "#ef4444", marginTop: 12 }, children: [
      progress.errors.length,
      " asset(s) failed to load"
    ] })
  ] });
}

// src/components/ParticleEmitter.tsx
import { useRef as useRef6, useEffect as useEffect9, forwardRef as forwardRef3 } from "react";
import { useThree as useThree2 } from "@react-three/fiber";
import { jsx as jsx8 } from "react/jsx-runtime";
var ParticleEmitter = forwardRef3(
  function ParticleEmitter2({ preset, position, rotation, enabled = true, ...configOverrides }, ref) {
    const scene = useThree2((s) => s.scene);
    const groupRef = useRef6(null);
    const emitterIdRef = useRef6("");
    const configRef = useRef6(null);
    if (configRef.current === null) {
      const base = preset ? getParticlePreset(preset) : {};
      configRef.current = { ...base, ...configOverrides };
    }
    const setRefs = (node) => {
      groupRef.current = node;
      if (typeof ref === "function") ref(node);
      else if (ref)
        ref.current = node;
    };
    useEffect9(() => {
      const mgr = getParticleManager();
      const config = configRef.current;
      const id = mgr.createEmitter(config);
      emitterIdRef.current = id;
      const emitter = mgr.getEmitter(id);
      emitter.source = groupRef.current;
      const space = config.space ?? "world";
      if (space === "world") {
        scene.add(emitter.mesh);
      } else if (groupRef.current) {
        groupRef.current.add(emitter.mesh);
      }
      return () => {
        mgr.destroyEmitter(id);
        emitterIdRef.current = "";
      };
    }, [scene]);
    useEffect9(() => {
      const emitter = getParticleManager().getEmitter(emitterIdRef.current);
      if (!emitter) return;
      if (enabled) emitter.start();
      else emitter.stop();
    }, [enabled]);
    return /* @__PURE__ */ jsx8("group", { ref: setRefs, position, rotation });
  }
);
export {
  Actor,
  AssetLoader,
  AudioListener,
  Camera,
  Game,
  LoadingScreen,
  ParticleEmitter,
  Scene,
  SceneManager,
  World,
  useGameContext
};
