// @carverjs/core/dist/hooks.mjs
import {
  useAnimation,
  useCamera
} from "./chunk-HPYY6PNE.mjs";
import {
  destroyGridCollisionManager,
  getGridCollisionManager
} from "./chunk-332ISZNQ.mjs";
import {
  getActorRegistry,
  getAssetManager,
  getAudioManager,
  getCollisionManager,
  getInputManager,
  getParticleManager,
  getParticlePreset,
  getTweenManager,
  usePhysicsContext,
  useRigidBodyRef
} from "./chunk-4JMHDLMK.mjs";
import {
  useGameStore,
  useSceneStore
} from "./chunk-333G7WFK.mjs";

// src/hooks/useGameLoop.ts
import { useRef, useEffect } from "react";
import { useFrame } from "@react-three/fiber";
var STAGE_PRIORITY = {
  earlyUpdate: -40,
  fixedUpdate: -30,
  update: -20,
  lateUpdate: -10
};
var DEFAULT_FIXED_DELTA = 1 / 60;
var DEFAULT_MAX_DELTA = 0.1;
function useGameLoop(callback, options = {}) {
  const {
    stage = "update",
    fixedTimestep = false,
    fixedDelta = DEFAULT_FIXED_DELTA,
    maxDelta = DEFAULT_MAX_DELTA,
    enabled = true
  } = options;
  const callbackRef = useRef(callback);
  useEffect(() => {
    callbackRef.current = callback;
  });
  const enabledRef = useRef(enabled);
  useEffect(() => {
    enabledRef.current = enabled;
  }, [enabled]);
  const accumulatorRef = useRef(0);
  useFrame((_, rawDelta) => {
    if (!enabledRef.current) return;
    const { phase: phase2, elapsed: elapsed2 } = useGameStore.getState();
    if (phase2 !== "playing") return;
    const delta = Math.min(rawDelta, maxDelta);
    if (fixedTimestep && stage === "fixedUpdate") {
      accumulatorRef.current += delta;
      let localElapsed = elapsed2;
      while (accumulatorRef.current >= fixedDelta) {
        callbackRef.current(fixedDelta, localElapsed);
        accumulatorRef.current -= fixedDelta;
        localElapsed += fixedDelta;
      }
    } else {
      callbackRef.current(delta, elapsed2);
    }
  }, STAGE_PRIORITY[stage]);
  const phase = useGameStore((s) => s.phase);
  const elapsed = useGameStore((s) => s.elapsed);
  return {
    phase,
    isPaused: phase === "paused" || phase === "gameover",
    elapsed
  };
}

// src/hooks/useInput.ts
import { useRef as useRef2, useEffect as useEffect2, useCallback, useMemo } from "react";
function useInput(options = {}) {
  const { actions = {}, enabled = true } = options;
  const enabledRef = useRef2(enabled);
  useEffect2(() => {
    enabledRef.current = enabled;
  }, [enabled]);
  const actionsRef = useRef2(actions);
  useEffect2(() => {
    actionsRef.current = actions;
  }, [actions]);
  const isPressed = useCallback((code) => {
    if (!enabledRef.current) return false;
    return getInputManager().getKey(code).pressed;
  }, []);
  const isJustPressed = useCallback((code) => {
    if (!enabledRef.current) return false;
    return getInputManager().getKey(code).justPressed;
  }, []);
  const isJustReleased = useCallback((code) => {
    if (!enabledRef.current) return false;
    return getInputManager().getKey(code).justReleased;
  }, []);
  const isAction = useCallback((action) => {
    if (!enabledRef.current) return false;
    const bindings = actionsRef.current[action];
    if (!bindings) return false;
    const mgr = getInputManager();
    return bindings.some((code) => mgr.getKey(code).pressed);
  }, []);
  const isActionJustPressed = useCallback((action) => {
    if (!enabledRef.current) return false;
    const bindings = actionsRef.current[action];
    if (!bindings) return false;
    const mgr = getInputManager();
    return bindings.some((code) => mgr.getKey(code).justPressed);
  }, []);
  const getAxis = useCallback((negative, positive) => {
    if (!enabledRef.current) return 0;
    const mgr = getInputManager();
    let value = 0;
    if (mgr.getKey(negative).pressed) value -= 1;
    if (mgr.getKey(positive).pressed) value += 1;
    return value;
  }, []);
  const pointer = useMemo(() => getInputManager().getPointer(), []);
  return {
    isPressed,
    isJustPressed,
    isJustReleased,
    isAction,
    isActionJustPressed,
    pointer,
    getAxis
  };
}

// src/hooks/useCollision.ts
import { useRef as useRef3, useEffect as useEffect3 } from "react";
var _idCounter = 0;
function useCollision(options) {
  const {
    ref,
    name = "",
    userData = {},
    collider,
    layer = 4294967295,
    mask = 4294967295,
    sensor = false,
    onCollisionEnter,
    onCollisionExit,
    onCollisionStay,
    enabled = true
  } = options;
  const idRef = useRef3("");
  if (idRef.current === "") {
    idRef.current = `col_${++_idCounter}`;
  }
  const id = idRef.current;
  const onEnterRef = useRef3(onCollisionEnter);
  const onExitRef = useRef3(onCollisionExit);
  const onStayRef = useRef3(onCollisionStay);
  useEffect3(() => {
    onEnterRef.current = onCollisionEnter;
    onExitRef.current = onCollisionExit;
    onStayRef.current = onCollisionStay;
  });
  useEffect3(() => {
    const mgr = getCollisionManager();
    mgr.register(id, {
      ref,
      collider,
      name,
      userData,
      layer,
      mask,
      sensor,
      onEnter: (e) => onEnterRef.current?.(e) ?? void 0,
      onExit: (e) => onExitRef.current?.(e) ?? void 0,
      onStay: (e) => onStayRef.current?.(e) ?? void 0,
      enabled
    });
    return () => {
      mgr.unregister(id);
    };
  }, [id, collider, name, layer, mask, sensor]);
  useEffect3(() => {
    getCollisionManager().updateEnabled(id, enabled);
  }, [id, enabled]);
  const returnRef = useRef3({
    isOverlapping: (targetName) => getCollisionManager().isOverlapping(id, targetName),
    getOverlaps: () => getCollisionManager().getOverlaps(id)
  });
  return returnRef.current;
}

// src/hooks/useGridCollision.ts
import { useRef as useRef4, useEffect as useEffect4 } from "react";
function useGridCollision(options) {
  const { config } = options;
  useEffect4(() => {
    getGridCollisionManager().create(config);
    return () => {
      destroyGridCollisionManager();
    };
  }, [config.width, config.height, config.cellSize, config.origin?.[0], config.origin?.[1]]);
  const returnRef = useRef4({
    setCell: (x, y, value) => getGridCollisionManager().setCell(x, y, value),
    getCell: (x, y) => getGridCollisionManager().getCell(x, y),
    clearCell: (x, y) => getGridCollisionManager().clearCell(x, y),
    isCellOccupied: (x, y) => getGridCollisionManager().isCellOccupied(x, y),
    worldToGrid: (wx, wy) => getGridCollisionManager().worldToGrid(wx, wy),
    gridToWorld: (gx, gy) => getGridCollisionManager().gridToWorld(gx, gy),
    getNeighbors4: (x, y) => getGridCollisionManager().getNeighbors4(x, y),
    getNeighbors8: (x, y) => getGridCollisionManager().getNeighbors8(x, y),
    clearAll: () => getGridCollisionManager().clearAll()
  });
  return returnRef.current;
}

// src/hooks/usePhysics.ts
import { useRef as useRef5 } from "react";
function usePhysics() {
  const ctx = usePhysicsContext();
  const rigidBodyRef = useRigidBodyRef();
  const returnRef = useRef5(null);
  if (!ctx?.available || !rigidBodyRef) {
    return null;
  }
  if (!returnRef.current) {
    returnRef.current = {
      applyImpulse: (impulse) => {
        const rb = rigidBodyRef.current;
        if (rb) rb.applyImpulse({ x: impulse[0], y: impulse[1], z: impulse[2] }, true);
      },
      applyForce: (force) => {
        const rb = rigidBodyRef.current;
        if (rb) rb.addForce({ x: force[0], y: force[1], z: force[2] }, true);
      },
      setLinearVelocity: (velocity) => {
        const rb = rigidBodyRef.current;
        if (rb) rb.setLinvel({ x: velocity[0], y: velocity[1], z: velocity[2] }, true);
      },
      getLinearVelocity: () => {
        const rb = rigidBodyRef.current;
        if (!rb) return [0, 0, 0];
        const v = rb.linvel();
        return [v.x, v.y, v.z];
      },
      setAngularVelocity: (velocity) => {
        const rb = rigidBodyRef.current;
        if (rb) rb.setAngvel({ x: velocity[0], y: velocity[1], z: velocity[2] }, true);
      },
      setTranslation: (position) => {
        const rb = rigidBodyRef.current;
        if (rb) rb.setTranslation({ x: position[0], y: position[1], z: position[2] }, true);
      },
      getTranslation: () => {
        const rb = rigidBodyRef.current;
        if (!rb) return [0, 0, 0];
        const t = rb.translation();
        return [t.x, t.y, t.z];
      },
      setRotation: (rotation) => {
        const rb = rigidBodyRef.current;
        if (rb) rb.setRotation({ x: rotation[0], y: rotation[1], z: rotation[2], w: rotation[3] }, true);
      },
      setEnabled: (enabled) => {
        const rb = rigidBodyRef.current;
        if (rb) rb.setEnabled(enabled);
      },
      resetForces: () => {
        const rb = rigidBodyRef.current;
        if (rb) rb.resetForces(true);
      }
    };
  }
  return returnRef.current;
}

// src/hooks/useCameraDirection.ts
import { useRef as useRef6, useCallback as useCallback2 } from "react";
import { useThree } from "@react-three/fiber";
import { Vector3 } from "three";
function useCameraDirection() {
  const { camera } = useThree();
  const _dir = useRef6(new Vector3());
  const getForward = useCallback2(() => {
    camera.getWorldDirection(_dir.current);
    _dir.current.y = 0;
    _dir.current.normalize();
    return [_dir.current.x, 0, _dir.current.z];
  }, [camera]);
  const getRight = useCallback2(() => {
    camera.getWorldDirection(_dir.current);
    _dir.current.y = 0;
    _dir.current.normalize();
    return [_dir.current.z, 0, -_dir.current.x];
  }, [camera]);
  return { getForward, getRight };
}

// src/hooks/useActorRegistry.ts
function useActorRegistry() {
  const registry = getActorRegistry();
  return {
    get: (id) => registry.get(id),
    getAll: () => registry.getAll(),
    getNetworked: () => registry.getNetworked(),
    onRegister: (cb) => registry.onRegister(cb),
    onUnregister: (cb) => registry.onUnregister(cb)
  };
}

// src/hooks/useAudio.ts
import { useRef as useRef7, useEffect as useEffect5, useCallback as useCallback3, useSyncExternalStore } from "react";
function useAudio(options = {}) {
  const { sounds = {}, enabled = true } = options;
  const enabledRef = useRef7(enabled);
  useEffect5(() => {
    enabledRef.current = enabled;
  }, [enabled]);
  const initialSoundsRef = useRef7(sounds);
  useEffect5(() => {
    const mgr2 = getAudioManager();
    const soundMap = initialSoundsRef.current;
    const names = Object.keys(soundMap);
    for (const name of names) {
      mgr2.registerSound(name, soundMap[name]);
    }
    return () => {
      for (const name of names) {
        mgr2.unregisterSound(name);
      }
    };
  }, []);
  const mgr = getAudioManager();
  useSyncExternalStore(mgr.subscribe, mgr.getSnapshot);
  const play = useCallback3((name, opts) => {
    if (!enabledRef.current) return null;
    return getAudioManager().play(name, opts);
  }, []);
  const stop = useCallback3((name) => {
    getAudioManager().stopByName(name);
  }, []);
  const pause = useCallback3((name) => {
    getAudioManager().pauseByName(name);
  }, []);
  const resume = useCallback3((name) => {
    getAudioManager().resumeByName(name);
  }, []);
  const isPlaying = useCallback3((name) => {
    return getAudioManager().isSoundPlaying(name);
  }, []);
  const playMusic = useCallback3((src, opts) => {
    if (!enabledRef.current) return;
    getAudioManager().playMusic(src, opts);
  }, []);
  const stopMusic = useCallback3((fadeOut) => {
    getAudioManager().stopMusic(fadeOut);
  }, []);
  const pauseMusic = useCallback3(() => {
    getAudioManager().pauseMusic();
  }, []);
  const resumeMusic = useCallback3(() => {
    getAudioManager().resumeMusic();
  }, []);
  const isMusicPlaying = useCallback3(() => {
    return getAudioManager().isMusicPlaying();
  }, []);
  const setVolume = useCallback3((channel, volume) => {
    getAudioManager().setChannelVolume(channel, volume);
  }, []);
  const getVolume = useCallback3((channel) => {
    return getAudioManager().getChannelVolume(channel);
  }, []);
  const setMute = useCallback3((channel, muted) => {
    getAudioManager().setChannelMute(channel, muted);
  }, []);
  const isMuted = useCallback3((channel) => {
    return getAudioManager().isChannelMuted(channel);
  }, []);
  const setMasterMute = useCallback3((muted) => {
    getAudioManager().setMasterMute(muted);
  }, []);
  const preload = useCallback3(async (name) => {
    await getAudioManager().preload(name);
  }, []);
  const preloadAll = useCallback3(async (names) => {
    await getAudioManager().preloadAll(names);
  }, []);
  return {
    play,
    stop,
    pause,
    resume,
    isPlaying,
    playMusic,
    stopMusic,
    pauseMusic,
    resumeMusic,
    isMusicPlaying,
    setVolume,
    getVolume,
    setMute,
    isMuted,
    setMasterMute,
    preload,
    preloadAll,
    isUnlocked: mgr.isUnlocked,
    isReady: mgr.isReady
  };
}

// src/hooks/useScene.ts
import { useRef as useRef8, useEffect as useEffect6, useCallback as useCallback4 } from "react";
function useScene() {
  const current = useSceneStore(
    (s) => s.stack[s.stack.length - 1] ?? null
  );
  const stack = useSceneStore((s) => s.stack);
  const isTransitioning = useSceneStore((s) => s.transition.active);
  const go = useCallback4(
    (name, data, transition) => {
      useSceneStore.getState().go(name, data, transition);
    },
    []
  );
  const push = useCallback4(
    (name, data, transition) => {
      useSceneStore.getState().push(name, data, transition);
    },
    []
  );
  const pop = useCallback4((transition) => {
    useSceneStore.getState().pop(transition);
  }, []);
  const replace = useCallback4(
    (name, data, transition) => {
      useSceneStore.getState().replace(name, data, transition);
    },
    []
  );
  return { go, push, pop, replace, current, stack, isTransitioning };
}
function useSceneData() {
  const currentName = useSceneStore(
    (s) => s.stack[s.stack.length - 1] ?? null
  );
  const data = useSceneStore((s) => {
    if (!currentName) return void 0;
    return s.scenes.get(currentName)?.data;
  });
  const setShared = useCallback4((key, value) => {
    useSceneStore.getState().setShared(key, value);
  }, []);
  const getShared = useCallback4((key) => {
    return useSceneStore.getState().getShared(key);
  }, []);
  return { data, setShared, getShared };
}
function useSceneTransition() {
  const active = useSceneStore((s) => s.transition.active);
  const progress = useSceneStore((s) => s.transition.progress);
  const from = useSceneStore((s) => s.transition.from);
  const to = useSceneStore((s) => s.transition.to);
  return { active, progress, from, to };
}
function useSceneLifecycle(sceneName, callbacks) {
  const callbacksRef = useRef8(callbacks);
  useEffect6(() => {
    callbacksRef.current = callbacks;
  });
  const prevStatusRef = useRef8(null);
  const cleanupRef = useRef8(null);
  useEffect6(() => {
    const unsub = useSceneStore.subscribe((state) => {
      const entry = state.scenes.get(sceneName);
      if (!entry) return;
      const prev = prevStatusRef.current;
      const next = entry.status;
      if (prev === next) return;
      prevStatusRef.current = next;
      const cbs = callbacksRef.current;
      if (next === "running" && prev !== "running") {
        if (prev === "sleeping") {
          cbs.onWake?.();
        } else if (prev === "paused") {
          cbs.onResume?.();
        } else {
          const result = cbs.onEnter?.();
          if (typeof result === "function") {
            cleanupRef.current = result;
          }
        }
      }
      if (prev === "running" && next !== "running") {
        cbs.onExit?.();
      }
      if (next === "sleeping" && prev !== "sleeping") {
        cbs.onSleep?.();
      }
      if (next === "paused" && prev !== "paused") {
        cbs.onPause?.();
      }
      if (next === "destroyed") {
        cleanupRef.current?.();
        cleanupRef.current = null;
      }
    });
    return () => {
      unsub();
      cleanupRef.current?.();
      cleanupRef.current = null;
    };
  }, [sceneName]);
}

// src/hooks/useAssets.ts
import { useEffect as useEffect7, useRef as useRef9 } from "react";
function useAssets(keyOrKeys) {
  const manager = getAssetManager();
  const keysRef = useRef9(keyOrKeys);
  useEffect7(() => {
    const keys = Array.isArray(keysRef.current) ? keysRef.current : [keysRef.current];
    const mgr = getAssetManager();
    for (const k of keys) mgr.retain(k);
    return () => {
      for (const k of keys) mgr.release(k);
    };
  }, []);
  if (Array.isArray(keyOrKeys)) {
    return keyOrKeys.map((k) => {
      const asset2 = manager.get(k);
      if (asset2 === void 0) {
        throw new Error(
          `[CarverJS] Asset "${k}" not loaded. Preload via <AssetLoader>.`
        );
      }
      return asset2;
    });
  }
  const asset = manager.get(keyOrKeys);
  if (asset === void 0) {
    throw new Error(
      `[CarverJS] Asset "${keyOrKeys}" not loaded. Preload via <AssetLoader>.`
    );
  }
  return asset;
}

// src/hooks/useAssetProgress.ts
import { useSyncExternalStore as useSyncExternalStore2 } from "react";
function useAssetProgress() {
  const manager = getAssetManager();
  useSyncExternalStore2(manager.subscribe, manager.getSnapshot);
  return manager.progress;
}

// src/hooks/useTween.ts
import { useRef as useRef10, useEffect as useEffect8, useCallback as useCallback5 } from "react";
function useTween() {
  const ownedIdsRef = useRef10(/* @__PURE__ */ new Set());
  useEffect8(() => {
    return () => {
      const mgr = getTweenManager();
      for (const id of ownedIdsRef.current) {
        mgr.killById(id);
      }
      ownedIdsRef.current.clear();
    };
  }, []);
  const tween = useCallback5(
    (config) => {
      const controls = getTweenManager().create(config);
      ownedIdsRef.current.add(controls.id);
      return controls;
    },
    []
  );
  const tweenNumber = useCallback5(
    (config) => {
      const controls = getTweenManager().createNumber(config);
      ownedIdsRef.current.add(controls.id);
      return controls;
    },
    []
  );
  const timeline = useCallback5(
    (config) => {
      return getTweenManager().createTimeline(config);
    },
    []
  );
  const killAll = useCallback5(() => {
    const mgr = getTweenManager();
    for (const id of ownedIdsRef.current) {
      mgr.killById(id);
    }
    ownedIdsRef.current.clear();
  }, []);
  const pauseAll = useCallback5(() => {
    getTweenManager().pauseAll();
  }, []);
  const resumeAll = useCallback5(() => {
    getTweenManager().resumeAll();
  }, []);
  const returnRef = useRef10({
    tween: null,
    tweenNumber: null,
    timeline: null,
    killAll: null,
    pauseAll: null,
    resumeAll: null
  });
  returnRef.current.tween = tween;
  returnRef.current.tweenNumber = tweenNumber;
  returnRef.current.timeline = timeline;
  returnRef.current.killAll = killAll;
  returnRef.current.pauseAll = pauseAll;
  returnRef.current.resumeAll = resumeAll;
  return returnRef.current;
}

// src/hooks/useParticles.ts
import { useRef as useRef11, useEffect as useEffect9, useCallback as useCallback6 } from "react";
import { useThree as useThree2 } from "@react-three/fiber";
function useParticles(options) {
  const { preset, enabled = true, ...configOverrides } = options;
  const scene = useThree2((s) => s.scene);
  const emitterIdRef = useRef11("");
  const sourceRef = useRef11(null);
  const configRef = useRef11(null);
  if (configRef.current === null) {
    const base = preset ? getParticlePreset(preset) : {};
    configRef.current = { ...base, ...configOverrides };
  }
  useEffect9(() => {
    const mgr = getParticleManager();
    const config = configRef.current;
    const id = mgr.createEmitter(config);
    emitterIdRef.current = id;
    const emitter = mgr.getEmitter(id);
    emitter.source = sourceRef.current;
    const space = config.space ?? "world";
    if (space === "world") {
      scene.add(emitter.mesh);
    } else if (sourceRef.current) {
      sourceRef.current.add(emitter.mesh);
    }
    return () => {
      mgr.destroyEmitter(id);
      emitterIdRef.current = "";
    };
  }, [scene]);
  useEffect9(() => {
    const emitter = getParticleManager().getEmitter(emitterIdRef.current);
    if (!emitter) return;
    if (enabled) emitter.start();
    else emitter.stop();
  }, [enabled]);
  const burst = useCallback6((count) => {
    getParticleManager().getEmitter(emitterIdRef.current)?.burst(count);
  }, []);
  const start = useCallback6(() => {
    getParticleManager().getEmitter(emitterIdRef.current)?.start();
  }, []);
  const stop = useCallback6(() => {
    getParticleManager().getEmitter(emitterIdRef.current)?.stop();
  }, []);
  const clear = useCallback6(() => {
    getParticleManager().getEmitter(emitterIdRef.current)?.clear();
  }, []);
  const setRate = useCallback6((rate) => {
    getParticleManager().getEmitter(emitterIdRef.current)?.setRate(rate);
  }, []);
  const getActiveCount = useCallback6(() => {
    return getParticleManager().getEmitter(emitterIdRef.current)?.getActiveCount() ?? 0;
  }, []);
  const isEmitting = useCallback6(() => {
    return getParticleManager().getEmitter(emitterIdRef.current)?.isEmitting() ?? false;
  }, []);
  const reset = useCallback6(() => {
    getParticleManager().getEmitter(emitterIdRef.current)?.reset();
  }, []);
  const returnRef = useRef11({
    ref: sourceRef,
    burst: null,
    start: null,
    stop: null,
    clear: null,
    setRate: null,
    getActiveCount: null,
    isEmitting: null,
    reset: null
  });
  returnRef.current.burst = burst;
  returnRef.current.start = start;
  returnRef.current.stop = stop;
  returnRef.current.clear = clear;
  returnRef.current.setRate = setRate;
  returnRef.current.getActiveCount = getActiveCount;
  returnRef.current.isEmitting = isEmitting;
  returnRef.current.reset = reset;
  return returnRef.current;
}
export {
  useActorRegistry,
  useAnimation,
  useAssetProgress,
  useAssets,
  useAudio,
  useCamera,
  useCameraDirection,
  useCollision,
  useGameLoop,
  useGridCollision,
  useInput,
  useParticles,
  usePhysics,
  useScene,
  useSceneData,
  useSceneLifecycle,
  useSceneTransition,
  useTween
};
