// @carverjs/core/dist/store.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/store/index.ts
var store_exports = {};
__export(store_exports, {
  useGameStore: () => useGameStore,
  useSceneStore: () => useSceneStore
});
module.exports = __toCommonJS(store_exports);

// src/store/gameStore.ts
var import_zustand = require("zustand");
var useGameStore = (0, import_zustand.create)()((set, get) => ({
  phase: "loading",
  elapsed: 0,
  frameCount: 0,
  setPhase: (phase) => set({ phase }),
  _tick: (delta) => {
    if (get().phase !== "playing") return;
    set((s) => ({
      elapsed: s.elapsed + delta,
      frameCount: s.frameCount + 1
    }));
  },
  pause: () => {
    if (get().phase === "playing") set({ phase: "paused" });
  },
  resume: () => {
    if (get().phase === "paused") set({ phase: "playing" });
  },
  reset: () => set({ phase: "playing", elapsed: 0, frameCount: 0 })
}));

// src/store/sceneStore.ts
var import_zustand2 = require("zustand");
var DEFAULT_TRANSITION = {
  active: false,
  from: null,
  to: null,
  progress: 0,
  config: { type: "none" },
  swapped: false,
  targetStack: null
};
function resolveTransition(config, sceneConfig) {
  return config ?? sceneConfig?.transition ?? { type: "none" };
}
var useSceneStore = (0, import_zustand2.create)()((set, get) => ({
  scenes: /* @__PURE__ */ new Map(),
  stack: [],
  transition: { ...DEFAULT_TRANSITION },
  shared: {},
  // ── Navigation Actions ────────────────────────────────────────────────────
  go: (name, data, transition) => {
    const state = get();
    const entry = state.scenes.get(name);
    if (!entry) {
      console.warn(`[CarverJS Scene] Scene "${name}" is not registered`);
      return;
    }
    if (state.transition.active) return;
    const currentName = state.stack[state.stack.length - 1] ?? null;
    const resolved = resolveTransition(transition, entry.config);
    if (resolved.type === "none") {
      _exitStack(state, set);
      entry.status = "running";
      entry.data = data;
      set({
        scenes: new Map(state.scenes),
        stack: [name]
      });
      return;
    }
    if (currentName) {
      const current = state.scenes.get(currentName);
      if (current) current.status = "shutting_down";
    }
    entry.data = data;
    set({
      scenes: new Map(state.scenes),
      transition: {
        active: true,
        from: currentName,
        to: name,
        progress: 0,
        config: resolved,
        swapped: false,
        targetStack: [name]
      }
    });
  },
  push: (name, data, transition) => {
    const state = get();
    const entry = state.scenes.get(name);
    if (!entry) {
      console.warn(`[CarverJS Scene] Scene "${name}" is not registered`);
      return;
    }
    if (state.transition.active) return;
    const currentName = state.stack[state.stack.length - 1] ?? null;
    const resolved = resolveTransition(transition, entry.config);
    if (resolved.type === "none") {
      if (currentName) {
        const current = state.scenes.get(currentName);
        if (current) current.status = "sleeping";
      }
      entry.status = "running";
      entry.data = data;
      set({
        scenes: new Map(state.scenes),
        stack: [...state.stack, name]
      });
      return;
    }
    if (currentName) {
      const current = state.scenes.get(currentName);
      if (current) current.status = "shutting_down";
    }
    entry.data = data;
    set({
      scenes: new Map(state.scenes),
      stack: [...state.stack, name],
      transition: {
        active: true,
        from: currentName,
        to: name,
        progress: 0,
        config: resolved,
        swapped: false,
        targetStack: null
      }
    });
  },
  pop: (transition) => {
    const state = get();
    if (state.stack.length <= 1) return;
    if (state.transition.active) return;
    const currentName = state.stack[state.stack.length - 1];
    const previousName = state.stack[state.stack.length - 2];
    const currentEntry = state.scenes.get(currentName);
    const previousEntry = state.scenes.get(previousName);
    const resolved = resolveTransition(
      transition,
      previousEntry?.config
    );
    if (resolved.type === "none") {
      _destroyOrSleep(currentEntry);
      if (previousEntry) previousEntry.status = "running";
      set({
        scenes: new Map(state.scenes),
        stack: state.stack.slice(0, -1)
      });
      return;
    }
    if (currentEntry) currentEntry.status = "shutting_down";
    set({
      scenes: new Map(state.scenes),
      transition: {
        active: true,
        from: currentName,
        to: previousName,
        progress: 0,
        config: resolved,
        swapped: false,
        targetStack: state.stack.slice(0, -1)
      }
    });
  },
  replace: (name, data, transition) => {
    const state = get();
    const entry = state.scenes.get(name);
    if (!entry) {
      console.warn(`[CarverJS Scene] Scene "${name}" is not registered`);
      return;
    }
    if (state.transition.active) return;
    const currentName = state.stack[state.stack.length - 1] ?? null;
    const resolved = resolveTransition(transition, entry.config);
    if (resolved.type === "none") {
      if (currentName) {
        const current = state.scenes.get(currentName);
        _destroyOrSleep(current);
      }
      entry.status = "running";
      entry.data = data;
      const newStack2 = [...state.stack];
      if (newStack2.length > 0) {
        newStack2[newStack2.length - 1] = name;
      } else {
        newStack2.push(name);
      }
      set({
        scenes: new Map(state.scenes),
        stack: newStack2
      });
      return;
    }
    if (currentName) {
      const current = state.scenes.get(currentName);
      if (current) current.status = "shutting_down";
    }
    entry.data = data;
    const newStack = [...state.stack];
    if (newStack.length > 0) {
      newStack[newStack.length - 1] = name;
    } else {
      newStack.push(name);
    }
    set({
      scenes: new Map(state.scenes),
      stack: newStack,
      transition: {
        active: true,
        from: currentName,
        to: name,
        progress: 0,
        config: resolved,
        swapped: false,
        targetStack: null
      }
    });
  },
  // ── Shared Data ───────────────────────────────────────────────────────────
  setShared: (key, value) => {
    set((s) => ({ shared: { ...s.shared, [key]: value } }));
  },
  getShared: (key) => {
    return get().shared[key];
  },
  // ── Internal Actions ──────────────────────────────────────────────────────
  _registerScene: (config) => {
    const state = get();
    const existing = state.scenes.get(config.name);
    if (existing) {
      existing.config = config;
      set({ scenes: new Map(state.scenes) });
      return;
    }
    const entry = {
      config,
      status: "created",
      data: void 0
    };
    state.scenes.set(config.name, entry);
    set({ scenes: new Map(state.scenes) });
  },
  _unregisterScene: (name) => {
    const state = get();
    state.scenes.delete(name);
    set({
      scenes: new Map(state.scenes),
      stack: state.stack.filter((s) => s !== name)
    });
  },
  _setStatus: (name, status) => {
    const state = get();
    const entry = state.scenes.get(name);
    if (!entry) return;
    entry.status = status;
    set({ scenes: new Map(state.scenes) });
  },
  _tickTransition: (delta) => {
    const state = get();
    const t = state.transition;
    if (!t.active || !t.to) return;
    const duration = t.config.duration ?? 0.5;
    const newProgress = Math.min(t.progress + delta / duration, 1);
    if (!t.swapped && newProgress >= 0.5) {
      const fromEntry = t.from ? state.scenes.get(t.from) : null;
      const toEntry = state.scenes.get(t.to);
      if (fromEntry) {
        _destroyOrSleep(fromEntry);
      }
      if (toEntry) {
        toEntry.status = "running";
      }
      const updates = {
        scenes: new Map(state.scenes),
        transition: { ...t, progress: newProgress, swapped: true }
      };
      if (t.targetStack) {
        updates.stack = t.targetStack;
      }
      set(updates);
      return;
    }
    if (newProgress >= 1) {
      set({
        transition: { ...DEFAULT_TRANSITION }
      });
      return;
    }
    set({
      transition: { ...t, progress: newProgress }
    });
  },
  _startInitial: (name) => {
    const state = get();
    const entry = state.scenes.get(name);
    if (!entry) {
      console.warn(`[CarverJS Scene] Initial scene "${name}" is not registered`);
      return;
    }
    entry.status = "running";
    set({
      scenes: new Map(state.scenes),
      stack: [name]
    });
  }
}));
function _destroyOrSleep(entry) {
  if (!entry) return;
  entry.status = entry.config.persistent ? "sleeping" : "destroyed";
}
function _exitStack(state, set) {
  for (const name of state.stack) {
    const entry = state.scenes.get(name);
    _destroyOrSleep(entry);
  }
  set({ scenes: new Map(state.scenes) });
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  useGameStore,
  useSceneStore
});
