// @carverjs/core/dist/systems.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/systems/index.ts
var systems_exports = {};
__export(systems_exports, {
  Easing: () => Easing,
  cubicBezier: () => cubicBezier,
  destroyActorRegistry: () => destroyActorRegistry,
  destroyAssetManager: () => destroyAssetManager,
  destroyAudioManager: () => destroyAudioManager,
  destroyCollisionManager: () => destroyCollisionManager,
  destroyGridCollisionManager: () => destroyGridCollisionManager,
  destroyInputManager: () => destroyInputManager,
  destroyParticleManager: () => destroyParticleManager,
  destroySceneManager: () => destroySceneManager,
  destroyTweenManager: () => destroyTweenManager,
  detectAssetType: () => detectAssetType,
  getActorRegistry: () => getActorRegistry,
  getAssetManager: () => getAssetManager,
  getAudioManager: () => getAudioManager,
  getCollisionManager: () => getCollisionManager,
  getGridCollisionManager: () => getGridCollisionManager,
  getInputManager: () => getInputManager,
  getParticleManager: () => getParticleManager,
  getParticlePreset: () => getParticlePreset,
  getSceneManager: () => getSceneManager,
  getTweenManager: () => getTweenManager,
  registerParticlePreset: () => registerParticlePreset,
  resolveEasing: () => resolveEasing,
  steps: () => steps,
  usePhysicsContext: () => usePhysicsContext
});
module.exports = __toCommonJS(systems_exports);

// src/systems/InputManager.ts
var DEFAULT_KEY_STATE = {
  pressed: false,
  justPressed: false,
  justReleased: false
};
var InputManager = class {
  constructor() {
    this._keys = /* @__PURE__ */ new Map();
    this._pendingDown = /* @__PURE__ */ new Set();
    this._pendingUp = /* @__PURE__ */ new Set();
    this._pointer = {
      position: { x: 0, y: 0 },
      isDown: false,
      justDown: false,
      justUp: false
    };
    this._pendingPointerDown = false;
    this._pendingPointerUp = false;
    this._attached = false;
    // Bound handlers stored as arrow-function properties for clean removal
    this._onKeyDown = (e) => {
      if (e.repeat) return;
      this._pendingDown.add(e.code);
      this._pendingUp.delete(e.code);
    };
    this._onKeyUp = (e) => {
      this._pendingUp.add(e.code);
      this._pendingDown.delete(e.code);
    };
    this._onPointerDown = (e) => {
      if (e.button !== 0) return;
      this._pendingPointerDown = true;
      this._pendingPointerUp = false;
      this._pointer.position.x = e.clientX;
      this._pointer.position.y = e.clientY;
    };
    this._onPointerUp = (e) => {
      if (e.button !== 0) return;
      this._pendingPointerUp = true;
      this._pendingPointerDown = false;
      this._pointer.position.x = e.clientX;
      this._pointer.position.y = e.clientY;
    };
    this._onPointerMove = (e) => {
      this._pointer.position.x = e.clientX;
      this._pointer.position.y = e.clientY;
    };
    this._onTouchStart = (e) => {
      const touch = e.touches[0];
      if (!touch) return;
      this._pendingPointerDown = true;
      this._pendingPointerUp = false;
      this._pointer.position.x = touch.clientX;
      this._pointer.position.y = touch.clientY;
    };
    this._onTouchMove = (e) => {
      const touch = e.touches[0];
      if (!touch) return;
      this._pointer.position.x = touch.clientX;
      this._pointer.position.y = touch.clientY;
    };
    this._onTouchEnd = () => {
      this._pendingPointerUp = true;
      this._pendingPointerDown = false;
    };
    this._onBlur = () => {
      for (const [code, state] of this._keys) {
        if (state.pressed) {
          this._pendingUp.add(code);
        }
      }
      this._pendingDown.clear();
      if (this._pointer.isDown) {
        this._pendingPointerUp = true;
        this._pendingPointerDown = false;
      }
    };
  }
  /** Attach DOM listeners. Keyboard + blur on window, pointer/touch on target element. */
  attach(target) {
    if (this._attached) return;
    this._attached = true;
    window.addEventListener("keydown", this._onKeyDown);
    window.addEventListener("keyup", this._onKeyUp);
    window.addEventListener("blur", this._onBlur);
    target.addEventListener("pointerdown", this._onPointerDown);
    target.addEventListener("pointerup", this._onPointerUp);
    target.addEventListener("pointermove", this._onPointerMove);
    target.addEventListener("touchstart", this._onTouchStart, { passive: true });
    target.addEventListener("touchmove", this._onTouchMove, { passive: true });
    target.addEventListener("touchend", this._onTouchEnd, { passive: true });
    target.addEventListener("touchcancel", this._onTouchEnd, { passive: true });
  }
  /** Remove all DOM listeners */
  detach() {
    if (!this._attached) return;
    this._attached = false;
    window.removeEventListener("keydown", this._onKeyDown);
    window.removeEventListener("keyup", this._onKeyUp);
    window.removeEventListener("blur", this._onBlur);
  }
  /**
   * Called once per frame at priority -50, BEFORE all game logic stages.
   * Clears previous frame's transient state, then applies queued DOM events.
   */
  flush() {
    for (const [, state] of this._keys) {
      state.justPressed = false;
      state.justReleased = false;
    }
    for (const code of this._pendingDown) {
      let state = this._keys.get(code);
      if (!state) {
        state = { pressed: false, justPressed: false, justReleased: false };
        this._keys.set(code, state);
      }
      state.pressed = true;
      state.justPressed = true;
    }
    this._pendingDown.clear();
    for (const code of this._pendingUp) {
      let state = this._keys.get(code);
      if (!state) {
        state = { pressed: false, justPressed: false, justReleased: false };
        this._keys.set(code, state);
      }
      state.pressed = false;
      state.justReleased = true;
    }
    this._pendingUp.clear();
    this._pointer.justDown = false;
    this._pointer.justUp = false;
    if (this._pendingPointerDown) {
      this._pointer.isDown = true;
      this._pointer.justDown = true;
      this._pendingPointerDown = false;
    }
    if (this._pendingPointerUp) {
      this._pointer.isDown = false;
      this._pointer.justUp = true;
      this._pendingPointerUp = false;
    }
  }
  // ── Query API ──
  /** Get the state of a key by its KeyboardEvent.code. Returns default (all false) if never seen. */
  getKey(code) {
    return this._keys.get(code) ?? DEFAULT_KEY_STATE;
  }
  /** Get the current pointer state. Returns a live reference — always up to date. */
  getPointer() {
    return this._pointer;
  }
};
var _instance = null;
function getInputManager() {
  if (!_instance) _instance = new InputManager();
  return _instance;
}
function destroyInputManager() {
  _instance?.detach();
  _instance = null;
}

// src/systems/CollisionManager.ts
var import_three = require("three");
var _posA = new import_three.Vector3();
var _posB = new import_three.Vector3();
function aabbOverlap(posA, a, posB, b, is2D) {
  const aoX = a.offset?.[0] ?? 0;
  const aoY = a.offset?.[1] ?? 0;
  const aoZ = a.offset?.[2] ?? 0;
  const boX = b.offset?.[0] ?? 0;
  const boY = b.offset?.[1] ?? 0;
  const boZ = b.offset?.[2] ?? 0;
  const ax = posA.x + aoX;
  const ay = posA.y + aoY;
  const bx = posB.x + boX;
  const by = posB.y + boY;
  if (Math.abs(ax - bx) > a.halfExtents[0] + b.halfExtents[0] || Math.abs(ay - by) > a.halfExtents[1] + b.halfExtents[1]) {
    return false;
  }
  if (!is2D) {
    const az = posA.z + aoZ;
    const bz = posB.z + boZ;
    if (Math.abs(az - bz) > a.halfExtents[2] + b.halfExtents[2]) {
      return false;
    }
  }
  return true;
}
function sphereOverlap(posA, a, posB, b, is2D) {
  const aoX = a.offset?.[0] ?? 0;
  const aoY = a.offset?.[1] ?? 0;
  const boX = b.offset?.[0] ?? 0;
  const boY = b.offset?.[1] ?? 0;
  const dx = posA.x + aoX - (posB.x + boX);
  const dy = posA.y + aoY - (posB.y + boY);
  let distSq = dx * dx + dy * dy;
  if (!is2D) {
    const aoZ = a.offset?.[2] ?? 0;
    const boZ = b.offset?.[2] ?? 0;
    const dz = posA.z + aoZ - (posB.z + boZ);
    distSq += dz * dz;
  }
  const r = a.radius + b.radius;
  return distSq <= r * r;
}
function aabbSphereOverlap(posAABB, aabb, posSphere, sphere, is2D) {
  const aoX = aabb.offset?.[0] ?? 0;
  const aoY = aabb.offset?.[1] ?? 0;
  const soX = sphere.offset?.[0] ?? 0;
  const soY = sphere.offset?.[1] ?? 0;
  const cx = posAABB.x + aoX;
  const cy = posAABB.y + aoY;
  const sx = posSphere.x + soX;
  const sy = posSphere.y + soY;
  const clampedX = Math.max(cx - aabb.halfExtents[0], Math.min(sx, cx + aabb.halfExtents[0]));
  const clampedY = Math.max(cy - aabb.halfExtents[1], Math.min(sy, cy + aabb.halfExtents[1]));
  let distSq = (sx - clampedX) ** 2 + (sy - clampedY) ** 2;
  if (!is2D) {
    const aoZ = aabb.offset?.[2] ?? 0;
    const soZ = sphere.offset?.[2] ?? 0;
    const cz = posAABB.z + aoZ;
    const sz = posSphere.z + soZ;
    const clampedZ = Math.max(cz - aabb.halfExtents[2], Math.min(sz, cz + aabb.halfExtents[2]));
    distSq += (sz - clampedZ) ** 2;
  }
  return distSq <= sphere.radius * sphere.radius;
}
function testOverlap(posA, colA, posB, colB, is2D) {
  const shapeA = colA.shape;
  const shapeB = colB.shape;
  if (shapeA === "aabb" && shapeB === "aabb") {
    return aabbOverlap(posA, colA, posB, colB, is2D);
  }
  if ((shapeA === "sphere" || shapeA === "circle") && (shapeB === "sphere" || shapeB === "circle")) {
    return sphereOverlap(posA, colA, posB, colB, is2D);
  }
  if (shapeA === "aabb" && (shapeB === "sphere" || shapeB === "circle")) {
    return aabbSphereOverlap(posA, colA, posB, colB, is2D);
  }
  if ((shapeA === "sphere" || shapeA === "circle") && shapeB === "aabb") {
    return aabbSphereOverlap(posB, colB, posA, colA, is2D);
  }
  return false;
}
var SpatialHash = class {
  constructor() {
    this._cellSize = 2;
    this._cells = /* @__PURE__ */ new Map();
  }
  setCellSize(size) {
    this._cellSize = Math.max(size, 0.01);
  }
  clear() {
    this._cells.clear();
  }
  _key(cx, cy) {
    return cx * 92837111 ^ cy * 689287499 | 0;
  }
  insert(id, x, y, radius) {
    const cs = this._cellSize;
    const minCX = Math.floor((x - radius) / cs);
    const maxCX = Math.floor((x + radius) / cs);
    const minCY = Math.floor((y - radius) / cs);
    const maxCY = Math.floor((y + radius) / cs);
    for (let cx = minCX; cx <= maxCX; cx++) {
      for (let cy = minCY; cy <= maxCY; cy++) {
        const key = this._key(cx, cy);
        let bucket = this._cells.get(key);
        if (!bucket) {
          bucket = [];
          this._cells.set(key, bucket);
        }
        bucket.push(id);
      }
    }
  }
  getPotentialPairs() {
    const pairs = /* @__PURE__ */ new Set();
    for (const bucket of this._cells.values()) {
      if (bucket.length < 2) continue;
      for (let i = 0; i < bucket.length; i++) {
        for (let j = i + 1; j < bucket.length; j++) {
          const a = bucket[i];
          const b = bucket[j];
          const pairKey = a < b ? `${a}|${b}` : `${b}|${a}`;
          pairs.add(pairKey);
        }
      }
    }
    return pairs;
  }
};
var CollisionManager = class {
  constructor() {
    this._entries = /* @__PURE__ */ new Map();
    this._mode = "3d";
    this._spatialHash = new SpatialHash();
  }
  setMode(mode) {
    this._mode = mode;
  }
  register(id, entry) {
    this._entries.set(id, { ...entry, id, overlaps: /* @__PURE__ */ new Set() });
  }
  unregister(id) {
    const entry = this._entries.get(id);
    if (!entry) return;
    for (const otherId of entry.overlaps) {
      const other = this._entries.get(otherId);
      if (other) {
        other.overlaps.delete(id);
        if (other.onExit && other.enabled) {
          other.onExit({
            otherName: entry.name,
            otherUserData: entry.userData,
            otherRef: entry.ref,
            isSensor: entry.sensor
          });
        }
      }
    }
    this._entries.delete(id);
  }
  updateCallbacks(id, onEnter, onExit, onStay) {
    const entry = this._entries.get(id);
    if (entry) {
      entry.onEnter = onEnter;
      entry.onExit = onExit;
      entry.onStay = onStay;
    }
  }
  updateEnabled(id, enabled) {
    const entry = this._entries.get(id);
    if (entry) entry.enabled = enabled;
  }
  getOverlaps(id) {
    const entry = this._entries.get(id);
    if (!entry) return [];
    const result = [];
    for (const otherId of entry.overlaps) {
      const other = this._entries.get(otherId);
      if (other) result.push(other.name);
    }
    return result;
  }
  isOverlapping(id, name) {
    const entry = this._entries.get(id);
    if (!entry) return false;
    for (const otherId of entry.overlaps) {
      const other = this._entries.get(otherId);
      if (other && other.name === name) return true;
    }
    return false;
  }
  /** Called once per frame at priority -25. */
  tick() {
    const entries = this._entries;
    if (entries.size < 2) return;
    const is2D = this._mode === "2d";
    this._spatialHash.clear();
    let maxRadius = 0;
    for (const [id, entry] of entries) {
      if (!entry.enabled || !entry.ref.current) continue;
      entry.ref.current.getWorldPosition(_posA);
      const col = entry.collider;
      let radius;
      if (col.shape === "aabb") {
        radius = Math.sqrt(
          col.halfExtents[0] ** 2 + col.halfExtents[1] ** 2 + (is2D ? 0 : col.halfExtents[2] ** 2)
        );
      } else {
        radius = col.radius;
      }
      if (radius > maxRadius) maxRadius = radius;
      this._spatialHash.insert(id, _posA.x, _posA.y, radius);
    }
    this._spatialHash.setCellSize(maxRadius * 2 || 2);
    const potentialPairs = this._spatialHash.getPotentialPairs();
    const currentOverlaps = /* @__PURE__ */ new Set();
    for (const pairKey of potentialPairs) {
      const [idA, idB] = pairKey.split("|");
      const entryA = entries.get(idA);
      const entryB = entries.get(idB);
      if (!entryA || !entryB) continue;
      if (!entryA.enabled || !entryB.enabled) continue;
      if (!entryA.ref.current || !entryB.ref.current) continue;
      if ((entryA.layer & entryB.mask) === 0 || (entryB.layer & entryA.mask) === 0) {
        continue;
      }
      entryA.ref.current.getWorldPosition(_posA);
      entryB.ref.current.getWorldPosition(_posB);
      if (!testOverlap(_posA, entryA.collider, _posB, entryB.collider, is2D)) {
        continue;
      }
      currentOverlaps.add(pairKey);
      const wasOverlapping = entryA.overlaps.has(idB);
      if (!wasOverlapping) {
        entryA.overlaps.add(idB);
        entryB.overlaps.add(idA);
        const eventForA = {
          otherName: entryB.name,
          otherUserData: entryB.userData,
          otherRef: entryB.ref,
          isSensor: entryB.sensor
        };
        const eventForB = {
          otherName: entryA.name,
          otherUserData: entryA.userData,
          otherRef: entryA.ref,
          isSensor: entryA.sensor
        };
        entryA.onEnter?.(eventForA);
        entryB.onEnter?.(eventForB);
      } else {
        const eventForA = {
          otherName: entryB.name,
          otherUserData: entryB.userData,
          otherRef: entryB.ref,
          isSensor: entryB.sensor
        };
        const eventForB = {
          otherName: entryA.name,
          otherUserData: entryA.userData,
          otherRef: entryA.ref,
          isSensor: entryA.sensor
        };
        entryA.onStay?.(eventForA);
        entryB.onStay?.(eventForB);
      }
    }
    for (const [id, entry] of entries) {
      if (!entry.enabled) continue;
      const toRemove = [];
      for (const otherId of entry.overlaps) {
        const pairKey = id < otherId ? `${id}|${otherId}` : `${otherId}|${id}`;
        if (!currentOverlaps.has(pairKey)) {
          toRemove.push(otherId);
        }
      }
      for (const otherId of toRemove) {
        entry.overlaps.delete(otherId);
        const other = entries.get(otherId);
        if (other) {
          other.overlaps.delete(id);
          if (entry.onExit) {
            entry.onExit({
              otherName: other.name,
              otherUserData: other.userData,
              otherRef: other.ref,
              isSensor: other.sensor
            });
          }
        }
      }
    }
  }
};
var _instance2 = null;
function getCollisionManager() {
  if (!_instance2) _instance2 = new CollisionManager();
  return _instance2;
}
function destroyCollisionManager() {
  _instance2 = null;
}

// src/systems/GridCollisionManager.ts
var GridCollisionManager = class {
  constructor() {
    this._grid = new Int32Array(0);
    this._width = 0;
    this._height = 0;
    this._cellSize = 1;
    this._originX = 0;
    this._originY = 0;
  }
  create(config) {
    this._width = config.width;
    this._height = config.height;
    this._cellSize = config.cellSize ?? 1;
    this._originX = config.origin?.[0] ?? 0;
    this._originY = config.origin?.[1] ?? 0;
    this._grid = new Int32Array(this._width * this._height);
  }
  _inBounds(x, y) {
    return x >= 0 && x < this._width && y >= 0 && y < this._height;
  }
  _index(x, y) {
    return y * this._width + x;
  }
  setCell(x, y, value) {
    if (!this._inBounds(x, y)) return;
    this._grid[this._index(x, y)] = value;
  }
  getCell(x, y) {
    if (!this._inBounds(x, y)) return 0;
    return this._grid[this._index(x, y)];
  }
  clearCell(x, y) {
    this.setCell(x, y, 0);
  }
  isCellOccupied(x, y) {
    return this.getCell(x, y) !== 0;
  }
  worldToGrid(worldX, worldY) {
    return [
      Math.floor((worldX - this._originX) / this._cellSize),
      Math.floor((worldY - this._originY) / this._cellSize)
    ];
  }
  gridToWorld(gridX, gridY) {
    return [
      this._originX + gridX * this._cellSize + this._cellSize / 2,
      this._originY + gridY * this._cellSize + this._cellSize / 2
    ];
  }
  getNeighbors4(x, y) {
    return [
      this.getCell(x, y - 1),
      // up
      this.getCell(x + 1, y),
      // right
      this.getCell(x, y + 1),
      // down
      this.getCell(x - 1, y)
      // left
    ];
  }
  getNeighbors8(x, y) {
    return [
      this.getCell(x, y - 1),
      // up
      this.getCell(x + 1, y - 1),
      // up-right
      this.getCell(x + 1, y),
      // right
      this.getCell(x + 1, y + 1),
      // down-right
      this.getCell(x, y + 1),
      // down
      this.getCell(x - 1, y + 1),
      // down-left
      this.getCell(x - 1, y),
      // left
      this.getCell(x - 1, y - 1)
      // up-left
    ];
  }
  clearAll() {
    this._grid.fill(0);
  }
  get width() {
    return this._width;
  }
  get height() {
    return this._height;
  }
  get cellSize() {
    return this._cellSize;
  }
};
var _instance3 = null;
function getGridCollisionManager() {
  if (!_instance3) _instance3 = new GridCollisionManager();
  return _instance3;
}
function destroyGridCollisionManager() {
  _instance3 = null;
}

// src/systems/PhysicsProvider.tsx
var import_react = require("react");
var import_jsx_runtime = require("react/jsx-runtime");
var PhysicsContext = (0, import_react.createContext)(null);
function usePhysicsContext() {
  return (0, import_react.useContext)(PhysicsContext);
}
var RigidBodyRefContext = (0, import_react.createContext)(null);

// src/systems/ActorRegistry.ts
var ActorRegistry = class {
  constructor() {
    this._entries = /* @__PURE__ */ new Map();
    this._onRegisterCallbacks = [];
    this._onUnregisterCallbacks = [];
  }
  register(id, ref) {
    this._entries.set(id, ref);
    for (const cb of this._onRegisterCallbacks) cb(id, ref);
  }
  unregister(id) {
    this._entries.delete(id);
    for (const cb of this._onUnregisterCallbacks) cb(id);
  }
  get(id) {
    return this._entries.get(id);
  }
  getAll() {
    return this._entries;
  }
  getNetworked() {
    const result = /* @__PURE__ */ new Map();
    for (const [id, ref] of this._entries) {
      if (ref.userData.networked) {
        result.set(id, ref);
      }
    }
    return result;
  }
  onRegister(cb) {
    this._onRegisterCallbacks.push(cb);
    return () => {
      const idx = this._onRegisterCallbacks.indexOf(cb);
      if (idx >= 0) this._onRegisterCallbacks.splice(idx, 1);
    };
  }
  onUnregister(cb) {
    this._onUnregisterCallbacks.push(cb);
    return () => {
      const idx = this._onUnregisterCallbacks.indexOf(cb);
      if (idx >= 0) this._onUnregisterCallbacks.splice(idx, 1);
    };
  }
};
var _instance4 = null;
function getActorRegistry() {
  if (!_instance4) _instance4 = new ActorRegistry();
  return _instance4;
}
function destroyActorRegistry() {
  _instance4 = null;
}

// src/systems/AudioManager.ts
var import_three2 = require("three");

// src/systems/MusicEngine.ts
var MusicEngine = class {
  constructor(ctx, output) {
    this._source = null;
    this._gain = null;
    this._buffer = null;
    this._volume = 1;
    this._loop = true;
    this._playing = false;
    this._paused = false;
    this._startTime = 0;
    this._pauseOffset = 0;
    this._ctx = ctx;
    this._output = output;
  }
  /**
   * Play a music track with optional crossfade from the current track.
   * If a track is already playing, the old track fades out while the new one fades in.
   */
  play(buffer, volume, loop, crossfadeDuration) {
    const now = this._ctx.currentTime;
    if (this._source && this._gain) {
      const oldGain = this._gain;
      const oldSource = this._source;
      oldGain.gain.setValueAtTime(oldGain.gain.value, now);
      oldGain.gain.linearRampToValueAtTime(0, now + crossfadeDuration);
      setTimeout(() => {
        try {
          oldSource.stop();
        } catch {
        }
        oldGain.disconnect();
      }, (crossfadeDuration + 0.1) * 1e3);
    }
    this._buffer = buffer;
    this._volume = volume;
    this._loop = loop;
    const gain = this._ctx.createGain();
    const source = this._ctx.createBufferSource();
    source.buffer = buffer;
    source.loop = loop;
    source.connect(gain);
    gain.connect(this._output);
    if (this._playing) {
      gain.gain.setValueAtTime(0, now);
      gain.gain.linearRampToValueAtTime(volume, now + crossfadeDuration);
    } else {
      gain.gain.setValueAtTime(volume, now);
    }
    source.start(0);
    this._source = source;
    this._gain = gain;
    this._startTime = now;
    this._pauseOffset = 0;
    this._playing = true;
    this._paused = false;
    if (!loop) {
      source.addEventListener("ended", () => {
        if (this._source === source) {
          this._playing = false;
          this._source = null;
          this._gain = null;
        }
      });
    }
  }
  /** Stop the current music track with optional fade-out. */
  stop(fadeOut = 0) {
    if (!this._source || !this._gain) return;
    const now = this._ctx.currentTime;
    if (fadeOut > 0) {
      this._gain.gain.setValueAtTime(this._gain.gain.value, now);
      this._gain.gain.linearRampToValueAtTime(0, now + fadeOut);
      const source = this._source;
      const gain = this._gain;
      setTimeout(() => {
        try {
          source.stop();
        } catch {
        }
        gain.disconnect();
      }, (fadeOut + 0.1) * 1e3);
    } else {
      try {
        this._source.stop();
      } catch {
      }
      this._gain.disconnect();
    }
    this._source = null;
    this._gain = null;
    this._playing = false;
    this._paused = false;
  }
  /** Pause the current music track. Records position for resume. */
  pause() {
    if (!this._playing || this._paused || !this._source) return;
    this._pauseOffset += this._ctx.currentTime - this._startTime;
    this._paused = true;
    try {
      this._source.stop();
    } catch {
    }
    this._source = null;
  }
  /** Resume from the paused position. */
  resume() {
    if (!this._paused || !this._buffer || !this._gain) return;
    const source = this._ctx.createBufferSource();
    source.buffer = this._buffer;
    source.loop = this._loop;
    source.connect(this._gain);
    this._gain.gain.cancelScheduledValues(this._ctx.currentTime);
    this._gain.gain.setValueAtTime(this._volume, this._ctx.currentTime);
    const offset = this._pauseOffset % this._buffer.duration;
    source.start(0, offset);
    this._source = source;
    this._startTime = this._ctx.currentTime;
    this._paused = false;
    if (!this._loop) {
      source.addEventListener("ended", () => {
        if (this._source === source) {
          this._playing = false;
          this._source = null;
          this._gain = null;
        }
      });
    }
  }
  isPlaying() {
    return this._playing && !this._paused;
  }
  destroy() {
    this.stop();
  }
};

// src/systems/HTML5AudioPool.ts
var UNLOCK_EVENTS = [
  "click",
  "touchstart",
  "touchend",
  "keydown",
  "keyup",
  "mousedown",
  "pointerdown",
  "pointerup"
];
var HTML5AudioPool = class {
  constructor(poolSize = 10) {
    this._pool = [];
    this._active = /* @__PURE__ */ new Map();
    this._nextId = 0;
    this._unlocked = false;
    // Music state (two elements for crossfade)
    this._musicElement = null;
    this._musicFadingOut = null;
    this._musicPlaying = false;
    this._musicPaused = false;
    this._musicVolume = 1;
    this._fadeOutStart = 0;
    this._fadeOutDuration = 0;
    this._poolSize = poolSize;
    this._createPool();
    this._setupUnlock();
  }
  _createPool() {
    for (let i = 0; i < this._poolSize; i++) {
      this._pool.push(new Audio());
    }
  }
  _setupUnlock() {
    const unlock = () => {
      if (this._unlocked) return;
      for (const audio of this._pool) {
        audio.load();
      }
      this._unlocked = true;
      UNLOCK_EVENTS.forEach(
        (e) => document.removeEventListener(e, unlock, true)
      );
    };
    UNLOCK_EVENTS.forEach(
      (e) => document.addEventListener(e, unlock, true)
    );
  }
  get isUnlocked() {
    return this._unlocked;
  }
  /** Acquire a free audio element from the pool. Steals oldest if full. */
  _acquire() {
    for (const audio of this._pool) {
      if (audio.paused && !audio.src) return audio;
    }
    for (const audio of this._pool) {
      if (audio.paused) return audio;
    }
    const stolen = this._pool[0];
    if (stolen) {
      stolen.pause();
      stolen.currentTime = 0;
      stolen.src = "";
      return stolen;
    }
    return null;
  }
  play(url, volume, loop, rate, spriteOffset, spriteDuration, onEnd) {
    const audio = this._acquire();
    if (!audio) return null;
    const id = ++this._nextId;
    audio.src = url;
    audio.volume = Math.max(0, Math.min(1, volume));
    audio.loop = loop && !spriteDuration;
    audio.playbackRate = rate;
    let spriteTimer = null;
    if (spriteOffset !== void 0 && spriteOffset > 0) {
      audio.currentTime = spriteOffset;
    }
    const instance = { id, element: audio, state: "playing", onEnd: onEnd ?? null };
    this._active.set(id, instance);
    const cleanup = () => {
      if (instance.state !== "playing") return;
      instance.state = "stopped";
      instance.onEnd?.();
      this._active.delete(id);
      audio.pause();
      audio.src = "";
      if (spriteTimer) clearTimeout(spriteTimer);
    };
    if (spriteDuration && spriteDuration > 0) {
      spriteTimer = setTimeout(cleanup, spriteDuration / rate * 1e3);
    }
    audio.onended = () => {
      if (spriteTimer) clearTimeout(spriteTimer);
      cleanup();
    };
    audio.play().catch(() => {
      instance.state = "stopped";
      this._active.delete(id);
    });
    return instance;
  }
  stop(instance) {
    if (instance.state === "stopped") return;
    instance.state = "stopped";
    this._active.delete(instance.id);
    instance.element.pause();
    instance.element.currentTime = 0;
    instance.element.src = "";
  }
  pause(instance) {
    if (instance.state !== "playing") return;
    instance.state = "paused";
    instance.element.pause();
  }
  resume(instance) {
    if (instance.state !== "paused") return;
    instance.state = "playing";
    instance.element.play().catch(() => {
      instance.state = "stopped";
      this._active.delete(instance.id);
    });
  }
  setVolume(instance, volume) {
    instance.element.volume = Math.max(0, Math.min(1, volume));
  }
  setRate(instance, rate) {
    instance.element.playbackRate = rate;
  }
  // ── Music (HTML5 fallback) ──────────────────────────────────────────────────
  playMusic(url, volume, loop, crossfadeDuration) {
    if (this._musicElement && this._musicPlaying) {
      this._musicFadingOut = this._musicElement;
      this._fadeOutStart = performance.now();
      this._fadeOutDuration = crossfadeDuration * 1e3;
    }
    const audio = new Audio(url);
    audio.loop = loop;
    audio.volume = this._musicPlaying ? 0 : volume;
    this._musicElement = audio;
    this._musicVolume = volume;
    this._musicPlaying = true;
    this._musicPaused = false;
    audio.play().catch(() => {
      this._musicPlaying = false;
    });
    audio.onended = () => {
      if (this._musicElement === audio) {
        this._musicPlaying = false;
        this._musicElement = null;
      }
    };
  }
  stopMusic(fadeOut = 0) {
    if (!this._musicElement) return;
    if (fadeOut > 0) {
      this._musicFadingOut = this._musicElement;
      this._fadeOutStart = performance.now();
      this._fadeOutDuration = fadeOut * 1e3;
    } else {
      this._musicElement.pause();
      this._musicElement.src = "";
    }
    this._musicElement = null;
    this._musicPlaying = false;
    this._musicPaused = false;
  }
  pauseMusic() {
    if (!this._musicPlaying || this._musicPaused || !this._musicElement) return;
    this._musicElement.pause();
    this._musicPaused = true;
  }
  resumeMusic() {
    if (!this._musicPaused || !this._musicElement) return;
    this._musicElement.play().catch(() => {
    });
    this._musicPaused = false;
  }
  isMusicPlaying() {
    return this._musicPlaying && !this._musicPaused;
  }
  /** Called per-frame by AudioManager to drive HTML5 music crossfade interpolation. */
  flushMusic() {
    if (this._musicFadingOut && this._fadeOutDuration > 0) {
      const elapsed = performance.now() - this._fadeOutStart;
      const t = Math.min(elapsed / this._fadeOutDuration, 1);
      this._musicFadingOut.volume = Math.max(0, (1 - t) * this._musicVolume);
      if (this._musicElement) {
        this._musicElement.volume = Math.min(1, t * this._musicVolume);
      }
      if (t >= 1) {
        this._musicFadingOut.pause();
        this._musicFadingOut.src = "";
        this._musicFadingOut = null;
      }
    }
  }
  /** Pause all active sounds (for game phase pause). */
  pauseAll() {
    for (const [, inst] of this._active) {
      if (inst.state === "playing") {
        inst.element.pause();
      }
    }
    if (this._musicElement && this._musicPlaying && !this._musicPaused) {
      this._musicElement.pause();
    }
  }
  /** Resume all sounds (for game phase resume). */
  resumeAll() {
    for (const [, inst] of this._active) {
      if (inst.state === "playing") {
        inst.element.play().catch(() => {
        });
      }
    }
    if (this._musicElement && this._musicPlaying && !this._musicPaused) {
      this._musicElement.play().catch(() => {
      });
    }
  }
  destroy() {
    for (const [, inst] of this._active) {
      inst.element.pause();
      inst.element.src = "";
    }
    this._active.clear();
    if (this._musicElement) {
      this._musicElement.pause();
      this._musicElement.src = "";
      this._musicElement = null;
    }
    if (this._musicFadingOut) {
      this._musicFadingOut.pause();
      this._musicFadingOut.src = "";
      this._musicFadingOut = null;
    }
    this._pool = [];
    this._musicPlaying = false;
    UNLOCK_EVENTS.forEach(
      (e) => document.removeEventListener(e, () => {
      }, true)
    );
  }
};

// src/systems/AudioManager.ts
var _emitterPos = new import_three2.Vector3();
var _listenerPos = new import_three2.Vector3();
var _listenerDir = new import_three2.Vector3();
var _listenerUp = new import_three2.Vector3();
var _supportedFormats = /* @__PURE__ */ (() => {
  if (typeof Audio === "undefined") return /* @__PURE__ */ new Map();
  const audio = new Audio();
  return /* @__PURE__ */ new Map([
    ["mp3", audio.canPlayType("audio/mpeg") !== ""],
    ["ogg", audio.canPlayType('audio/ogg; codecs="vorbis"') !== ""],
    ["wav", audio.canPlayType('audio/wav; codecs="1"') !== ""],
    ["webm", audio.canPlayType('audio/webm; codecs="vorbis"') !== ""],
    ["aac", audio.canPlayType("audio/aac") !== ""],
    ["flac", audio.canPlayType("audio/flac") !== ""]
  ]);
})();
function resolveUrl(src) {
  const urls = Array.isArray(src) ? src : [src];
  for (const url of urls) {
    const ext = url.split(".").pop()?.toLowerCase();
    if (ext && (_supportedFormats.get(ext) ?? true)) return url;
  }
  return urls[0] ?? null;
}
var CHANNELS = ["master", "sfx", "music", "ui", "ambient", "voice"];
var UNLOCK_EVENTS2 = [
  "click",
  "touchstart",
  "touchend",
  "keydown",
  "keyup",
  "mousedown",
  "pointerdown",
  "pointerup"
];
var AudioManager = class {
  constructor() {
    // Web Audio API state
    this._context = null;
    this._masterGain = null;
    this._channelGains = /* @__PURE__ */ new Map();
    this._musicEngine = null;
    // HTML5 fallback
    this._useHTML5 = false;
    this._html5Pool = null;
    this._html5Refs = /* @__PURE__ */ new Map();
    // Shared state
    this._channelVolumes = /* @__PURE__ */ new Map();
    this._channelMuted = /* @__PURE__ */ new Map();
    this._bufferCache = /* @__PURE__ */ new Map();
    this._pendingLoads = /* @__PURE__ */ new Map();
    this._definitions = /* @__PURE__ */ new Map();
    this._active = /* @__PURE__ */ new Map();
    this._activeByName = /* @__PURE__ */ new Map();
    this._spatialTrackers = /* @__PURE__ */ new Map();
    this._listenerSource = null;
    this._customListener = null;
    this._unlocked = false;
    this._ready = false;
    this._gamePaused = false;
    this._nextId = 0;
    this._lastPlayed = /* @__PURE__ */ new Map();
    this._queuedPlays = [];
    this._queuedMusic = null;
    // ── useSyncExternalStore subscriber pattern ──
    this._stateVersion = 0;
    this._subscribers = /* @__PURE__ */ new Set();
    this.subscribe = (cb) => {
      this._subscribers.add(cb);
      return () => {
        this._subscribers.delete(cb);
      };
    };
    this.getSnapshot = () => this._stateVersion;
    // ── AudioContext Unlock ─────────────────────────────────────────────────────
    this._unlockHandler = () => {
      if (this._unlocked || !this._context) return;
      if (this._context.state === "suspended") {
        this._context.resume().then(() => {
          if (!this._unlocked) this._onUnlocked();
        });
      } else {
        this._onUnlocked();
      }
      const silentBuffer = this._context.createBuffer(1, 1, 22050);
      const source = this._context.createBufferSource();
      source.buffer = silentBuffer;
      source.connect(this._context.destination);
      source.start(0);
      source.stop(0);
    };
    this._visibilityHandler = () => {
      if (!this._context) return;
      if (!document.hidden && this._context.state === "suspended" && this._unlocked) {
        this._context.resume();
      }
    };
  }
  _notify() {
    this._stateVersion++;
    for (const cb of this._subscribers) cb();
  }
  get isUnlocked() {
    return this._unlocked;
  }
  get isReady() {
    return this._ready;
  }
  // ── Initialization ──────────────────────────────────────────────────────────
  init() {
    if (this._context || this._html5Pool) return;
    for (const ch of CHANNELS) {
      this._channelVolumes.set(ch, 1);
      this._channelMuted.set(ch, false);
    }
    if (typeof AudioContext === "undefined") {
      this._useHTML5 = true;
      this._html5Pool = new HTML5AudioPool();
      this._ready = true;
      this._unlocked = true;
      this._notify();
      return;
    }
    this._context = new AudioContext();
    this._masterGain = this._context.createGain();
    this._masterGain.connect(this._context.destination);
    for (const ch of CHANNELS) {
      if (ch === "master") {
        this._channelGains.set(ch, this._masterGain);
      } else {
        const gain = this._context.createGain();
        gain.connect(this._masterGain);
        this._channelGains.set(ch, gain);
      }
    }
    this._musicEngine = new MusicEngine(
      this._context,
      this._channelGains.get("music")
    );
    this._setupUnlock();
    if (this._context.state === "running") {
      this._onUnlocked();
    }
  }
  destroy() {
    for (const [, inst] of this._active) {
      try {
        inst.source.stop();
      } catch {
      }
      inst.gainNode.disconnect();
      if (inst.pannerNode) inst.pannerNode.disconnect();
    }
    this._active.clear();
    this._activeByName.clear();
    this._spatialTrackers.clear();
    this._musicEngine?.destroy();
    this._context?.close();
    this._context = null;
    this._masterGain = null;
    this._channelGains.clear();
    this._musicEngine = null;
    this._html5Pool?.destroy();
    this._html5Pool = null;
    this._html5Refs.clear();
    this._bufferCache.clear();
    this._pendingLoads.clear();
    this._definitions.clear();
    this._lastPlayed.clear();
    this._queuedPlays = [];
    this._queuedMusic = null;
    this._unlocked = false;
    this._ready = false;
    this._gamePaused = false;
    UNLOCK_EVENTS2.forEach(
      (e) => document.removeEventListener(e, this._unlockHandler, true)
    );
    document.removeEventListener("visibilitychange", this._visibilityHandler);
  }
  _setupUnlock() {
    UNLOCK_EVENTS2.forEach(
      (e) => document.addEventListener(e, this._unlockHandler, true)
    );
    document.addEventListener("visibilitychange", this._visibilityHandler);
    if (this._context) {
      this._context.onstatechange = () => {
        if (!this._context) return;
        if (this._context.state === "running" && !this._unlocked) {
          this._onUnlocked();
        }
        if (this._context.state === "suspended" || this._context.state === "interrupted") {
          if (this._unlocked) {
            this._unlocked = false;
            this._notify();
            UNLOCK_EVENTS2.forEach(
              (e) => document.addEventListener(e, this._unlockHandler, true)
            );
          }
        }
      };
    }
  }
  _onUnlocked() {
    this._unlocked = true;
    this._ready = true;
    UNLOCK_EVENTS2.forEach(
      (e) => document.removeEventListener(e, this._unlockHandler, true)
    );
    if (this._queuedMusic) {
      const { src, options } = this._queuedMusic;
      this._queuedMusic = null;
      this.playMusic(src, options);
    }
    const queued = this._queuedPlays.splice(0);
    for (const { name, options } of queued) {
      this.play(name, options);
    }
    this._notify();
  }
  // ── Sound Registration ──────────────────────────────────────────────────────
  registerSound(name, def) {
    this._definitions.set(name, def);
    if (def.preload !== false && !this._useHTML5) {
      this._loadByName(name).catch(() => {
      });
    }
  }
  unregisterSound(name) {
    this._definitions.delete(name);
    this.stopByName(name);
  }
  // ── Buffer Loading (Web Audio only) ─────────────────────────────────────────
  async _loadByName(name) {
    const def = this._definitions.get(name);
    if (!def) throw new Error(`Sound "${name}" is not registered`);
    return this._loadBuffer(def.src);
  }
  async _loadBuffer(src) {
    if (!this._context) throw new Error("AudioManager not initialized");
    const url = resolveUrl(src);
    if (!url) throw new Error("No supported audio format found");
    const cached = this._bufferCache.get(url);
    if (cached) return cached;
    const pending = this._pendingLoads.get(url);
    if (pending) return pending;
    const promise = (async () => {
      const response = await fetch(url);
      if (!response.ok) throw new Error(`HTTP ${response.status} loading ${url}`);
      const arrayBuffer = await response.arrayBuffer();
      const audioBuffer = await this._context.decodeAudioData(arrayBuffer);
      this._bufferCache.set(url, audioBuffer);
      return audioBuffer;
    })();
    this._pendingLoads.set(url, promise);
    try {
      return await promise;
    } finally {
      this._pendingLoads.delete(url);
    }
  }
  // ── Sound Playback ──────────────────────────────────────────────────────────
  play(name, options = {}) {
    const def = this._definitions.get(name);
    if (!def) {
      console.warn(`[CarverJS Audio] Sound "${name}" is not registered`);
      return null;
    }
    if (!this._ready) {
      if (options.loop || def.loop || def.channel === "music" || def.channel === "ambient") {
        this._queuedPlays.push({ name, options });
      }
      return null;
    }
    if (this._useHTML5) {
      return this._playHTML5(name, def, options);
    }
    return this._playWebAudio(name, def, options);
  }
  _playHTML5(name, def, options) {
    if (!this._html5Pool) return null;
    const url = resolveUrl(def.src);
    if (!url) return null;
    const volume = options.volume ?? def.volume ?? 1;
    const rate = options.rate ?? def.rate ?? 1;
    const loop = options.loop ?? def.loop ?? false;
    let spriteOffset;
    let spriteDuration;
    if (options.sprite && def.sprites?.[options.sprite]) {
      const region = def.sprites[options.sprite];
      spriteOffset = region.start;
      spriteDuration = region.duration;
    }
    const channelVol = this._channelMuted.get(def.channel ?? "sfx") ? 0 : this._channelVolumes.get(def.channel ?? "sfx") ?? 1;
    const masterVol = this._channelMuted.get("master") ? 0 : this._channelVolumes.get("master") ?? 1;
    const finalVol = volume * channelVol * masterVol;
    const instance = this._html5Pool.play(
      url,
      finalVol,
      loop,
      rate,
      spriteOffset,
      spriteDuration,
      options.onEnd
    );
    if (!instance) return null;
    const ref = { id: instance.id, name, instance };
    this._html5Refs.set(instance.id, ref);
    let nameSet = this._activeByName.get(name);
    if (!nameSet) {
      nameSet = /* @__PURE__ */ new Set();
      this._activeByName.set(name, nameSet);
    }
    nameSet.add(instance.id);
    return this._createHTML5Handle(ref);
  }
  _createHTML5Handle(ref) {
    const pool = this._html5Pool;
    return {
      get id() {
        return ref.instance.id;
      },
      get state() {
        return ref.instance.state;
      },
      stop: () => {
        pool.stop(ref.instance);
        this._html5Refs.delete(ref.id);
        this._activeByName.get(ref.name)?.delete(ref.id);
      },
      pause: () => pool.pause(ref.instance),
      resume: () => pool.resume(ref.instance),
      fade: () => {
      },
      setVolume: (v) => pool.setVolume(ref.instance, v),
      setRate: (r) => pool.setRate(ref.instance, r)
    };
  }
  _playWebAudio(name, def, options) {
    if (!this._context) return null;
    const ctx = this._context;
    if (def.cooldown) {
      const now = ctx.currentTime;
      const lastPlayed = this._lastPlayed.get(name) ?? 0;
      if (now - lastPlayed < def.cooldown) return null;
      this._lastPlayed.set(name, now);
    }
    const maxInstances = def.maxInstances ?? 5;
    const activeSet = this._activeByName.get(name);
    if (activeSet && activeSet.size >= maxInstances) {
      const oldestId = activeSet.values().next().value;
      if (oldestId !== void 0) {
        const oldest = this._active.get(oldestId);
        if (oldest) this._stopInstance(oldest);
      }
    }
    const url = resolveUrl(def.src);
    if (!url) return null;
    const buffer = this._bufferCache.get(url);
    if (!buffer) {
      this._loadByName(name).then((buf) => this._playWithBuffer(name, buf, def, options)).catch((err) => {
        console.warn(`[CarverJS Audio] Failed to load "${name}":`, err);
        options.onError?.(err instanceof Error ? err : new Error(String(err)));
      });
      return null;
    }
    return this._playWithBuffer(name, buffer, def, options);
  }
  _playWithBuffer(name, buffer, def, options) {
    if (!this._context) return null;
    const ctx = this._context;
    const channel = def.channel ?? "sfx";
    const channelGain = this._channelGains.get(channel);
    if (!channelGain) return null;
    let spriteOffset = 0;
    let spriteDuration = 0;
    let spriteLoop = false;
    if (options.sprite && def.sprites?.[options.sprite]) {
      const region = def.sprites[options.sprite];
      spriteOffset = region.start;
      spriteDuration = region.duration;
      spriteLoop = region.loop ?? false;
    }
    const instanceVolume = options.volume ?? def.volume ?? 1;
    const instanceRate = options.rate ?? def.rate ?? 1;
    const instanceLoop = options.loop ?? spriteLoop ?? def.loop ?? false;
    const source = ctx.createBufferSource();
    source.buffer = buffer;
    source.playbackRate.value = instanceRate;
    if (instanceLoop) {
      source.loop = true;
      if (spriteDuration > 0) {
        source.loopStart = spriteOffset;
        source.loopEnd = spriteOffset + spriteDuration;
      }
    }
    const gainNode = ctx.createGain();
    if (options.fadeIn && options.fadeIn > 0) {
      gainNode.gain.setValueAtTime(0, ctx.currentTime);
      gainNode.gain.linearRampToValueAtTime(instanceVolume, ctx.currentTime + options.fadeIn);
    } else {
      gainNode.gain.setValueAtTime(instanceVolume, ctx.currentTime);
    }
    let pannerNode = null;
    if (options.spatial) {
      pannerNode = new PannerNode(ctx, {
        panningModel: options.spatial.panningModel ?? "HRTF",
        distanceModel: options.spatial.distanceModel ?? "inverse",
        refDistance: options.spatial.refDistance ?? 1,
        maxDistance: options.spatial.maxDistance ?? 100,
        rolloffFactor: options.spatial.rolloffFactor ?? 1,
        coneInnerAngle: options.spatial.coneInnerAngle ?? 360,
        coneOuterAngle: options.spatial.coneOuterAngle ?? 360,
        coneOuterGain: options.spatial.coneOuterGain ?? 0
      });
      const obj = options.spatial.ref.current;
      if (obj) {
        obj.getWorldPosition(_emitterPos);
        pannerNode.positionX.value = _emitterPos.x;
        pannerNode.positionY.value = _emitterPos.y;
        pannerNode.positionZ.value = _emitterPos.z;
      }
      source.connect(pannerNode);
      pannerNode.connect(gainNode);
    } else {
      source.connect(gainNode);
    }
    gainNode.connect(channelGain);
    const id = ++this._nextId;
    const instance = {
      id,
      name,
      source,
      gainNode,
      pannerNode,
      state: "playing",
      startTime: ctx.currentTime,
      pauseOffset: 0,
      buffer,
      loop: instanceLoop,
      rate: instanceRate,
      volume: instanceVolume,
      channel,
      spriteOffset,
      spriteDuration,
      onEnd: options.onEnd ?? null
    };
    this._active.set(id, instance);
    let nameSet = this._activeByName.get(name);
    if (!nameSet) {
      nameSet = /* @__PURE__ */ new Set();
      this._activeByName.set(name, nameSet);
    }
    nameSet.add(id);
    if (pannerNode && options.spatial && options.spatial.trackPosition !== false) {
      this._spatialTrackers.set(id, {
        instanceId: id,
        ref: options.spatial.ref,
        pannerNode
      });
    }
    const delay = options.delay ?? 0;
    if (spriteDuration > 0 && !instanceLoop) {
      source.start(ctx.currentTime + delay, spriteOffset, spriteDuration);
    } else {
      source.start(ctx.currentTime + delay, spriteOffset);
    }
    source.addEventListener("ended", () => {
      if (instance.state !== "playing") return;
      instance.state = "stopped";
      instance.onEnd?.();
      this._removeInstance(instance);
    });
    return this._createHandle(instance);
  }
  // ── Sound Handle (Web Audio) ────────────────────────────────────────────────
  _createHandle(instance) {
    return {
      get id() {
        return instance.id;
      },
      get state() {
        return instance.state;
      },
      stop: () => this._stopInstance(instance),
      pause: () => this._pauseInstance(instance),
      resume: () => this._resumeInstance(instance),
      fade: (from, to, dur) => this._fadeInstance(instance, from, to, dur),
      setVolume: (v) => this._setInstanceVolume(instance, v),
      setRate: (r) => this._setInstanceRate(instance, r)
    };
  }
  _stopInstance(instance) {
    if (instance.state === "stopped") return;
    instance.state = "stopped";
    this._removeInstance(instance);
    try {
      instance.source.stop();
    } catch {
    }
    instance.gainNode.disconnect();
    if (instance.pannerNode) instance.pannerNode.disconnect();
  }
  _pauseInstance(instance) {
    if (instance.state !== "playing" || !this._context) return;
    instance.pauseOffset += this._context.currentTime - instance.startTime;
    instance.state = "paused";
    try {
      instance.source.stop();
    } catch {
    }
  }
  _resumeInstance(instance) {
    if (instance.state !== "paused" || !this._context) return;
    const ctx = this._context;
    const source = ctx.createBufferSource();
    source.buffer = instance.buffer;
    source.playbackRate.value = instance.rate;
    if (instance.loop) {
      source.loop = true;
      if (instance.spriteDuration > 0) {
        source.loopStart = instance.spriteOffset;
        source.loopEnd = instance.spriteOffset + instance.spriteDuration;
      }
    }
    if (instance.pannerNode) {
      source.connect(instance.pannerNode);
    } else {
      source.connect(instance.gainNode);
    }
    const offset = instance.spriteOffset + instance.pauseOffset;
    if (instance.spriteDuration > 0 && !instance.loop) {
      const remaining = instance.spriteDuration - instance.pauseOffset;
      if (remaining <= 0) {
        instance.state = "stopped";
        instance.onEnd?.();
        this._removeInstance(instance);
        return;
      }
      source.start(0, offset, remaining);
    } else {
      source.start(0, offset);
    }
    instance.source = source;
    instance.startTime = ctx.currentTime;
    instance.state = "playing";
    source.addEventListener("ended", () => {
      if (instance.state !== "playing") return;
      instance.state = "stopped";
      instance.onEnd?.();
      this._removeInstance(instance);
    });
  }
  _fadeInstance(instance, from, to, duration) {
    if (!this._context || instance.state === "stopped") return;
    instance.gainNode.gain.setValueAtTime(from, this._context.currentTime);
    instance.gainNode.gain.linearRampToValueAtTime(to, this._context.currentTime + duration);
    instance.volume = to;
  }
  _setInstanceVolume(instance, volume) {
    if (!this._context || instance.state === "stopped") return;
    instance.gainNode.gain.setValueAtTime(volume, this._context.currentTime);
    instance.volume = volume;
  }
  _setInstanceRate(instance, rate) {
    if (instance.state === "stopped") return;
    instance.source.playbackRate.value = rate;
    instance.rate = rate;
  }
  _removeInstance(instance) {
    this._active.delete(instance.id);
    const nameSet = this._activeByName.get(instance.name);
    if (nameSet) {
      nameSet.delete(instance.id);
      if (nameSet.size === 0) this._activeByName.delete(instance.name);
    }
    this._spatialTrackers.delete(instance.id);
  }
  // ── Global Sound Operations ─────────────────────────────────────────────────
  stopByName(name) {
    if (this._useHTML5) {
      for (const [id, ref] of this._html5Refs) {
        if (ref.name === name) {
          this._html5Pool.stop(ref.instance);
          this._html5Refs.delete(id);
        }
      }
      this._activeByName.delete(name);
      return;
    }
    const nameSet = this._activeByName.get(name);
    if (!nameSet) return;
    for (const id of [...nameSet]) {
      const inst = this._active.get(id);
      if (inst) this._stopInstance(inst);
    }
  }
  pauseByName(name) {
    if (this._useHTML5) {
      for (const [, ref] of this._html5Refs) {
        if (ref.name === name) this._html5Pool.pause(ref.instance);
      }
      return;
    }
    const nameSet = this._activeByName.get(name);
    if (!nameSet) return;
    for (const id of nameSet) {
      const inst = this._active.get(id);
      if (inst) this._pauseInstance(inst);
    }
  }
  resumeByName(name) {
    if (this._useHTML5) {
      for (const [, ref] of this._html5Refs) {
        if (ref.name === name) this._html5Pool.resume(ref.instance);
      }
      return;
    }
    const nameSet = this._activeByName.get(name);
    if (!nameSet) return;
    for (const id of nameSet) {
      const inst = this._active.get(id);
      if (inst) this._resumeInstance(inst);
    }
  }
  isSoundPlaying(name) {
    if (this._useHTML5) {
      for (const [, ref] of this._html5Refs) {
        if (ref.name === name && ref.instance.state === "playing") return true;
      }
      return false;
    }
    const nameSet = this._activeByName.get(name);
    if (!nameSet) return false;
    for (const id of nameSet) {
      const inst = this._active.get(id);
      if (inst && inst.state === "playing") return true;
    }
    return false;
  }
  // ── Volume Control ──────────────────────────────────────────────────────────
  setChannelVolume(channel, volume) {
    this._channelVolumes.set(channel, volume);
    if (this._channelMuted.get(channel)) return;
    if (!this._useHTML5) {
      const gain = this._channelGains.get(channel);
      if (gain && this._context) {
        gain.gain.setValueAtTime(volume, this._context.currentTime);
      }
    }
  }
  getChannelVolume(channel) {
    return this._channelVolumes.get(channel) ?? 1;
  }
  setChannelMute(channel, muted) {
    this._channelMuted.set(channel, muted);
    if (!this._useHTML5) {
      const gain = this._channelGains.get(channel);
      if (gain && this._context) {
        const volume = muted ? 0 : this._channelVolumes.get(channel) ?? 1;
        gain.gain.setValueAtTime(volume, this._context.currentTime);
      }
    }
  }
  isChannelMuted(channel) {
    return this._channelMuted.get(channel) ?? false;
  }
  setMasterMute(muted) {
    this.setChannelMute("master", muted);
  }
  // ── Music ───────────────────────────────────────────────────────────────────
  async playMusic(src, options = {}) {
    if (!this._ready) {
      this._queuedMusic = { src, options };
      return;
    }
    const url = resolveUrl(src);
    if (!url) return;
    const volume = options.volume ?? 1;
    const loop = options.loop ?? true;
    const crossfade = options.crossfade?.duration ?? 2;
    if (this._useHTML5) {
      this._html5Pool.playMusic(url, volume, loop, crossfade);
      return;
    }
    try {
      const buffer = await this._loadBuffer(src);
      this._musicEngine?.play(buffer, volume, loop, crossfade);
    } catch (err) {
      console.warn("[CarverJS Audio] Failed to load music:", err);
    }
  }
  stopMusic(fadeOut = 0) {
    this._queuedMusic = null;
    if (this._useHTML5) {
      this._html5Pool?.stopMusic(fadeOut);
    } else {
      this._musicEngine?.stop(fadeOut);
    }
  }
  pauseMusic() {
    if (this._useHTML5) {
      this._html5Pool?.pauseMusic();
    } else {
      this._musicEngine?.pause();
    }
  }
  resumeMusic() {
    if (this._useHTML5) {
      this._html5Pool?.resumeMusic();
    } else {
      this._musicEngine?.resume();
    }
  }
  isMusicPlaying() {
    if (this._useHTML5) return this._html5Pool?.isMusicPlaying() ?? false;
    return this._musicEngine?.isPlaying() ?? false;
  }
  // ── Listener ────────────────────────────────────────────────────────────────
  setListenerSource(source) {
    this._listenerSource = source;
  }
  setCustomListener(source) {
    this._customListener = source;
  }
  // ── Preloading ──────────────────────────────────────────────────────────────
  async preload(name) {
    if (this._useHTML5) return;
    await this._loadByName(name);
  }
  async preloadAll(names) {
    if (this._useHTML5) return;
    await Promise.all(names.map((n) => this._loadByName(n)));
  }
  /** Unload a sound's buffer from memory. Re-fetched on next play. */
  unloadBuffer(name) {
    const def = this._definitions.get(name);
    if (!def) return;
    const urls = Array.isArray(def.src) ? def.src : [def.src];
    for (const url of urls) this._bufferCache.delete(url);
  }
  // ── Per-Frame Flush ─────────────────────────────────────────────────────────
  flush(phase) {
    try {
      if (this._useHTML5) {
        const shouldPause2 = phase === "paused" || phase === "gameover";
        if (shouldPause2 && !this._gamePaused) {
          this._gamePaused = true;
          this._html5Pool?.pauseAll();
        } else if (!shouldPause2 && this._gamePaused) {
          this._gamePaused = false;
          this._html5Pool?.resumeAll();
        }
        if (!this._gamePaused) {
          this._html5Pool?.flushMusic();
        }
        return;
      }
      if (!this._context || !this._unlocked) return;
      const shouldPause = phase === "paused" || phase === "gameover";
      if (shouldPause && !this._gamePaused) {
        this._gamePaused = true;
        this._context.suspend();
      } else if (!shouldPause && this._gamePaused) {
        this._gamePaused = false;
        this._context.resume();
      }
      if (this._gamePaused) return;
      this._updateSpatialPositions();
      this._syncListener();
    } catch {
    }
  }
  // ── Spatial Updates ─────────────────────────────────────────────────────────
  _updateSpatialPositions() {
    for (const [, tracker] of this._spatialTrackers) {
      const obj = tracker.ref.current;
      if (!obj) continue;
      obj.getWorldPosition(_emitterPos);
      const panner = tracker.pannerNode;
      panner.positionX.value = _emitterPos.x;
      panner.positionY.value = _emitterPos.y;
      panner.positionZ.value = _emitterPos.z;
    }
  }
  _syncListener() {
    if (!this._context) return;
    const source = this._customListener ?? this._listenerSource;
    if (!source) return;
    const listener = this._context.listener;
    source.getWorldPosition(_listenerPos);
    source.getWorldDirection(_listenerDir);
    _listenerUp.set(0, 1, 0).applyQuaternion(source.quaternion);
    if (listener.positionX !== void 0) {
      listener.positionX.value = _listenerPos.x;
      listener.positionY.value = _listenerPos.y;
      listener.positionZ.value = _listenerPos.z;
      listener.forwardX.value = _listenerDir.x;
      listener.forwardY.value = _listenerDir.y;
      listener.forwardZ.value = _listenerDir.z;
      listener.upX.value = _listenerUp.x;
      listener.upY.value = _listenerUp.y;
      listener.upZ.value = _listenerUp.z;
    } else {
      listener.setPosition(_listenerPos.x, _listenerPos.y, _listenerPos.z);
      listener.setOrientation(
        _listenerDir.x,
        _listenerDir.y,
        _listenerDir.z,
        _listenerUp.x,
        _listenerUp.y,
        _listenerUp.z
      );
    }
  }
};
var _instance5 = null;
function getAudioManager() {
  if (!_instance5) _instance5 = new AudioManager();
  return _instance5;
}
function destroyAudioManager() {
  _instance5?.destroy();
  _instance5 = null;
}

// src/store/sceneStore.ts
var import_zustand = require("zustand");
var DEFAULT_TRANSITION = {
  active: false,
  from: null,
  to: null,
  progress: 0,
  config: { type: "none" },
  swapped: false,
  targetStack: null
};
function resolveTransition(config, sceneConfig) {
  return config ?? sceneConfig?.transition ?? { type: "none" };
}
var useSceneStore = (0, import_zustand.create)()((set, get) => ({
  scenes: /* @__PURE__ */ new Map(),
  stack: [],
  transition: { ...DEFAULT_TRANSITION },
  shared: {},
  // ── Navigation Actions ────────────────────────────────────────────────────
  go: (name, data, transition) => {
    const state = get();
    const entry = state.scenes.get(name);
    if (!entry) {
      console.warn(`[CarverJS Scene] Scene "${name}" is not registered`);
      return;
    }
    if (state.transition.active) return;
    const currentName = state.stack[state.stack.length - 1] ?? null;
    const resolved = resolveTransition(transition, entry.config);
    if (resolved.type === "none") {
      _exitStack(state, set);
      entry.status = "running";
      entry.data = data;
      set({
        scenes: new Map(state.scenes),
        stack: [name]
      });
      return;
    }
    if (currentName) {
      const current = state.scenes.get(currentName);
      if (current) current.status = "shutting_down";
    }
    entry.data = data;
    set({
      scenes: new Map(state.scenes),
      transition: {
        active: true,
        from: currentName,
        to: name,
        progress: 0,
        config: resolved,
        swapped: false,
        targetStack: [name]
      }
    });
  },
  push: (name, data, transition) => {
    const state = get();
    const entry = state.scenes.get(name);
    if (!entry) {
      console.warn(`[CarverJS Scene] Scene "${name}" is not registered`);
      return;
    }
    if (state.transition.active) return;
    const currentName = state.stack[state.stack.length - 1] ?? null;
    const resolved = resolveTransition(transition, entry.config);
    if (resolved.type === "none") {
      if (currentName) {
        const current = state.scenes.get(currentName);
        if (current) current.status = "sleeping";
      }
      entry.status = "running";
      entry.data = data;
      set({
        scenes: new Map(state.scenes),
        stack: [...state.stack, name]
      });
      return;
    }
    if (currentName) {
      const current = state.scenes.get(currentName);
      if (current) current.status = "shutting_down";
    }
    entry.data = data;
    set({
      scenes: new Map(state.scenes),
      stack: [...state.stack, name],
      transition: {
        active: true,
        from: currentName,
        to: name,
        progress: 0,
        config: resolved,
        swapped: false,
        targetStack: null
      }
    });
  },
  pop: (transition) => {
    const state = get();
    if (state.stack.length <= 1) return;
    if (state.transition.active) return;
    const currentName = state.stack[state.stack.length - 1];
    const previousName = state.stack[state.stack.length - 2];
    const currentEntry = state.scenes.get(currentName);
    const previousEntry = state.scenes.get(previousName);
    const resolved = resolveTransition(
      transition,
      previousEntry?.config
    );
    if (resolved.type === "none") {
      _destroyOrSleep(currentEntry);
      if (previousEntry) previousEntry.status = "running";
      set({
        scenes: new Map(state.scenes),
        stack: state.stack.slice(0, -1)
      });
      return;
    }
    if (currentEntry) currentEntry.status = "shutting_down";
    set({
      scenes: new Map(state.scenes),
      transition: {
        active: true,
        from: currentName,
        to: previousName,
        progress: 0,
        config: resolved,
        swapped: false,
        targetStack: state.stack.slice(0, -1)
      }
    });
  },
  replace: (name, data, transition) => {
    const state = get();
    const entry = state.scenes.get(name);
    if (!entry) {
      console.warn(`[CarverJS Scene] Scene "${name}" is not registered`);
      return;
    }
    if (state.transition.active) return;
    const currentName = state.stack[state.stack.length - 1] ?? null;
    const resolved = resolveTransition(transition, entry.config);
    if (resolved.type === "none") {
      if (currentName) {
        const current = state.scenes.get(currentName);
        _destroyOrSleep(current);
      }
      entry.status = "running";
      entry.data = data;
      const newStack2 = [...state.stack];
      if (newStack2.length > 0) {
        newStack2[newStack2.length - 1] = name;
      } else {
        newStack2.push(name);
      }
      set({
        scenes: new Map(state.scenes),
        stack: newStack2
      });
      return;
    }
    if (currentName) {
      const current = state.scenes.get(currentName);
      if (current) current.status = "shutting_down";
    }
    entry.data = data;
    const newStack = [...state.stack];
    if (newStack.length > 0) {
      newStack[newStack.length - 1] = name;
    } else {
      newStack.push(name);
    }
    set({
      scenes: new Map(state.scenes),
      stack: newStack,
      transition: {
        active: true,
        from: currentName,
        to: name,
        progress: 0,
        config: resolved,
        swapped: false,
        targetStack: null
      }
    });
  },
  // ── Shared Data ───────────────────────────────────────────────────────────
  setShared: (key, value) => {
    set((s) => ({ shared: { ...s.shared, [key]: value } }));
  },
  getShared: (key) => {
    return get().shared[key];
  },
  // ── Internal Actions ──────────────────────────────────────────────────────
  _registerScene: (config) => {
    const state = get();
    const existing = state.scenes.get(config.name);
    if (existing) {
      existing.config = config;
      set({ scenes: new Map(state.scenes) });
      return;
    }
    const entry = {
      config,
      status: "created",
      data: void 0
    };
    state.scenes.set(config.name, entry);
    set({ scenes: new Map(state.scenes) });
  },
  _unregisterScene: (name) => {
    const state = get();
    state.scenes.delete(name);
    set({
      scenes: new Map(state.scenes),
      stack: state.stack.filter((s) => s !== name)
    });
  },
  _setStatus: (name, status) => {
    const state = get();
    const entry = state.scenes.get(name);
    if (!entry) return;
    entry.status = status;
    set({ scenes: new Map(state.scenes) });
  },
  _tickTransition: (delta) => {
    const state = get();
    const t = state.transition;
    if (!t.active || !t.to) return;
    const duration = t.config.duration ?? 0.5;
    const newProgress = Math.min(t.progress + delta / duration, 1);
    if (!t.swapped && newProgress >= 0.5) {
      const fromEntry = t.from ? state.scenes.get(t.from) : null;
      const toEntry = state.scenes.get(t.to);
      if (fromEntry) {
        _destroyOrSleep(fromEntry);
      }
      if (toEntry) {
        toEntry.status = "running";
      }
      const updates = {
        scenes: new Map(state.scenes),
        transition: { ...t, progress: newProgress, swapped: true }
      };
      if (t.targetStack) {
        updates.stack = t.targetStack;
      }
      set(updates);
      return;
    }
    if (newProgress >= 1) {
      set({
        transition: { ...DEFAULT_TRANSITION }
      });
      return;
    }
    set({
      transition: { ...t, progress: newProgress }
    });
  },
  _startInitial: (name) => {
    const state = get();
    const entry = state.scenes.get(name);
    if (!entry) {
      console.warn(`[CarverJS Scene] Initial scene "${name}" is not registered`);
      return;
    }
    entry.status = "running";
    set({
      scenes: new Map(state.scenes),
      stack: [name]
    });
  }
}));
function _destroyOrSleep(entry) {
  if (!entry) return;
  entry.status = entry.config.persistent ? "sleeping" : "destroyed";
}
function _exitStack(state, set) {
  for (const name of state.stack) {
    const entry = state.scenes.get(name);
    _destroyOrSleep(entry);
  }
  set({ scenes: new Map(state.scenes) });
}

// src/systems/SceneManagerImpl.ts
var SceneManagerImpl = class {
  /** Navigate to a scene, clearing the stack */
  go(name, data, transition) {
    useSceneStore.getState().go(name, data, transition);
  }
  /** Push a scene onto the stack (current scene sleeps) */
  push(name, data, transition) {
    useSceneStore.getState().push(name, data, transition);
  }
  /** Pop the top scene, returning to the previous one */
  pop(transition) {
    useSceneStore.getState().pop(transition);
  }
  /** Replace the current scene without changing stack depth */
  replace(name, data, transition) {
    useSceneStore.getState().replace(name, data, transition);
  }
  /** Get the current active scene name */
  getCurrent() {
    const stack = useSceneStore.getState().stack;
    return stack[stack.length - 1] ?? null;
  }
  /** Get the full navigation stack */
  getStack() {
    return useSceneStore.getState().stack;
  }
  /** Whether a transition is currently in progress */
  isTransitioning() {
    return useSceneStore.getState().transition.active;
  }
  /** Set shared cross-scene data */
  setShared(key, value) {
    useSceneStore.getState().setShared(key, value);
  }
  /** Get shared cross-scene data */
  getShared(key) {
    return useSceneStore.getState().getShared(key);
  }
  /** Reset the store (called on destroy) */
  destroy() {
    useSceneStore.setState({
      scenes: /* @__PURE__ */ new Map(),
      stack: [],
      transition: {
        active: false,
        from: null,
        to: null,
        progress: 0,
        config: { type: "none" },
        swapped: false,
        targetStack: null
      },
      shared: {}
    });
  }
};
var _instance6 = null;
function getSceneManager() {
  if (!_instance6) _instance6 = new SceneManagerImpl();
  return _instance6;
}
function destroySceneManager() {
  _instance6?.destroy();
  _instance6 = null;
}

// src/systems/AssetManager.ts
var EXTENSION_MAP = /* @__PURE__ */ new Map([
  [".gltf", "gltf"],
  [".glb", "gltf"],
  [".png", "texture"],
  [".jpg", "texture"],
  [".jpeg", "texture"],
  [".webp", "texture"],
  [".svg", "texture"],
  [".mp3", "audio"],
  [".ogg", "audio"],
  [".wav", "audio"],
  [".m4a", "audio"],
  [".json", "json"],
  [".bin", "binary"],
  [".dat", "binary"]
]);
function detectAssetType(url) {
  const clean = url.split("?")[0].split("#")[0];
  const dot = clean.lastIndexOf(".");
  if (dot !== -1) {
    const ext = clean.slice(dot).toLowerCase();
    return EXTENSION_MAP.get(ext);
  }
  if (url.startsWith("data:")) {
    const mime = url.slice(5, url.indexOf(";"));
    if (mime.includes("gltf")) return "gltf";
    if (mime.startsWith("image/")) return "texture";
    if (mime.startsWith("audio/")) return "audio";
    if (mime === "application/json") return "json";
  }
  return void 0;
}
function resolvePriority(p) {
  if (typeof p === "number") return p;
  switch (p) {
    case "critical":
      return 100;
    case "high":
      return 75;
    case "low":
      return 25;
    case "lazy":
      return 0;
    case "normal":
    default:
      return 50;
  }
}
var _loaders = {
  gltf: async (url, options) => {
    const { GLTFLoader } = await import("three-stdlib");
    const loader = new GLTFLoader();
    if (options.draco) {
      const { DRACOLoader } = await import("three-stdlib");
      const draco = new DRACOLoader();
      draco.setDecoderPath(
        typeof options.draco === "string" ? options.draco : "https://www.gstatic.com/draco/versioned/decoders/1.5.7/"
      );
      loader.setDRACOLoader(draco);
    }
    if (options.meshopt) {
      const { MeshoptDecoder } = await import("three-stdlib");
      loader.setMeshoptDecoder(MeshoptDecoder);
    }
    return new Promise((resolve, reject) => {
      loader.load(url, resolve, void 0, reject);
    });
  },
  texture: async (url) => {
    const { TextureLoader: TextureLoader2 } = await import("three");
    return new Promise((resolve, reject) => {
      new TextureLoader2().load(url, resolve, void 0, reject);
    });
  },
  audio: async (url) => {
    const res = await fetch(url);
    if (!res.ok) throw new Error(`HTTP ${res.status}: ${url}`);
    return res.arrayBuffer();
  },
  json: async (url) => {
    const res = await fetch(url);
    if (!res.ok) throw new Error(`HTTP ${res.status}: ${url}`);
    return res.json();
  },
  binary: async (url) => {
    const res = await fetch(url);
    if (!res.ok) throw new Error(`HTTP ${res.status}: ${url}`);
    return res.arrayBuffer();
  }
};
function estimateSize(data, type) {
  switch (type) {
    case "audio":
    case "binary":
      return data.byteLength;
    case "texture": {
      const tex = data;
      if (tex.image)
        return (tex.image.width ?? 256) * (tex.image.height ?? 256) * 4;
      return 1024;
    }
    case "json":
      try {
        return JSON.stringify(data).length * 2;
      } catch {
        return 1024;
      }
    case "gltf":
      return 1024 * 1024;
    // Rough estimate; GLTF size varies widely
    default:
      return 1024;
  }
}
function disposeAsset(data, type) {
  try {
    if (type === "texture") {
      data.dispose?.();
    } else if (type === "gltf") {
      const gltf = data;
      gltf.scene?.traverse?.((child) => {
        const c = child;
        c.geometry?.dispose?.();
        if (c.material) {
          const mats = Array.isArray(c.material) ? c.material : [c.material];
          for (const m of mats) {
            const mat = m;
            mat.map?.dispose?.();
            mat.normalMap?.dispose?.();
            mat.roughnessMap?.dispose?.();
            mat.metalnessMap?.dispose?.();
            mat.dispose?.();
          }
        }
      });
    }
  } catch {
  }
}
var AssetManager = class {
  constructor() {
    // ── Cache ──
    this._cache = /* @__PURE__ */ new Map();
    this._groups = /* @__PURE__ */ new Map();
    this._totalBytes = 0;
    this._maxBytes = 256 * 1024 * 1024;
    // 256 MB
    // ── Custom loaders (override built-in) ──
    this._customLoaders = /* @__PURE__ */ new Map();
    // ── Loading state ──
    this._activeLoads = /* @__PURE__ */ new Map();
    // ── Configuration ──
    this._maxConcurrent = 6;
    this._retries = 3;
    this._retryDelay = 1e3;
    this._timeout = 3e4;
    // ── Progress (pre-allocated, mutated in place for zero-GC) ──
    this._progress = {
      phase: "idle",
      loaded: 0,
      total: 0,
      progress: 0,
      currentAsset: null,
      currentGroup: null,
      errors: []
    };
    // ── useSyncExternalStore subscriber pattern ──
    this._stateVersion = 0;
    this._subscribers = /* @__PURE__ */ new Set();
    this.subscribe = (cb) => {
      this._subscribers.add(cb);
      return () => {
        this._subscribers.delete(cb);
      };
    };
    this.getSnapshot = () => this._stateVersion;
  }
  _notify() {
    this._stateVersion++;
    for (const cb of this._subscribers) cb();
  }
  /** Current loading progress (read-only snapshot) */
  get progress() {
    return this._progress;
  }
  // ── Configuration ───────────────────────────────────────────────────────────
  configure(options) {
    if (options.maxConcurrent !== void 0)
      this._maxConcurrent = options.maxConcurrent;
    if (options.maxCacheBytes !== void 0)
      this._maxBytes = options.maxCacheBytes;
    if (options.retries !== void 0) this._retries = options.retries;
    if (options.retryDelay !== void 0) this._retryDelay = options.retryDelay;
    if (options.timeout !== void 0) this._timeout = options.timeout;
  }
  // ── Cache Access ────────────────────────────────────────────────────────────
  /** Get a loaded asset synchronously. Returns undefined if not cached. */
  get(key) {
    const entry = this._cache.get(key);
    if (!entry) return void 0;
    entry.lastAccessedAt = Date.now();
    return entry.data;
  }
  /** Check if an asset is loaded and cached. */
  has(key) {
    return this._cache.has(key);
  }
  /** Check if an asset is currently loading. */
  isLoading(key) {
    return this._activeLoads.has(key);
  }
  // ── Single Asset Load ───────────────────────────────────────────────────────
  /**
   * Load a single asset by URL. Returns cached result if available.
   * Deduplicates concurrent requests for the same key.
   */
  async load(url, options) {
    const key = options?.key ?? url;
    const cached = this.get(key);
    if (cached !== void 0) return cached;
    const type = options?.type ?? detectAssetType(url);
    if (!type)
      throw new Error(
        `[CarverJS] Cannot detect asset type for "${url}". Specify type explicitly.`
      );
    return this._loadWithRetry(
      url,
      key,
      type,
      options?.loaderOptions ?? {}
    );
  }
  // ── Manifest Load ──────────────────────────────────────────────────────────
  /**
   * Load all assets from a manifest with progress tracking.
   * Assets with priority "lazy" are skipped (loaded on-demand via useAssets).
   * Already-cached assets are skipped.
   */
  async loadManifest(manifest, callbacks) {
    const entries = this._parseManifest(manifest);
    const toLoad = entries.filter(
      (e) => resolvePriority(e.priority) > 0 && !this.has(e.key ?? e.url)
    );
    if (toLoad.length === 0) {
      const p2 = this._progress;
      p2.phase = "complete";
      p2.progress = 1;
      p2.loaded = 0;
      p2.total = 0;
      this._notify();
      callbacks?.onProgress?.(p2);
      callbacks?.onComplete?.();
      return;
    }
    toLoad.sort(
      (a, b) => resolvePriority(b.priority) - resolvePriority(a.priority)
    );
    for (const entry of toLoad) {
      if (entry.group) {
        let set = this._groups.get(entry.group);
        if (!set) {
          set = /* @__PURE__ */ new Set();
          this._groups.set(entry.group, set);
        }
        set.add(entry.key ?? entry.url);
      }
    }
    const p = this._progress;
    p.phase = "loading";
    p.loaded = 0;
    p.total = toLoad.length;
    p.progress = 0;
    p.currentAsset = null;
    p.currentGroup = null;
    p.errors = [];
    this._notify();
    callbacks?.onProgress?.(p);
    let idx = 0;
    const workers = [];
    const workerCount = Math.min(this._maxConcurrent, toLoad.length);
    for (let w = 0; w < workerCount; w++) {
      workers.push(
        (async () => {
          while (idx < toLoad.length) {
            const entry = toLoad[idx++];
            const key = entry.key ?? entry.url;
            const type = entry.type ?? detectAssetType(entry.url);
            if (!type) {
              p.errors.push({
                key,
                url: entry.url,
                message: `Cannot detect asset type for "${entry.url}"`,
                retries: 0,
                recoverable: false
              });
              p.loaded++;
              p.progress = p.total > 0 ? p.loaded / p.total : 1;
              this._notify();
              callbacks?.onProgress?.(p);
              continue;
            }
            p.currentAsset = key;
            p.currentGroup = entry.group ?? null;
            this._notify();
            callbacks?.onProgress?.(p);
            try {
              await this._loadWithRetry(
                entry.url,
                key,
                type,
                entry.loaderOptions ?? {}
              );
            } catch (err) {
              p.errors.push({
                key,
                url: entry.url,
                message: err instanceof Error ? err.message : String(err),
                retries: this._retries,
                recoverable: false
              });
            }
            p.loaded++;
            p.progress = p.total > 0 ? p.loaded / p.total : 1;
            this._notify();
            callbacks?.onProgress?.(p);
          }
        })()
      );
    }
    await Promise.all(workers);
    p.phase = p.errors.length > 0 ? "error" : "complete";
    p.currentAsset = null;
    p.currentGroup = null;
    this._notify();
    callbacks?.onProgress?.(p);
    if (p.errors.length > 0) callbacks?.onError?.(p.errors);
    callbacks?.onComplete?.();
  }
  // ── Load Group ──────────────────────────────────────────────────────────────
  /**
   * Load all assets in a named group.
   * Only works for groups that were registered via a previous loadManifest call.
   */
  async loadGroup(name) {
    const keys = this._groups.get(name);
    if (!keys) return;
    const toLoad = [];
    for (const key of keys) {
      if (!this.has(key)) {
        const type = detectAssetType(key);
        if (type) toLoad.push({ url: key, key, type });
      }
    }
    if (toLoad.length === 0) return;
    await this.loadManifest(toLoad);
  }
  // ── Retry Logic ─────────────────────────────────────────────────────────────
  _loadWithRetry(url, key, type, options) {
    const cached = this._cache.get(key);
    if (cached) return Promise.resolve(cached.data);
    const existing = this._activeLoads.get(key);
    if (existing) return existing;
    const promise = this._executeLoad(url, key, type, options);
    this._activeLoads.set(key, promise);
    return promise.finally(() => this._activeLoads.delete(key));
  }
  async _executeLoad(url, key, type, options) {
    const loader = this._customLoaders.get(type) ?? _loaders[type];
    if (!loader)
      throw new Error(`[CarverJS] No loader registered for type "${type}"`);
    let lastErr = null;
    for (let attempt = 0; attempt <= this._retries; attempt++) {
      try {
        const result = await Promise.race([
          loader(url, options),
          new Promise(
            (_, reject) => setTimeout(
              () => reject(new Error(`Timeout loading "${url}"`)),
              this._timeout
            )
          )
        ]);
        const sizeBytes = estimateSize(result, type);
        this._evictIfNeeded(sizeBytes);
        this._cache.set(key, {
          data: result,
          key,
          type,
          sizeBytes,
          lastAccessedAt: Date.now(),
          refCount: 0
        });
        this._totalBytes += sizeBytes;
        return result;
      } catch (err) {
        lastErr = err instanceof Error ? err : new Error(String(err));
        if (attempt < this._retries) {
          const delay = this._retryDelay * Math.pow(2, attempt) * (0.5 + Math.random());
          await new Promise((r) => setTimeout(r, delay));
        }
      }
    }
    throw lastErr;
  }
  // ── Manifest Parsing ────────────────────────────────────────────────────────
  _parseManifest(input) {
    const manifest = Array.isArray(input) ? { assets: input } : input;
    const base = manifest.baseUrl ?? "";
    return manifest.assets.map((entry) => {
      const isAbsolute = entry.url.startsWith("http") || entry.url.startsWith("/") || entry.url.startsWith("data:");
      const url = isAbsolute ? entry.url : base + entry.url;
      return { ...entry, url, key: entry.key ?? url };
    });
  }
  // ── Cache Eviction (LRU with reference counting) ────────────────────────────
  _evictIfNeeded(neededBytes) {
    while (this._totalBytes + neededBytes > this._maxBytes && this._cache.size > 0) {
      let oldest = null;
      for (const entry of this._cache.values()) {
        if (entry.refCount > 0) continue;
        if (!oldest || entry.lastAccessedAt < oldest.lastAccessedAt) {
          oldest = entry;
        }
      }
      if (!oldest) break;
      disposeAsset(oldest.data, oldest.type);
      this._totalBytes -= oldest.sizeBytes;
      this._cache.delete(oldest.key);
    }
  }
  // ── Cache Management ────────────────────────────────────────────────────────
  /** Unload a specific asset, freeing memory. Three.js resources are disposed. */
  unload(key) {
    const entry = this._cache.get(key);
    if (!entry) return;
    disposeAsset(entry.data, entry.type);
    this._totalBytes -= entry.sizeBytes;
    this._cache.delete(key);
  }
  /** Unload all assets in a group. */
  unloadGroup(group) {
    const keys = this._groups.get(group);
    if (!keys) return;
    for (const key of keys) this.unload(key);
    this._groups.delete(group);
  }
  /** Clear all cached assets. */
  clearAll() {
    for (const entry of this._cache.values()) {
      disposeAsset(entry.data, entry.type);
    }
    this._cache.clear();
    this._groups.clear();
    this._totalBytes = 0;
  }
  /** Get current memory usage estimate in bytes. */
  getMemoryUsage() {
    return this._totalBytes;
  }
  // ── Reference Counting ──────────────────────────────────────────────────────
  /** Increment reference count. Assets with refCount > 0 are never evicted. */
  retain(key) {
    const entry = this._cache.get(key);
    if (entry) entry.refCount++;
  }
  /** Decrement reference count. */
  release(key) {
    const entry = this._cache.get(key);
    if (entry && entry.refCount > 0) entry.refCount--;
  }
  // ── Custom Loaders ──────────────────────────────────────────────────────────
  /** Register a custom loader for an asset type. Overrides built-in loaders. */
  registerLoader(type, loader) {
    this._customLoaders.set(type, loader);
  }
  // ── Preload Helpers ─────────────────────────────────────────────────────────
  /** Fire-and-forget preload for one or more URLs. */
  preload(urls) {
    const list = Array.isArray(urls) ? urls : [urls];
    for (const url of list) {
      if (!this.has(url) && !this.isLoading(url)) {
        this.load(url).catch(() => {
        });
      }
    }
  }
  // ── Cleanup ─────────────────────────────────────────────────────────────────
  destroy() {
    this.clearAll();
    this._activeLoads.clear();
    this._customLoaders.clear();
    this._subscribers.clear();
    const p = this._progress;
    p.phase = "idle";
    p.loaded = 0;
    p.total = 0;
    p.progress = 0;
    p.currentAsset = null;
    p.currentGroup = null;
    p.errors = [];
  }
};
var _instance7 = null;
function getAssetManager() {
  if (!_instance7) _instance7 = new AssetManager();
  return _instance7;
}
function destroyAssetManager() {
  _instance7?.destroy();
  _instance7 = null;
}

// src/systems/Easing.ts
var PI = Math.PI;
var HALF_PI = PI / 2;
var c1 = 1.70158;
var c3 = c1 + 1;
var c2 = c1 * 1.525;
var c4 = 2 * PI / 3;
var c5 = 2 * PI / 4.5;
var n1 = 7.5625;
var d1 = 2.75;
function linear(t) {
  return t;
}
function quadIn(t) {
  return t * t;
}
function quadOut(t) {
  return 1 - (1 - t) * (1 - t);
}
function quadInOut(t) {
  return t < 0.5 ? 2 * t * t : 1 - (-2 * t + 2) * (-2 * t + 2) / 2;
}
function cubicIn(t) {
  return t * t * t;
}
function cubicOut(t) {
  return 1 - (1 - t) * (1 - t) * (1 - t);
}
function cubicInOut(t) {
  return t < 0.5 ? 4 * t * t * t : 1 - (-2 * t + 2) * (-2 * t + 2) * (-2 * t + 2) / 2;
}
function quartIn(t) {
  return t * t * t * t;
}
function quartOut(t) {
  return 1 - (1 - t) * (1 - t) * (1 - t) * (1 - t);
}
function quartInOut(t) {
  return t < 0.5 ? 8 * t * t * t * t : 1 - (-2 * t + 2) * (-2 * t + 2) * (-2 * t + 2) * (-2 * t + 2) / 2;
}
function quintIn(t) {
  return t * t * t * t * t;
}
function quintOut(t) {
  return 1 - (1 - t) * (1 - t) * (1 - t) * (1 - t) * (1 - t);
}
function quintInOut(t) {
  return t < 0.5 ? 16 * t * t * t * t * t : 1 - (-2 * t + 2) * (-2 * t + 2) * (-2 * t + 2) * (-2 * t + 2) * (-2 * t + 2) / 2;
}
function sineIn(t) {
  return 1 - Math.cos(t * HALF_PI);
}
function sineOut(t) {
  return Math.sin(t * HALF_PI);
}
function sineInOut(t) {
  return -(Math.cos(PI * t) - 1) / 2;
}
function expoIn(t) {
  return t === 0 ? 0 : Math.pow(2, 10 * t - 10);
}
function expoOut(t) {
  return t === 1 ? 1 : 1 - Math.pow(2, -10 * t);
}
function expoInOut(t) {
  if (t === 0) return 0;
  if (t === 1) return 1;
  return t < 0.5 ? Math.pow(2, 20 * t - 10) / 2 : (2 - Math.pow(2, -20 * t + 10)) / 2;
}
function circIn(t) {
  return 1 - Math.sqrt(1 - t * t);
}
function circOut(t) {
  return Math.sqrt(1 - (t - 1) * (t - 1));
}
function circInOut(t) {
  return t < 0.5 ? (1 - Math.sqrt(1 - 2 * t * (2 * t))) / 2 : (Math.sqrt(1 - (-2 * t + 2) * (-2 * t + 2)) + 1) / 2;
}
function backIn(t) {
  return c3 * t * t * t - c1 * t * t;
}
function backOut(t) {
  return 1 + c3 * (t - 1) * (t - 1) * (t - 1) + c1 * (t - 1) * (t - 1);
}
function backInOut(t) {
  return t < 0.5 ? 2 * t * (2 * t) * ((c2 + 1) * 2 * t - c2) / 2 : ((2 * t - 2) * (2 * t - 2) * ((c2 + 1) * (t * 2 - 2) + c2) + 2) / 2;
}
function elasticIn(t) {
  if (t === 0) return 0;
  if (t === 1) return 1;
  return -Math.pow(2, 10 * t - 10) * Math.sin((t * 10 - 10.75) * c4);
}
function elasticOut(t) {
  if (t === 0) return 0;
  if (t === 1) return 1;
  return Math.pow(2, -10 * t) * Math.sin((t * 10 - 0.75) * c4) + 1;
}
function elasticInOut(t) {
  if (t === 0) return 0;
  if (t === 1) return 1;
  return t < 0.5 ? -(Math.pow(2, 20 * t - 10) * Math.sin((20 * t - 11.125) * c5)) / 2 : Math.pow(2, -20 * t + 10) * Math.sin((20 * t - 11.125) * c5) / 2 + 1;
}
function bounceOut(t) {
  if (t < 1 / d1) return n1 * t * t;
  if (t < 2 / d1) return n1 * (t -= 1.5 / d1) * t + 0.75;
  if (t < 2.5 / d1) return n1 * (t -= 2.25 / d1) * t + 0.9375;
  return n1 * (t -= 2.625 / d1) * t + 0.984375;
}
function bounceIn(t) {
  return 1 - bounceOut(1 - t);
}
function bounceInOut(t) {
  return t < 0.5 ? (1 - bounceOut(1 - 2 * t)) / 2 : (1 + bounceOut(2 * t - 1)) / 2;
}
function spring(t) {
  return 1 - Math.exp(-6.9 * t) * Math.cos(6.9 * t);
}
var EASING_MAP = {
  linear,
  "quad.in": quadIn,
  "quad.out": quadOut,
  "quad.inOut": quadInOut,
  "cubic.in": cubicIn,
  "cubic.out": cubicOut,
  "cubic.inOut": cubicInOut,
  "quart.in": quartIn,
  "quart.out": quartOut,
  "quart.inOut": quartInOut,
  "quint.in": quintIn,
  "quint.out": quintOut,
  "quint.inOut": quintInOut,
  "sine.in": sineIn,
  "sine.out": sineOut,
  "sine.inOut": sineInOut,
  "expo.in": expoIn,
  "expo.out": expoOut,
  "expo.inOut": expoInOut,
  "circ.in": circIn,
  "circ.out": circOut,
  "circ.inOut": circInOut,
  "back.in": backIn,
  "back.out": backOut,
  "back.inOut": backInOut,
  "elastic.in": elasticIn,
  "elastic.out": elasticOut,
  "elastic.inOut": elasticInOut,
  "bounce.in": bounceIn,
  "bounce.out": bounceOut,
  "bounce.inOut": bounceInOut,
  spring
};
var BEZIER_SAMPLES = 11;
var BEZIER_STEP = 1 / (BEZIER_SAMPLES - 1);
var NEWTON_ITERATIONS = 4;
var NEWTON_MIN_SLOPE = 1e-3;
var SUBDIVISION_PRECISION = 1e-7;
var SUBDIVISION_MAX_ITERATIONS = 10;
function calcBezier(aT, a1, a2) {
  return (((1 - 3 * a2 + 3 * a1) * aT + (3 * a2 - 6 * a1)) * aT + 3 * a1) * aT;
}
function getSlope(aT, a1, a2) {
  return 3 * (1 - 3 * a2 + 3 * a1) * aT * aT + 2 * (3 * a2 - 6 * a1) * aT + 3 * a1;
}
function binarySubdivide(aX, aA, aB, x1, x2) {
  let currentX;
  let currentT;
  let i = 0;
  do {
    currentT = aA + (aB - aA) / 2;
    currentX = calcBezier(currentT, x1, x2) - aX;
    if (currentX > 0) {
      aB = currentT;
    } else {
      aA = currentT;
    }
  } while (Math.abs(currentX) > SUBDIVISION_PRECISION && ++i < SUBDIVISION_MAX_ITERATIONS);
  return currentT;
}
function newtonRaphson(aX, aGuessT, x1, x2) {
  for (let i = 0; i < NEWTON_ITERATIONS; ++i) {
    const slope = getSlope(aGuessT, x1, x2);
    if (slope === 0) return aGuessT;
    const currentX = calcBezier(aGuessT, x1, x2) - aX;
    aGuessT -= currentX / slope;
  }
  return aGuessT;
}
function cubicBezier(x1, y1, x2, y2) {
  if (x1 === y1 && x2 === y2) return linear;
  const sampleValues = new Float32Array(BEZIER_SAMPLES);
  for (let i = 0; i < BEZIER_SAMPLES; ++i) {
    sampleValues[i] = calcBezier(i * BEZIER_STEP, x1, x2);
  }
  function getTForX(aX) {
    let intervalStart = 0;
    let currentSample = 1;
    const lastSample = BEZIER_SAMPLES - 1;
    for (; currentSample !== lastSample && sampleValues[currentSample] <= aX; ++currentSample) {
      intervalStart += BEZIER_STEP;
    }
    --currentSample;
    const dist = (aX - sampleValues[currentSample]) / (sampleValues[currentSample + 1] - sampleValues[currentSample]);
    const guessForT = intervalStart + dist * BEZIER_STEP;
    const initialSlope = getSlope(guessForT, x1, x2);
    if (initialSlope >= NEWTON_MIN_SLOPE) {
      return newtonRaphson(aX, guessForT, x1, x2);
    }
    if (initialSlope === 0) {
      return guessForT;
    }
    return binarySubdivide(aX, intervalStart, intervalStart + BEZIER_STEP, x1, x2);
  }
  return (t) => {
    if (t === 0) return 0;
    if (t === 1) return 1;
    return calcBezier(getTForX(t), y1, y2);
  };
}
function steps(count, jumpEnd = true) {
  return (t) => {
    if (jumpEnd) {
      return Math.min(Math.floor(t * count), count) / count;
    }
    return Math.ceil(t * count - 1) / count;
  };
}
function resolveEasing(input) {
  if (typeof input === "function") return input;
  if (Array.isArray(input)) return cubicBezier(input[0], input[1], input[2], input[3]);
  return EASING_MAP[input] ?? linear;
}
var Easing = EASING_MAP;

// src/store/gameStore.ts
var import_zustand2 = require("zustand");
var useGameStore = (0, import_zustand2.create)()((set, get) => ({
  phase: "loading",
  elapsed: 0,
  frameCount: 0,
  setPhase: (phase) => set({ phase }),
  _tick: (delta) => {
    if (get().phase !== "playing") return;
    set((s) => ({
      elapsed: s.elapsed + delta,
      frameCount: s.frameCount + 1
    }));
  },
  pause: () => {
    if (get().phase === "playing") set({ phase: "paused" });
  },
  resume: () => {
    if (get().phase === "paused") set({ phase: "playing" });
  },
  reset: () => set({ phase: "playing", elapsed: 0, frameCount: 0 })
}));

// src/systems/TweenManager.ts
var INITIAL_POOL_SIZE = 64;
var MAX_POOL_SIZE = 512;
var MAX_PROPERTIES = 16;
var _tweenIdCounter = 0;
var Tween = class {
  constructor() {
    this.id = 0;
    this.state = "idle";
    this.direction = "forward";
    // Target
    this.target = null;
    // Config
    this.duration = 0.3;
    this.delay = 0;
    this.easeFn = (t) => t;
    this.repeat = 0;
    this.repeatDelay = 0;
    this.yoyo = false;
    this.speed = 1;
    this.group = "";
    this.persist = false;
    this.ignorePause = false;
    this.propCount = 0;
    // Timing
    this.elapsed = 0;
    this.delayElapsed = 0;
    this.repeatCount = 0;
    this.progress = 0;
    this._totalRepeat = 0;
    // original repeat value for restart
    // Callbacks
    this.onStart = null;
    this.onUpdate = null;
    this.onComplete = null;
    this.onRepeat = null;
    this.onYoyo = null;
    this.onKill = null;
    // Number tween
    this._isNumberTween = false;
    this._numberOnUpdate = null;
    // Chain
    this._chainConfigs = [];
    // Promise (.finished) — lazy
    this._resolve = null;
    this._promise = null;
    this.props = [];
    for (let i = 0; i < MAX_PROPERTIES; i++) {
      this.props.push({ key: "", parts: [], from: 0, to: 0, delta: 0 });
    }
  }
  reset() {
    this.id = 0;
    this.state = "idle";
    this.direction = "forward";
    this.target = null;
    this.duration = 0.3;
    this.delay = 0;
    this.easeFn = (t) => t;
    this.repeat = 0;
    this.repeatDelay = 0;
    this.yoyo = false;
    this.speed = 1;
    this.group = "";
    this.persist = false;
    this.ignorePause = false;
    this.propCount = 0;
    this.elapsed = 0;
    this.delayElapsed = 0;
    this.repeatCount = 0;
    this.progress = 0;
    this._totalRepeat = 0;
    this.onStart = null;
    this.onUpdate = null;
    this.onComplete = null;
    this.onRepeat = null;
    this.onYoyo = null;
    this.onKill = null;
    this._isNumberTween = false;
    this._numberOnUpdate = null;
    this._chainConfigs.length = 0;
    if (this._resolve) this._resolve();
    this._resolve = null;
    this._promise = null;
  }
};
function createControls(tween, manager) {
  const capturedId = tween.id;
  function isAlive() {
    return tween.id === capturedId && tween.state !== "killed";
  }
  const controls = {
    get id() {
      return capturedId;
    },
    get state() {
      return tween.id === capturedId ? tween.state : "killed";
    },
    get progress() {
      return tween.id === capturedId ? tween.progress : 0;
    },
    pause() {
      if (isAlive() && tween.state === "playing") {
        tween.state = "pending";
      }
      return controls;
    },
    resume() {
      if (isAlive() && tween.state === "pending") {
        tween.state = "playing";
      }
      return controls;
    },
    kill() {
      if (isAlive()) {
        manager._killTween(tween);
      }
      return controls;
    },
    restart() {
      if (tween.id === capturedId) {
        tween.elapsed = 0;
        tween.delayElapsed = 0;
        tween.repeatCount = 0;
        tween.direction = "forward";
        tween.repeat = tween._totalRepeat;
        tween.state = tween.delay > 0 ? "delayed" : "playing";
        tween.onStart?.();
      }
      return controls;
    },
    seek(p) {
      if (!isAlive()) return controls;
      const clamped = Math.max(0, Math.min(1, p));
      tween.elapsed = clamped * tween.duration;
      tween.progress = clamped;
      const eased = tween.easeFn(
        tween.direction === "reverse" ? 1 - clamped : clamped
      );
      for (let i = 0; i < tween.propCount; i++) {
        const prop = tween.props[i];
        writeProperty(
          tween.target,
          prop.parts,
          prop.from + prop.delta * eased
        );
      }
      return controls;
    },
    setSpeed(multiplier) {
      if (isAlive()) tween.speed = Math.max(0, multiplier);
      return controls;
    },
    chain(config) {
      if (isAlive()) tween._chainConfigs.push(config);
      return controls;
    },
    get finished() {
      if (tween.id !== capturedId) return Promise.resolve();
      if (tween.state === "completed") return Promise.resolve();
      if (!tween._promise) {
        tween._promise = new Promise((resolve) => {
          tween._resolve = resolve;
        });
      }
      return tween._promise;
    }
  };
  return controls;
}
function readProperty(target, parts) {
  let obj = target;
  for (let i = 0; i < parts.length; i++) {
    obj = obj[parts[i]];
  }
  return obj;
}
function writeProperty(target, parts, value) {
  let obj = target;
  for (let i = 0; i < parts.length - 1; i++) {
    obj = obj[parts[i]];
  }
  obj[parts[parts.length - 1]] = value;
}
var DEAD_CONTROLS = {
  get id() {
    return -1;
  },
  get state() {
    return "killed";
  },
  get progress() {
    return 0;
  },
  pause() {
    return DEAD_CONTROLS;
  },
  resume() {
    return DEAD_CONTROLS;
  },
  kill() {
    return DEAD_CONTROLS;
  },
  restart() {
    return DEAD_CONTROLS;
  },
  seek() {
    return DEAD_CONTROLS;
  },
  setSpeed() {
    return DEAD_CONTROLS;
  },
  chain() {
    return DEAD_CONTROLS;
  },
  get finished() {
    return Promise.resolve();
  }
};
var _timelineIdCounter = 0;
var Timeline = class {
  constructor(manager, config) {
    this.state = "idle";
    this.ignorePause = false;
    this.speed = 1;
    this.repeat = 0;
    this.repeatDelay = 0;
    this.yoyo = false;
    this.group = "";
    this._entries = [];
    this._labels = /* @__PURE__ */ new Map();
    this._playhead = 0;
    this._totalDuration = 0;
    this._lastEntryEnd = 0;
    this._direction = "forward";
    this._repeatCount = 0;
    this._totalRepeat = 0;
    this._started = false;
    // Callbacks
    this.onStart = null;
    this.onUpdate = null;
    this.onComplete = null;
    this.onRepeat = null;
    // Promise
    this._resolve = null;
    this._promise = null;
    this._manager = manager;
    this.id = ++_timelineIdCounter;
    if (config) {
      this.repeat = config.repeat ?? 0;
      this.repeatDelay = config.repeatDelay ?? 0;
      this.yoyo = config.yoyo ?? false;
      this.speed = config.speed ?? 1;
      this.ignorePause = config.ignorePause ?? false;
      this.group = config.group ?? "";
      this.onStart = config.onStart ?? null;
      this.onUpdate = config.onUpdate ?? null;
      this.onComplete = config.onComplete ?? null;
      this.onRepeat = config.onRepeat ?? null;
    }
    this._totalRepeat = this.repeat;
    this.state = config?.paused ? "pending" : "playing";
    this.controls = this._createControls();
  }
  _createControls() {
    const self = this;
    const ctrl = {
      get id() {
        return self.id;
      },
      get state() {
        return self.state;
      },
      get progress() {
        return self._totalDuration > 0 ? self._playhead / self._totalDuration : 0;
      },
      get duration() {
        return self._totalDuration;
      },
      add(config, position) {
        self._addTween(config, position);
        return ctrl;
      },
      addCallback(callback, position) {
        const time = self._resolvePosition(position);
        self._entries.push({
          type: "callback",
          startTime: time,
          duration: 0,
          callback,
          started: false,
          completed: false
        });
        self._recalcDuration();
        return ctrl;
      },
      addLabel(name, position) {
        const time = self._resolvePosition(position);
        self._labels.set(name, time);
        self._entries.push({
          type: "label",
          startTime: time,
          duration: 0,
          label: name,
          started: false,
          completed: false
        });
        if (time > self._lastEntryEnd) self._lastEntryEnd = time;
        return ctrl;
      },
      play() {
        if (self.state === "pending" || self.state === "idle") {
          self.state = "playing";
        }
        return ctrl;
      },
      pause() {
        if (self.state === "playing") self.state = "pending";
        return ctrl;
      },
      resume() {
        if (self.state === "pending") self.state = "playing";
        return ctrl;
      },
      kill() {
        self._killAll();
        self.state = "killed";
        if (self._resolve) self._resolve();
        return ctrl;
      },
      restart() {
        self._killAll();
        self._playhead = 0;
        self._repeatCount = 0;
        self._direction = "forward";
        self.repeat = self._totalRepeat;
        self._started = false;
        for (let i = 0; i < self._entries.length; i++) {
          self._entries[i].started = false;
          self._entries[i].completed = false;
          self._entries[i].controls = void 0;
        }
        self.state = "playing";
        return ctrl;
      },
      seek(timeOrLabel) {
        let targetTime;
        if (typeof timeOrLabel === "string") {
          targetTime = self._labels.get(timeOrLabel) ?? 0;
        } else {
          targetTime = timeOrLabel;
        }
        self._playhead = Math.max(0, Math.min(targetTime, self._totalDuration));
        return ctrl;
      },
      setSpeed(multiplier) {
        self.speed = Math.max(0, multiplier);
        return ctrl;
      },
      get finished() {
        if (self.state === "completed") return Promise.resolve();
        if (!self._promise) {
          self._promise = new Promise((resolve) => {
            self._resolve = resolve;
          });
        }
        return self._promise;
      }
    };
    return ctrl;
  }
  _addTween(config, position) {
    const time = this._resolvePosition(position);
    const dur = config.duration ?? 0.3;
    this._entries.push({
      type: "tween",
      startTime: time,
      duration: dur,
      tweenConfig: config,
      started: false,
      completed: false
    });
    const end = time + dur;
    if (end > this._lastEntryEnd) this._lastEntryEnd = end;
    this._recalcDuration();
  }
  _resolvePosition(pos) {
    if (pos === void 0) return this._lastEntryEnd;
    if (typeof pos === "number") return Math.max(0, pos);
    if (pos === "<") {
      return this._entries.length > 0 ? this._entries[this._entries.length - 1].startTime : 0;
    }
    const prevRelMatch = pos.match(/^<([+-]=)(\d+\.?\d*)$/);
    if (prevRelMatch) {
      const base = this._entries.length > 0 ? this._entries[this._entries.length - 1].startTime : 0;
      const sign = prevRelMatch[1] === "+=" ? 1 : -1;
      return Math.max(0, base + sign * parseFloat(prevRelMatch[2]));
    }
    const relMatch = pos.match(/^([+-]=)(\d+\.?\d*)$/);
    if (relMatch) {
      const sign = relMatch[1] === "+=" ? 1 : -1;
      return Math.max(0, this._lastEntryEnd + sign * parseFloat(relMatch[2]));
    }
    const labelMatch = pos.match(/^([a-zA-Z_]\w*)(?:([+-]=)(\d+\.?\d*))?$/);
    if (labelMatch) {
      const labelTime = this._labels.get(labelMatch[1]) ?? this._lastEntryEnd;
      if (labelMatch[2]) {
        const sign = labelMatch[2] === "+=" ? 1 : -1;
        return Math.max(0, labelTime + sign * parseFloat(labelMatch[3]));
      }
      return labelTime;
    }
    return this._lastEntryEnd;
  }
  _recalcDuration() {
    let max = 0;
    for (let i = 0; i < this._entries.length; i++) {
      const end = this._entries[i].startTime + this._entries[i].duration;
      if (end > max) max = end;
    }
    this._totalDuration = max;
  }
  _killAll() {
    for (let i = 0; i < this._entries.length; i++) {
      const entry = this._entries[i];
      if (entry.controls && entry.controls.state !== "completed" && entry.controls.state !== "killed") {
        entry.controls.kill();
      }
    }
  }
  tick(delta) {
    if (this.state !== "playing") return;
    if (!this._started) {
      this._started = true;
      this.onStart?.();
    }
    const dt = delta * this.speed;
    this._playhead += dt;
    for (let i = 0; i < this._entries.length; i++) {
      const entry = this._entries[i];
      if (entry.started || this._playhead < entry.startTime) continue;
      entry.started = true;
      if (entry.type === "tween" && entry.tweenConfig) {
        entry.controls = this._manager.create(entry.tweenConfig);
      } else if (entry.type === "callback" && entry.callback) {
        entry.callback();
        entry.completed = true;
      }
    }
    if (this._totalDuration > 0) {
      this.onUpdate?.(Math.min(this._playhead / this._totalDuration, 1));
    }
    if (this._playhead >= this._totalDuration) {
      if (this.repeat === -1 || this._repeatCount < this.repeat) {
        this._repeatCount++;
        this._playhead = 0;
        this._started = false;
        if (this.yoyo) {
          this._direction = this._direction === "forward" ? "reverse" : "forward";
        }
        for (let i = 0; i < this._entries.length; i++) {
          this._entries[i].started = false;
          this._entries[i].completed = false;
          this._entries[i].controls = void 0;
        }
        this.onRepeat?.(this._repeatCount);
      } else {
        this.state = "completed";
        this.onComplete?.();
        if (this._resolve) this._resolve();
      }
    }
  }
};
var TweenManager = class {
  constructor() {
    // Object pool
    this._pool = [];
    this._freeStack = [];
    // Active list (swap-and-pop)
    this._active = [];
    this._activeCount = 0;
    // Groups
    this._groups = /* @__PURE__ */ new Map();
    // Timelines
    this._timelines = [];
    this._timelineCount = 0;
    this._growPool(INITIAL_POOL_SIZE);
  }
  // ── Pool Management ──
  _growPool(count) {
    const newTotal = Math.min(this._pool.length + count, MAX_POOL_SIZE);
    for (let i = this._pool.length; i < newTotal; i++) {
      this._pool.push(new Tween());
      this._freeStack.push(i);
    }
  }
  _acquire() {
    if (this._freeStack.length === 0) {
      if (this._pool.length < MAX_POOL_SIZE) {
        this._growPool(Math.min(64, MAX_POOL_SIZE - this._pool.length));
      }
    }
    if (this._freeStack.length === 0) {
      return null;
    }
    return this._pool[this._freeStack.pop()];
  }
  _release(tween) {
    const idx = this._pool.indexOf(tween);
    if (idx !== -1) {
      tween.reset();
      this._freeStack.push(idx);
    }
  }
  // ── Active List (swap-and-pop) ──
  _addActive(tween) {
    if (this._activeCount < this._active.length) {
      this._active[this._activeCount] = tween;
    } else {
      this._active.push(tween);
    }
    this._activeCount++;
  }
  _removeActive(index) {
    const last = this._activeCount - 1;
    if (index !== last) {
      this._active[index] = this._active[last];
    }
    this._activeCount--;
  }
  // ── Create API ──
  create(config) {
    const tween = this._acquire();
    if (!tween) return DEAD_CONTROLS;
    tween.id = ++_tweenIdCounter;
    tween.target = config.target;
    tween.duration = config.duration ?? 0.3;
    tween.delay = config.delay ?? 0;
    tween.easeFn = resolveEasing(config.ease ?? "quad.out");
    tween.repeat = config.repeat ?? 0;
    tween.repeatDelay = config.repeatDelay ?? 0;
    tween.yoyo = config.yoyo ?? false;
    tween.speed = config.speed ?? 1;
    tween.group = config.group ?? "";
    tween.persist = config.persist ?? false;
    tween.ignorePause = config.ignorePause ?? false;
    tween._totalRepeat = tween.repeat;
    tween.onStart = config.onStart ?? null;
    tween.onUpdate = config.onUpdate ?? null;
    tween.onComplete = config.onComplete ?? null;
    tween.onRepeat = config.onRepeat ?? null;
    tween.onYoyo = config.onYoyo ?? null;
    tween.onKill = config.onKill ?? null;
    this._resolveProps(tween, config.to, config.from);
    if (tween.group) {
      let set = this._groups.get(tween.group);
      if (!set) {
        set = /* @__PURE__ */ new Set();
        this._groups.set(tween.group, set);
      }
      set.add(tween);
    }
    if (tween.delay > 0) {
      tween.state = "delayed";
    } else {
      tween.state = "playing";
      tween.onStart?.();
    }
    this._addActive(tween);
    return createControls(tween, this);
  }
  createNumber(config) {
    const tween = this._acquire();
    if (!tween) return DEAD_CONTROLS;
    tween.id = ++_tweenIdCounter;
    tween._isNumberTween = true;
    const scratchTarget = { value: config.from };
    tween.target = scratchTarget;
    tween.duration = config.duration ?? 0.3;
    tween.delay = config.delay ?? 0;
    tween.easeFn = resolveEasing(config.ease ?? "quad.out");
    tween.repeat = config.repeat ?? 0;
    tween.repeatDelay = config.repeatDelay ?? 0;
    tween.yoyo = config.yoyo ?? false;
    tween.speed = config.speed ?? 1;
    tween.group = config.group ?? "";
    tween.persist = false;
    tween.ignorePause = config.ignorePause ?? false;
    tween._totalRepeat = tween.repeat;
    tween.onStart = config.onStart ?? null;
    tween.onComplete = config.onComplete ?? null;
    tween.onRepeat = config.onRepeat ?? null;
    tween.onYoyo = config.onYoyo ?? null;
    tween.onKill = config.onKill ?? null;
    tween._numberOnUpdate = config.onUpdate ?? null;
    tween.propCount = 1;
    const prop = tween.props[0];
    prop.key = "value";
    prop.parts = ["value"];
    prop.from = config.from;
    prop.to = config.to;
    prop.delta = config.to - config.from;
    if (tween.group) {
      let set = this._groups.get(tween.group);
      if (!set) {
        set = /* @__PURE__ */ new Set();
        this._groups.set(tween.group, set);
      }
      set.add(tween);
    }
    if (tween.delay > 0) {
      tween.state = "delayed";
    } else {
      tween.state = "playing";
      tween.onStart?.();
    }
    this._addActive(tween);
    return createControls(tween, this);
  }
  createTimeline(config) {
    const tl = new Timeline(this, config);
    if (this._timelineCount < this._timelines.length) {
      this._timelines[this._timelineCount] = tl;
    } else {
      this._timelines.push(tl);
    }
    this._timelineCount++;
    return tl.controls;
  }
  // ── Property Resolution ──
  _resolveProps(tween, to, from) {
    let count = 0;
    for (const key in to) {
      if (count >= MAX_PROPERTIES) break;
      const prop = tween.props[count];
      prop.key = key;
      const parts = key.split(".");
      prop.parts = parts;
      prop.to = to[key];
      prop.from = from?.[key] ?? readProperty(tween.target, parts);
      prop.delta = prop.to - prop.from;
      count++;
    }
    tween.propCount = count;
  }
  // ── Tick (Hot Path) ──
  tick(delta) {
    const phase = useGameStore.getState().phase;
    if (phase === "loading") return;
    const isPlaying = phase === "playing";
    for (let i = this._activeCount - 1; i >= 0; i--) {
      const tw = this._active[i];
      if (!isPlaying && !tw.ignorePause) continue;
      if (tw.state === "pending") continue;
      const dt = delta * tw.speed;
      if (tw.state === "delayed") {
        tw.delayElapsed += dt;
        if (tw.delayElapsed < tw.delay) continue;
        const overflow = tw.delayElapsed - tw.delay;
        tw.state = "playing";
        tw.delayElapsed = 0;
        tw.onStart?.();
        if (overflow > 0) {
          tw.elapsed += overflow;
        }
      }
      if (tw.state === "playing") {
        tw.elapsed += dt;
        let rawProgress;
        if (tw.duration <= 0) {
          rawProgress = 1;
        } else {
          rawProgress = Math.min(tw.elapsed / tw.duration, 1);
        }
        const effectiveProgress = tw.direction === "reverse" ? 1 - rawProgress : rawProgress;
        const eased = tw.easeFn(effectiveProgress);
        tw.progress = eased;
        for (let p = 0; p < tw.propCount; p++) {
          const prop = tw.props[p];
          const value = prop.from + prop.delta * eased;
          writeProperty(
            tw.target,
            prop.parts,
            value
          );
        }
        if (tw._isNumberTween && tw._numberOnUpdate) {
          tw._numberOnUpdate(
            tw.target.value,
            eased
          );
        }
        tw.onUpdate?.(eased);
        if (rawProgress >= 1) {
          this._onComplete(tw, i);
        }
      }
    }
    for (let i = this._timelineCount - 1; i >= 0; i--) {
      const tl = this._timelines[i];
      if (!isPlaying && !tl.ignorePause) continue;
      tl.tick(delta);
      if (tl.state === "completed" || tl.state === "killed") {
        const last = this._timelineCount - 1;
        if (i !== last) this._timelines[i] = this._timelines[last];
        this._timelineCount--;
      }
    }
  }
  _onComplete(tw, activeIndex) {
    if (tw.repeat === -1 || tw.repeatCount < tw.repeat) {
      tw.repeatCount++;
      const overflow = tw.elapsed - tw.duration;
      tw.elapsed = 0;
      if (tw.yoyo) {
        tw.direction = tw.direction === "forward" ? "reverse" : "forward";
        tw.onYoyo?.(tw.direction);
      }
      if (tw.repeatDelay > 0) {
        tw.state = "delayed";
        tw.delay = tw.repeatDelay;
        tw.delayElapsed = 0;
      } else if (overflow > 0) {
        tw.elapsed = overflow;
      }
      tw.onRepeat?.(tw.repeatCount);
      return;
    }
    tw.state = "completed";
    tw.onComplete?.();
    if (tw._resolve) tw._resolve();
    for (let c = 0; c < tw._chainConfigs.length; c++) {
      this.create(tw._chainConfigs[c]);
    }
    this._removeActive(activeIndex);
    if (tw.group) {
      this._groups.get(tw.group)?.delete(tw);
    }
    if (!tw.persist) {
      this._release(tw);
    }
  }
  // ── Kill ──
  _killTween(tween) {
    if (tween.state === "killed" || tween.state === "idle") return;
    tween.state = "killed";
    tween.onKill?.();
    if (tween._resolve) tween._resolve();
    if (tween.group) {
      this._groups.get(tween.group)?.delete(tween);
    }
    for (let i = 0; i < this._activeCount; i++) {
      if (this._active[i] === tween) {
        this._removeActive(i);
        break;
      }
    }
    this._release(tween);
  }
  // ── Kill by ID (for hook cleanup) ──
  killById(id) {
    for (let i = 0; i < this._activeCount; i++) {
      if (this._active[i].id === id) {
        this._killTween(this._active[i]);
        return;
      }
    }
  }
  // ── Group API ──
  getGroup(name) {
    const self = this;
    return {
      pause() {
        const set = self._groups.get(name);
        if (!set) return;
        for (const tw of set) {
          if (tw.state === "playing") tw.state = "pending";
        }
      },
      resume() {
        const set = self._groups.get(name);
        if (!set) return;
        for (const tw of set) {
          if (tw.state === "pending") tw.state = "playing";
        }
      },
      kill() {
        const set = self._groups.get(name);
        if (!set) return;
        const tweens = Array.from(set);
        for (const tw of tweens) {
          self._killTween(tw);
        }
      },
      get count() {
        return self._groups.get(name)?.size ?? 0;
      }
    };
  }
  // ── Global Controls ──
  pauseAll() {
    for (let i = 0; i < this._activeCount; i++) {
      if (this._active[i].state === "playing") {
        this._active[i].state = "pending";
      }
    }
  }
  resumeAll() {
    for (let i = 0; i < this._activeCount; i++) {
      if (this._active[i].state === "pending") {
        this._active[i].state = "playing";
      }
    }
  }
  killAll() {
    for (let i = this._activeCount - 1; i >= 0; i--) {
      this._killTween(this._active[i]);
    }
    for (let i = this._timelineCount - 1; i >= 0; i--) {
      this._timelines[i].controls.kill();
    }
    this._timelineCount = 0;
  }
  // ── Destroy ──
  destroy() {
    this.killAll();
    this._pool.length = 0;
    this._freeStack.length = 0;
    this._active.length = 0;
    this._activeCount = 0;
    this._groups.clear();
    this._timelines.length = 0;
    this._timelineCount = 0;
  }
};
var _instance8 = null;
function getTweenManager() {
  if (!_instance8) _instance8 = new TweenManager();
  return _instance8;
}
function destroyTweenManager() {
  _instance8?.destroy();
  _instance8 = null;
}

// src/systems/ParticleManager.ts
var import_three3 = require("three");
var VERTEX_SHADER = (
  /* glsl */
  `
precision highp float;

uniform mat4 projectionMatrix;
uniform mat4 viewMatrix;
uniform mat4 modelMatrix;

attribute vec3 position;
attribute vec2 uv;
attribute mat4 instanceMatrix;
attribute vec3 instanceColor;
attribute float instanceAlpha;
attribute float instanceSpriteFrame;
attribute float instanceRotation;

varying float vAlpha;
varying vec2 vUv;
varying vec3 vColor;

uniform float uColumns;
uniform float uRows;
uniform bool uBillboard;

void main() {
  vAlpha = instanceAlpha;
  vColor = instanceColor;

  // Sprite sheet UV
  float col = mod(instanceSpriteFrame, uColumns);
  float row = floor(instanceSpriteFrame / uColumns);
  vec2 uvScale = vec2(1.0 / uColumns, 1.0 / uRows);
  vec2 uvOffset = vec2(col / uColumns, 1.0 - (row + 1.0) / uRows);
  vUv = uv * uvScale + uvOffset;

  if (uBillboard) {
    // Extract position and scale from instanceMatrix
    vec3 instancePos = vec3(instanceMatrix[3]);
    float scaleX = length(instanceMatrix[0].xyz);
    float scaleY = length(instanceMatrix[1].xyz);

    // Transform to world space
    vec4 worldPos4 = modelMatrix * vec4(instancePos, 1.0);
    vec3 worldPos = worldPos4.xyz;

    // Camera vectors in world space
    vec3 camRight = vec3(viewMatrix[0][0], viewMatrix[1][0], viewMatrix[2][0]);
    vec3 camUp = vec3(viewMatrix[0][1], viewMatrix[1][1], viewMatrix[2][1]);

    // Apply particle rotation around view axis
    float c = cos(instanceRotation);
    float s = sin(instanceRotation);
    vec3 rotRight = camRight * c + camUp * s;
    vec3 rotUp = -camRight * s + camUp * c;

    // Construct billboard vertex
    vec3 finalPos = worldPos
      + rotRight * position.x * scaleX
      + rotUp * position.y * scaleY;

    gl_Position = projectionMatrix * viewMatrix * vec4(finalPos, 1.0);
  } else {
    gl_Position = projectionMatrix * viewMatrix * modelMatrix * instanceMatrix * vec4(position, 1.0);
  }
}
`
);
var FRAGMENT_SHADER = (
  /* glsl */
  `
precision highp float;

uniform sampler2D uTexture;
uniform bool uHasTexture;

varying float vAlpha;
varying vec2 vUv;
varying vec3 vColor;

void main() {
  vec4 texColor = uHasTexture ? texture2D(uTexture, vUv) : vec4(1.0);
  float alpha = texColor.a * vAlpha;
  if (alpha < 0.01) discard;
  gl_FragColor = vec4(vColor * texColor.rgb, alpha);
}
`
);
var _textureLoader = new import_three3.TextureLoader();
function sampleValueRange(range) {
  if (typeof range === "number") return range;
  return range[0] + Math.random() * (range[1] - range[0]);
}
var _colorA = new import_three3.Color();
var _colorB = new import_three3.Color();
function sampleColorRange(range) {
  if (Array.isArray(range) && range.length === 2) {
    _colorA.set(range[0]);
    _colorB.set(range[1]);
    const t = Math.random();
    return [
      _colorA.r + (_colorB.r - _colorA.r) * t,
      _colorA.g + (_colorB.g - _colorA.g) * t,
      _colorA.b + (_colorB.b - _colorA.b) * t
    ];
  }
  _colorA.set(range);
  return [_colorA.r, _colorA.g, _colorA.b];
}
function evaluateCurve(curve, t) {
  const len = curve.length;
  if (len === 0) return 1;
  if (t <= curve[0].t) return curve[0].value;
  if (t >= curve[len - 1].t) return curve[len - 1].value;
  let lo = 0;
  let hi = len - 1;
  while (hi - lo > 1) {
    const mid = lo + hi >> 1;
    if (curve[mid].t <= t) lo = mid;
    else hi = mid;
  }
  const a = curve[lo];
  const b = curve[hi];
  const frac = (t - a.t) / (b.t - a.t);
  return a.value + (b.value - a.value) * frac;
}
function evaluateColorGradient(gradient, t) {
  const len = gradient.length;
  if (len === 0) return [1, 1, 1];
  if (t <= gradient[0].t) {
    _colorA.set(gradient[0].color);
    return [_colorA.r, _colorA.g, _colorA.b];
  }
  if (t >= gradient[len - 1].t) {
    _colorA.set(gradient[len - 1].color);
    return [_colorA.r, _colorA.g, _colorA.b];
  }
  let lo = 0;
  let hi = len - 1;
  while (hi - lo > 1) {
    const mid = lo + hi >> 1;
    if (gradient[mid].t <= t) lo = mid;
    else hi = mid;
  }
  _colorA.set(gradient[lo].color);
  _colorB.set(gradient[hi].color);
  const frac = (t - gradient[lo].t) / (gradient[hi].t - gradient[lo].t);
  return [
    _colorA.r + (_colorB.r - _colorA.r) * frac,
    _colorA.g + (_colorB.g - _colorA.g) * frac,
    _colorA.b + (_colorB.b - _colorA.b) * frac
  ];
}
function resolveBlending(mode) {
  switch (mode) {
    case "additive":
      return { blending: import_three3.AdditiveBlending };
    case "multiply":
      return { blending: import_three3.MultiplyBlending };
    case "screen":
      return {
        blending: import_three3.CustomBlending,
        blendSrc: import_three3.OneFactor,
        blendDst: import_three3.OneMinusSrcColorFactor
      };
    default:
      return { blending: import_three3.NormalBlending };
  }
}
function loadTexture(src) {
  if (typeof src === "string") return _textureLoader.load(src);
  return src;
}
function createParticleData(max) {
  return {
    posX: new Float32Array(max),
    posY: new Float32Array(max),
    posZ: new Float32Array(max),
    velX: new Float32Array(max),
    velY: new Float32Array(max),
    velZ: new Float32Array(max),
    accX: new Float32Array(max),
    accY: new Float32Array(max),
    accZ: new Float32Array(max),
    age: new Float32Array(max),
    lifetime: new Float32Array(max),
    size: new Float32Array(max),
    initialSize: new Float32Array(max),
    rotation: new Float32Array(max),
    rotationSpeed: new Float32Array(max),
    initialRotationSpeed: new Float32Array(max),
    alpha: new Float32Array(max),
    initialAlpha: new Float32Array(max),
    initialSpeed: new Float32Array(max),
    drag: new Float32Array(max),
    colorR: new Float32Array(max),
    colorG: new Float32Array(max),
    colorB: new Float32Array(max),
    spriteFrame: new Float32Array(max),
    spriteElapsed: new Float32Array(max),
    alive: new Uint8Array(max)
  };
}
var _emitterIdCounter = 0;
var EmitterInstance = class {
  constructor(config, mode) {
    // Public: set by the React component to provide emitter world position
    this.source = null;
    this._mode = "3d";
    this._elapsed = 0;
    this._emissionAccumulator = 0;
    this._aliveCount = 0;
    this._emissionComplete = false;
    // Pre-allocated temporaries (zero GC)
    this._tmpMatrix = new import_three3.Matrix4();
    this._tmpPosition = new import_three3.Vector3();
    this._tmpQuaternion = new import_three3.Quaternion();
    this._tmpScale = new import_three3.Vector3();
    this._tmpColor = new import_three3.Color();
    this._emitterPos = new import_three3.Vector3();
    this._emitterQuat = new import_three3.Quaternion();
    this.id = `particle_${++_emitterIdCounter}`;
    this._mode = mode;
    this._maxParticles = config.maxParticles ?? 1e3;
    this._emission = config.emission ?? "stream";
    this._rate = config.rate ?? 50;
    this._bursts = config.bursts ?? [];
    this._duration = config.duration ?? Infinity;
    this._loop = config.loop ?? true;
    this._startDelay = config.startDelay ?? 0;
    this._shape = config.shape ?? { shape: "point" };
    this._particle = config.particle ?? {};
    this._overLifetime = config.overLifetime ?? {};
    this._billboard = config.billboard ?? true;
    this._space = config.space ?? "world";
    this._sortByDistance = config.sortByDistance ?? false;
    this._spriteSheet = config.spriteSheet ?? null;
    this._onParticleBorn = config.onParticleBorn ?? null;
    this._onParticleDeath = config.onParticleDeath ?? null;
    this._onComplete = config.onComplete ?? null;
    this._emitting = config.autoPlay ?? true;
    this._delayRemaining = this._startDelay;
    this._burstTimers = this._bursts.map(() => 0);
    this._burstCycles = this._bursts.map(() => 0);
    this._freeList = [];
    for (let i = this._maxParticles - 1; i >= 0; i--) {
      this._freeList.push(i);
    }
    this._data = createParticleData(this._maxParticles);
    const geometry = new import_three3.PlaneGeometry(1, 1);
    let texture = null;
    let columns = 1;
    let rows = 1;
    if (this._spriteSheet) {
      texture = loadTexture(this._spriteSheet.texture);
      columns = this._spriteSheet.columns;
      rows = this._spriteSheet.rows;
    } else if (config.texture) {
      texture = loadTexture(config.texture);
    }
    const blendConfig = resolveBlending(config.blendMode ?? "normal");
    this._material = new import_three3.RawShaderMaterial({
      uniforms: {
        uTexture: { value: texture },
        uHasTexture: { value: texture !== null },
        uColumns: { value: columns },
        uRows: { value: rows },
        uBillboard: { value: this._billboard }
      },
      vertexShader: VERTEX_SHADER,
      fragmentShader: FRAGMENT_SHADER,
      transparent: true,
      depthWrite: false,
      side: import_three3.DoubleSide,
      ...blendConfig
    });
    this.mesh = new import_three3.InstancedMesh(geometry, this._material, this._maxParticles);
    this.mesh.count = 0;
    this.mesh.frustumCulled = false;
    this._tmpColor.set(1, 1, 1);
    for (let i = 0; i < this._maxParticles; i++) {
      this.mesh.setColorAt(i, this._tmpColor);
    }
    this._alphaArray = new Float32Array(this._maxParticles);
    this._spriteFrameArray = new Float32Array(this._maxParticles);
    this._rotationArray = new Float32Array(this._maxParticles);
    this._alphaAttr = new import_three3.InstancedBufferAttribute(this._alphaArray, 1);
    this._spriteFrameAttr = new import_three3.InstancedBufferAttribute(
      this._spriteFrameArray,
      1
    );
    this._rotationAttr = new import_three3.InstancedBufferAttribute(this._rotationArray, 1);
    geometry.setAttribute("instanceAlpha", this._alphaAttr);
    geometry.setAttribute("instanceSpriteFrame", this._spriteFrameAttr);
    geometry.setAttribute("instanceRotation", this._rotationAttr);
  }
  // ── Public Controls ──
  start() {
    this._emitting = true;
    this._emissionComplete = false;
  }
  stop() {
    this._emitting = false;
  }
  clear() {
    this._emitting = false;
    for (let i = 0; i < this._maxParticles; i++) {
      if (this._data.alive[i]) {
        this._data.alive[i] = 0;
        this._freeList.push(i);
      }
    }
    this._aliveCount = 0;
    this.mesh.count = 0;
  }
  reset() {
    this.clear();
    this._elapsed = 0;
    this._emissionAccumulator = 0;
    this._delayRemaining = this._startDelay;
    this._emissionComplete = false;
    for (let b = 0; b < this._bursts.length; b++) {
      this._burstTimers[b] = 0;
      this._burstCycles[b] = 0;
    }
    this._emitting = true;
  }
  burst(count) {
    const c = count ?? 30;
    this._emit(typeof c === "number" ? c : sampleValueRange(c));
  }
  setRate(rate) {
    this._rate = rate;
  }
  getActiveCount() {
    return this._aliveCount;
  }
  isEmitting() {
    return this._emitting;
  }
  setMode(mode) {
    this._mode = mode;
  }
  // ── Per-Frame Tick ──
  tick(delta) {
    if (this.source) {
      if (this._space === "world") {
        this.source.getWorldPosition(this._emitterPos);
        this.source.getWorldQuaternion(this._emitterQuat);
      }
    }
    if (this._delayRemaining > 0) {
      this._delayRemaining -= delta;
      if (this._delayRemaining > 0) {
        this._updatePhase(delta);
        this._writePhase();
        return;
      }
      delta = -this._delayRemaining;
      this._delayRemaining = 0;
    }
    if (this._emitting) {
      this._emitPhase(delta);
    }
    this._updatePhase(delta);
    this._writePhase();
    if (!this._emitting && this._aliveCount === 0 && !this._emissionComplete) {
      this._emissionComplete = true;
      this._onComplete?.();
    }
  }
  // ── Emission Phase ──
  _emitPhase(delta) {
    this._elapsed += delta;
    if (this._emission === "stream") {
      const rate = sampleValueRange(this._rate);
      this._emissionAccumulator += rate * delta;
      const count = Math.floor(this._emissionAccumulator);
      if (count > 0) {
        this._emissionAccumulator -= count;
        this._emit(count);
      }
    } else {
      for (let b = 0; b < this._bursts.length; b++) {
        const burst = this._bursts[b];
        const time = burst.time ?? 0;
        const cycles = burst.cycles ?? 1;
        const interval = burst.interval ?? 1;
        if (this._elapsed < time) continue;
        const timeSinceBurstStart = this._elapsed - time;
        const expectedCycle = Math.floor(timeSinceBurstStart / interval);
        const maxCycles = cycles === 0 ? Infinity : cycles;
        while (this._burstCycles[b] <= expectedCycle && this._burstCycles[b] < maxCycles) {
          const count = Math.round(sampleValueRange(burst.count));
          this._emit(count);
          this._burstCycles[b]++;
        }
      }
    }
    if (this._duration !== Infinity && this._elapsed >= this._duration) {
      if (this._loop) {
        this._elapsed -= this._duration;
        for (let b = 0; b < this._bursts.length; b++) {
          this._burstTimers[b] = 0;
          this._burstCycles[b] = 0;
        }
      } else {
        this._emitting = false;
      }
    }
  }
  // ── Emit N Particles ──
  _emit(count) {
    const d = this._data;
    const p = this._particle;
    const is2D = this._mode === "2d";
    for (let n = 0; n < count; n++) {
      if (this._freeList.length === 0) return;
      const i = this._freeList.pop();
      const [sx, sy, sz, dx, dy, dz] = this._sampleShape(is2D);
      if (this._space === "world" && this.source) {
        this._tmpPosition.set(sx, sy, sz);
        this._tmpPosition.applyQuaternion(this._emitterQuat);
        this._tmpPosition.add(this._emitterPos);
        d.posX[i] = this._tmpPosition.x;
        d.posY[i] = this._tmpPosition.y;
        d.posZ[i] = this._tmpPosition.z;
      } else {
        d.posX[i] = sx;
        d.posY[i] = sy;
        d.posZ[i] = sz;
      }
      const speed = sampleValueRange(p.speed ?? 5);
      d.initialSpeed[i] = speed;
      if (this._space === "world" && this.source) {
        this._tmpPosition.set(dx, dy, dz);
        this._tmpPosition.applyQuaternion(this._emitterQuat);
        d.velX[i] = this._tmpPosition.x * speed;
        d.velY[i] = this._tmpPosition.y * speed;
        d.velZ[i] = this._tmpPosition.z * speed;
      } else {
        d.velX[i] = dx * speed;
        d.velY[i] = dy * speed;
        d.velZ[i] = dz * speed;
      }
      const acc = p.acceleration ?? [0, 0, 0];
      const gravity = p.gravity ?? 0;
      d.accX[i] = acc[0];
      d.accY[i] = acc[1] - gravity;
      d.accZ[i] = acc[2];
      d.age[i] = 0;
      d.lifetime[i] = Math.max(1e-3, sampleValueRange(p.lifetime ?? 1));
      const size = sampleValueRange(p.size ?? 1);
      d.size[i] = size;
      d.initialSize[i] = size;
      d.rotation[i] = sampleValueRange(p.rotation ?? 0);
      const rotSpeed = sampleValueRange(p.rotationSpeed ?? 0);
      d.rotationSpeed[i] = rotSpeed;
      d.initialRotationSpeed[i] = rotSpeed;
      const alpha = sampleValueRange(p.alpha ?? 1);
      d.alpha[i] = alpha;
      d.initialAlpha[i] = alpha;
      d.drag[i] = p.drag ?? 0;
      const [r, g, b] = sampleColorRange(p.color ?? "#ffffff");
      d.colorR[i] = r;
      d.colorG[i] = g;
      d.colorB[i] = b;
      if (this._spriteSheet) {
        const totalFrames = this._spriteSheet.totalFrames ?? this._spriteSheet.columns * this._spriteSheet.rows;
        d.spriteFrame[i] = this._spriteSheet.randomStart ? Math.floor(Math.random() * totalFrames) : this._spriteSheet.startFrame ?? 0;
        d.spriteElapsed[i] = 0;
      } else {
        d.spriteFrame[i] = 0;
        d.spriteElapsed[i] = 0;
      }
      d.alive[i] = 1;
      this._aliveCount++;
      this._onParticleBorn?.(i);
    }
  }
  // ── Update Phase ──
  _updatePhase(delta) {
    const d = this._data;
    const ol = this._overLifetime;
    for (let i = 0; i < this._maxParticles; i++) {
      if (!d.alive[i]) continue;
      d.age[i] += delta;
      if (d.age[i] >= d.lifetime[i]) {
        d.alive[i] = 0;
        this._aliveCount--;
        this._freeList.push(i);
        this._onParticleDeath?.(i);
        continue;
      }
      const t = d.age[i] / d.lifetime[i];
      const drag = d.drag[i];
      if (drag > 0) {
        const factor = 1 - drag * delta;
        d.velX[i] *= factor;
        d.velY[i] *= factor;
        d.velZ[i] *= factor;
      }
      d.velX[i] += d.accX[i] * delta;
      d.velY[i] += d.accY[i] * delta;
      d.velZ[i] += d.accZ[i] * delta;
      if (ol.speed) {
        const speedMul = evaluateCurve(ol.speed, t);
        const currentSpeed = Math.sqrt(
          d.velX[i] * d.velX[i] + d.velY[i] * d.velY[i] + d.velZ[i] * d.velZ[i]
        );
        if (currentSpeed > 1e-4) {
          const targetSpeed = d.initialSpeed[i] * speedMul;
          const scale = targetSpeed / currentSpeed;
          d.velX[i] *= scale;
          d.velY[i] *= scale;
          d.velZ[i] *= scale;
        }
      }
      d.posX[i] += d.velX[i] * delta;
      d.posY[i] += d.velY[i] * delta;
      d.posZ[i] += d.velZ[i] * delta;
      if (ol.size) {
        d.size[i] = d.initialSize[i] * evaluateCurve(ol.size, t);
      }
      if (ol.alpha) {
        d.alpha[i] = d.initialAlpha[i] * evaluateCurve(ol.alpha, t);
      }
      if (ol.color) {
        const [cr, cg, cb] = evaluateColorGradient(ol.color, t);
        d.colorR[i] = cr;
        d.colorG[i] = cg;
        d.colorB[i] = cb;
      }
      if (ol.rotationSpeed) {
        d.rotationSpeed[i] = d.initialRotationSpeed[i] * evaluateCurve(ol.rotationSpeed, t);
      }
      d.rotation[i] += d.rotationSpeed[i] * delta;
      if (this._spriteSheet) {
        const fps = this._spriteSheet.fps ?? 30;
        const totalFrames = this._spriteSheet.totalFrames ?? this._spriteSheet.columns * this._spriteSheet.rows;
        d.spriteElapsed[i] += delta;
        const frameAdvance = Math.floor(d.spriteElapsed[i] * fps);
        if (frameAdvance > 0) {
          d.spriteElapsed[i] -= frameAdvance / fps;
          let newFrame = d.spriteFrame[i] + frameAdvance;
          if (this._spriteSheet.loop ?? true) {
            newFrame = newFrame % totalFrames;
          } else {
            newFrame = Math.min(newFrame, totalFrames - 1);
          }
          d.spriteFrame[i] = newFrame;
        }
      }
    }
  }
  // ── Write GPU Buffers ──
  _writePhase() {
    if (this._aliveCount === 0) {
      this.mesh.count = 0;
      this.mesh.visible = false;
      return;
    }
    this.mesh.visible = true;
    const d = this._data;
    let writeIndex = 0;
    this._tmpQuaternion.identity();
    for (let i = 0; i < this._maxParticles; i++) {
      if (!d.alive[i]) continue;
      this._tmpPosition.set(d.posX[i], d.posY[i], d.posZ[i]);
      const s = d.size[i];
      this._tmpScale.set(s, s, s);
      this._tmpMatrix.compose(
        this._tmpPosition,
        this._tmpQuaternion,
        this._tmpScale
      );
      this.mesh.setMatrixAt(writeIndex, this._tmpMatrix);
      this._tmpColor.setRGB(d.colorR[i], d.colorG[i], d.colorB[i]);
      this.mesh.setColorAt(writeIndex, this._tmpColor);
      this._alphaArray[writeIndex] = d.alpha[i];
      this._spriteFrameArray[writeIndex] = d.spriteFrame[i];
      this._rotationArray[writeIndex] = d.rotation[i];
      writeIndex++;
    }
    this.mesh.count = writeIndex;
    this.mesh.instanceMatrix.needsUpdate = true;
    if (this.mesh.instanceColor) this.mesh.instanceColor.needsUpdate = true;
    this._alphaAttr.needsUpdate = true;
    this._spriteFrameAttr.needsUpdate = true;
    this._rotationAttr.needsUpdate = true;
  }
  // ── Shape Sampling ──
  _sampleShape(is2D) {
    const cfg = this._shape;
    switch (cfg.shape) {
      case "point":
        return [0, 0, 0, 0, 1, 0];
      case "cone": {
        const angle = cfg.angle ?? Math.PI / 4;
        const radius = cfg.radius ?? 1;
        const surface = cfg.surface ?? false;
        const theta = Math.random() * Math.PI * 2;
        const phi = Math.random() * angle;
        const r = surface ? radius : Math.random() * radius;
        const sinPhi = Math.sin(phi);
        const cosPhi = Math.cos(phi);
        const x = r * sinPhi * Math.cos(theta);
        const z = is2D ? 0 : r * sinPhi * Math.sin(theta);
        const y = r * cosPhi;
        const dx = sinPhi * Math.cos(theta);
        const dz = is2D ? 0 : sinPhi * Math.sin(theta);
        const dy = cosPhi;
        return [x, y, z, dx, dy, dz];
      }
      case "sphere": {
        const radius = cfg.radius ?? 1;
        const surface = cfg.surface ?? false;
        const theta = Math.random() * Math.PI * 2;
        const phi = Math.acos(2 * Math.random() - 1);
        const r = surface ? radius : Math.cbrt(Math.random()) * radius;
        const sinPhi = Math.sin(phi);
        const x = r * sinPhi * Math.cos(theta);
        const y = r * Math.cos(phi);
        const z = is2D ? 0 : r * sinPhi * Math.sin(theta);
        const len = Math.sqrt(x * x + y * y + z * z) || 1;
        return [x, y, z, x / len, y / len, z / len];
      }
      case "rectangle": {
        const w = cfg.width ?? 1;
        const h = cfg.height ?? 1;
        const x = (Math.random() - 0.5) * w;
        const y = (Math.random() - 0.5) * h;
        return [x, y, 0, 0, 0, is2D ? 0 : 1];
      }
      case "edge": {
        const from = cfg.from ?? [
          -0.5,
          0,
          0
        ];
        const to = cfg.to ?? [
          0.5,
          0,
          0
        ];
        const t = Math.random();
        const x = from[0] + (to[0] - from[0]) * t;
        const y = from[1] + (to[1] - from[1]) * t;
        const z = from[2] + (to[2] - from[2]) * t;
        return [x, y, z, 0, 1, 0];
      }
      case "ring": {
        const outerR = cfg.radius ?? 1;
        const innerR = cfg.innerRadius ?? outerR * 0.8;
        const theta = Math.random() * Math.PI * 2;
        const r = innerR + Math.random() * (outerR - innerR);
        const x = r * Math.cos(theta);
        const z = is2D ? 0 : r * Math.sin(theta);
        const y = is2D ? r * Math.sin(theta) : 0;
        const len = Math.sqrt(x * x + y * y + z * z) || 1;
        return [x, y, z, x / len, y / len, z / len];
      }
    }
  }
  // ── Dispose ──
  dispose() {
    this.mesh.geometry.dispose();
    this._material.dispose();
    if (this._material.uniforms.uTexture.value) {
      const tex = this._material.uniforms.uTexture.value;
      tex.dispose();
    }
    if (this.mesh.parent) {
      this.mesh.parent.remove(this.mesh);
    }
  }
};
var ParticleManager = class {
  constructor() {
    this._emitters = /* @__PURE__ */ new Map();
    this._mode = "3d";
  }
  setMode(mode) {
    this._mode = mode;
    for (const emitter of this._emitters.values()) {
      emitter.setMode(mode);
    }
  }
  createEmitter(config) {
    const emitter = new EmitterInstance(config, this._mode);
    this._emitters.set(emitter.id, emitter);
    return emitter.id;
  }
  destroyEmitter(id) {
    const emitter = this._emitters.get(id);
    if (!emitter) return;
    emitter.dispose();
    this._emitters.delete(id);
  }
  getEmitter(id) {
    return this._emitters.get(id);
  }
  tick(delta) {
    for (const emitter of this._emitters.values()) {
      emitter.tick(delta);
    }
  }
  dispose() {
    for (const emitter of this._emitters.values()) {
      emitter.dispose();
    }
    this._emitters.clear();
  }
};
var _instance9 = null;
function getParticleManager() {
  if (!_instance9) _instance9 = new ParticleManager();
  return _instance9;
}
function destroyParticleManager() {
  _instance9?.dispose();
  _instance9 = null;
}

// src/systems/ParticlePresets.ts
var PRESETS = {
  fire: {
    maxParticles: 500,
    emission: "stream",
    rate: 80,
    shape: { shape: "cone", angle: Math.PI / 8, radius: 0.3 },
    particle: {
      speed: [2, 5],
      lifetime: [0.5, 1.5],
      size: [0.3, 0.8],
      color: ["#ff4400", "#ffaa00"],
      alpha: 1,
      gravity: -1,
      drag: 0.05
    },
    overLifetime: {
      size: [
        { t: 0, value: 0.5 },
        { t: 0.3, value: 1 },
        { t: 1, value: 0 }
      ],
      alpha: [
        { t: 0, value: 1 },
        { t: 0.6, value: 0.8 },
        { t: 1, value: 0 }
      ],
      color: [
        { t: 0, color: "#ffffff" },
        { t: 0.2, color: "#ffaa00" },
        { t: 0.5, color: "#ff4400" },
        { t: 1, color: "#220000" }
      ]
    },
    blendMode: "additive",
    billboard: true
  },
  smoke: {
    maxParticles: 300,
    emission: "stream",
    rate: 30,
    shape: { shape: "cone", angle: Math.PI / 10, radius: 0.2 },
    particle: {
      speed: [1, 3],
      lifetime: [2, 4],
      size: [0.5, 1.5],
      color: ["#888888", "#aaaaaa"],
      alpha: 0.6,
      gravity: -0.5,
      drag: 0.1,
      rotationSpeed: [-0.5, 0.5]
    },
    overLifetime: {
      size: [
        { t: 0, value: 0.5 },
        { t: 1, value: 2.5 }
      ],
      alpha: [
        { t: 0, value: 0 },
        { t: 0.1, value: 1 },
        { t: 0.6, value: 0.6 },
        { t: 1, value: 0 }
      ]
    },
    blendMode: "normal",
    billboard: true
  },
  explosion: {
    maxParticles: 200,
    emission: "burst",
    bursts: [
      { time: 0, count: [60, 100] },
      { time: 0.05, count: [30, 50] }
    ],
    duration: 0,
    loop: false,
    shape: { shape: "sphere", radius: 0.2, surface: true },
    particle: {
      speed: [5, 15],
      lifetime: [0.3, 1],
      size: [0.2, 0.6],
      color: ["#ff8800", "#ffff00"],
      alpha: 1,
      drag: 0.15,
      gravity: 3
    },
    overLifetime: {
      size: [
        { t: 0, value: 1 },
        { t: 0.3, value: 1.5 },
        { t: 1, value: 0 }
      ],
      alpha: [
        { t: 0, value: 1 },
        { t: 0.5, value: 0.8 },
        { t: 1, value: 0 }
      ],
      color: [
        { t: 0, color: "#ffffff" },
        { t: 0.2, color: "#ffaa00" },
        { t: 0.6, color: "#ff4400" },
        { t: 1, color: "#330000" }
      ]
    },
    blendMode: "additive",
    billboard: true
  },
  sparks: {
    maxParticles: 300,
    emission: "burst",
    bursts: [{ time: 0, count: [20, 50] }],
    duration: 0,
    loop: false,
    shape: { shape: "point" },
    particle: {
      speed: [3, 12],
      lifetime: [0.2, 0.8],
      size: [0.02, 0.08],
      color: ["#ffcc00", "#ff8800"],
      alpha: 1,
      gravity: 5,
      drag: 0.02
    },
    overLifetime: {
      alpha: [
        { t: 0, value: 1 },
        { t: 0.8, value: 1 },
        { t: 1, value: 0 }
      ]
    },
    blendMode: "additive",
    billboard: true
  },
  rain: {
    maxParticles: 2e3,
    emission: "stream",
    rate: 500,
    shape: { shape: "rectangle", width: 30, height: 0 },
    particle: {
      speed: [8, 12],
      lifetime: [1, 2],
      size: [0.01, 0.03],
      color: "#aaccff",
      alpha: 0.4,
      acceleration: [0, -15, 0]
    },
    overLifetime: {
      alpha: [
        { t: 0, value: 0 },
        { t: 0.05, value: 1 },
        { t: 0.9, value: 1 },
        { t: 1, value: 0 }
      ]
    },
    blendMode: "additive",
    billboard: false
  },
  snow: {
    maxParticles: 1e3,
    emission: "stream",
    rate: 100,
    shape: { shape: "rectangle", width: 20, height: 0 },
    particle: {
      speed: [0.5, 2],
      lifetime: [3, 6],
      size: [0.05, 0.15],
      color: "#ffffff",
      alpha: 0.8,
      acceleration: [0, -1, 0],
      rotationSpeed: [-1, 1]
    },
    overLifetime: {
      alpha: [
        { t: 0, value: 0 },
        { t: 0.1, value: 1 },
        { t: 0.8, value: 1 },
        { t: 1, value: 0 }
      ]
    },
    blendMode: "normal",
    billboard: true
  },
  magic: {
    maxParticles: 400,
    emission: "stream",
    rate: 60,
    shape: { shape: "sphere", radius: 0.5 },
    particle: {
      speed: [0.5, 2],
      lifetime: [1, 2],
      size: [0.05, 0.2],
      color: ["#8844ff", "#44aaff"],
      alpha: 1,
      drag: 0.1,
      rotationSpeed: [-2, 2]
    },
    overLifetime: {
      size: [
        { t: 0, value: 0 },
        { t: 0.2, value: 1 },
        { t: 0.8, value: 1 },
        { t: 1, value: 0 }
      ],
      alpha: [
        { t: 0, value: 0 },
        { t: 0.15, value: 1 },
        { t: 0.7, value: 0.8 },
        { t: 1, value: 0 }
      ],
      color: [
        { t: 0, color: "#8844ff" },
        { t: 0.5, color: "#44aaff" },
        { t: 1, color: "#8844ff" }
      ]
    },
    blendMode: "additive",
    billboard: true
  },
  confetti: {
    maxParticles: 500,
    emission: "burst",
    bursts: [{ time: 0, count: [80, 150] }],
    duration: 0,
    loop: false,
    shape: { shape: "cone", angle: Math.PI / 3, radius: 0.5 },
    particle: {
      speed: [5, 12],
      lifetime: [2, 4],
      size: [0.1, 0.3],
      color: ["#ff0000", "#00ff00"],
      alpha: 1,
      gravity: 4,
      drag: 0.05,
      rotation: [0, Math.PI * 2],
      rotationSpeed: [-5, 5]
    },
    overLifetime: {
      alpha: [
        { t: 0, value: 1 },
        { t: 0.8, value: 1 },
        { t: 1, value: 0 }
      ]
    },
    blendMode: "normal",
    billboard: false
  }
};
var _customPresets = /* @__PURE__ */ new Map();
function getParticlePreset(name) {
  const builtin = PRESETS[name];
  if (builtin) return builtin;
  const custom = _customPresets.get(name);
  if (custom) return custom;
  return {};
}
function registerParticlePreset(name, config) {
  _customPresets.set(name, config);
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Easing,
  cubicBezier,
  destroyActorRegistry,
  destroyAssetManager,
  destroyAudioManager,
  destroyCollisionManager,
  destroyGridCollisionManager,
  destroyInputManager,
  destroyParticleManager,
  destroySceneManager,
  destroyTweenManager,
  detectAssetType,
  getActorRegistry,
  getAssetManager,
  getAudioManager,
  getCollisionManager,
  getGridCollisionManager,
  getInputManager,
  getParticleManager,
  getParticlePreset,
  getSceneManager,
  getTweenManager,
  registerParticlePreset,
  resolveEasing,
  steps,
  usePhysicsContext
});
